// @lit-protocol/vincent-ability-sdk/dist/src/lib/abilityCore/vincentAbility.js
"use strict";
// src/lib/abilityCore/vincentAbility.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.createVincentAbility = createVincentAbility;
const zod_1 = require("zod");
const assertSupportedAbilityVersion_1 = require("../assertSupportedAbilityVersion");
const constants_1 = require("../constants");
const utils_1 = require("../utils");
const abilityContext_1 = require("./abilityConfig/context/abilityContext");
const resultCreators_1 = require("./helpers/resultCreators");
const typeGuards_1 = require("./helpers/typeGuards");
const zod_2 = require("./helpers/zod");
/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * The `createVincentAbility()` method is used to define an ability's lifecycle methods and ensure that arguments provided to the ability's
 * lifecycle methods, as well as their return values, are validated and fully type-safe by defining ZOD schemas for them.
 *
 *```typescript
 * const exampleSimpleAbility = createVincentAbility({
 *     packageName: '@lit-protocol/yesability@1.0.0',
 *     abilityDescription: 'Yes Ability description',
 *     abilityParamsSchema: testSchema,
 *     supportedPolicies: supportedPoliciesForAbility([testPolicy]),
 *
 *     precheck: async (params, { succeed, fail }) => {
 *       // Should allow succeed() with no arguments
 *       succeed();
 *
 *       // Should allow fail() with string error
 *       fail('Error message');
 *
 *       // @ts-expect-error - Should not allow succeed() with arguments when no schema
 *       succeed({ message: 'test' });
 *
 *       // @ts-expect-error - Should not allow fail() with object when no schema
 *       fail({ error: 'test' });
 *
 *       return succeed();
 *     },
 *
 *     execute: async (params, { succeed }) => {
 *       // Should allow succeed() with no arguments
 *       succeed();
 *
 *       // @ts-expect-error - Should not allow succeed() with arguments when no schema
 *       return succeed({ data: 'test' });
 *     },
 *   });
 * ```
 *
 * @typeParam AbilityParamsSchema {@removeTypeParameterCompletely}
 * @typeParam PkgNames {@removeTypeParameterCompletely}
 * @typeParam PolicyMap {@removeTypeParameterCompletely}
 * @typeParam PolicyMapByPackageName {@removeTypeParameterCompletely}
 * @typeParam PrecheckSuccessSchema {@removeTypeParameterCompletely}
 * @typeParam PrecheckFailSchema {@removeTypeParameterCompletely}
 * @typeParam ExecuteSuccessSchema {@removeTypeParameterCompletely}
 * @typeParam ExecuteFailSchema {@removeTypeParameterCompletely}
 *
 * @category API Methods
 */
function createVincentAbility(AbilityConfig) {
    const { policyByPackageName, policyByIpfsCid } = AbilityConfig.supportedPolicies;
    for (const policyId in policyByIpfsCid) {
        const policy = policyByIpfsCid[policyId];
        const { vincentAbilityApiVersion } = policy;
        (0, assertSupportedAbilityVersion_1.assertSupportedAbilityVersion)(vincentAbilityApiVersion);
    }
    const executeSuccessSchema = (AbilityConfig.executeSuccessSchema ??
        zod_1.z.undefined());
    const executeFailSchema = (AbilityConfig.executeFailSchema ?? zod_1.z.undefined());
    const execute = async ({ abilityParams }, baseAbilityContext) => {
        try {
            const context = (0, abilityContext_1.createExecutionAbilityContext)({
                baseContext: baseAbilityContext,
                policiesByPackageName: policyByPackageName,
            });
            const parsedAbilityParams = (0, zod_2.validateOrFail)(abilityParams, AbilityConfig.abilityParamsSchema, 'execute', 'input');
            if ((0, typeGuards_1.isAbilityFailureResult)(parsedAbilityParams)) {
                // In this case, we have an invalid input to the ability -- return { success: fail, runtimeError, schemaValidationError }
                return parsedAbilityParams;
            }
            const result = await AbilityConfig.execute({ abilityParams: parsedAbilityParams }, {
                ...context,
                policiesContext: { ...context.policiesContext, allow: true },
            });
            console.log('AbilityConfig execute result', JSON.stringify(result, utils_1.bigintReplacer));
            const { schemaToUse } = (0, zod_2.getSchemaForAbilityResult)({
                value: result,
                successResultSchema: executeSuccessSchema,
                failureResultSchema: executeFailSchema,
            });
            const resultOrFailure = (0, zod_2.validateOrFail)(result.result, schemaToUse, 'execute', 'output');
            if ((0, typeGuards_1.isAbilityFailureResult)(resultOrFailure)) {
                return resultOrFailure;
            }
            // We parsed the result -- it may be a success or a failure; return appropriately.
            if ((0, typeGuards_1.isAbilityFailureResult)(result)) {
                return (0, resultCreators_1.wrapFailure)(resultOrFailure);
            }
            return (0, resultCreators_1.wrapSuccess)(resultOrFailure);
        }
        catch (err) {
            return (0, resultCreators_1.wrapNoResultFailure)(err instanceof Error ? err.message : 'Unknown error');
        }
    };
    const precheckSuccessSchema = (AbilityConfig.precheckSuccessSchema ??
        zod_1.z.undefined());
    const precheckFailSchema = (AbilityConfig.precheckFailSchema ??
        zod_1.z.undefined());
    const { precheck: precheckFn } = AbilityConfig;
    const precheck = precheckFn
        ? (async ({ abilityParams }, baseAbilityContext) => {
            try {
                const context = (0, abilityContext_1.createPrecheckAbilityContext)({
                    baseContext: baseAbilityContext,
                });
                const parsedAbilityParams = (0, zod_2.validateOrFail)(abilityParams, AbilityConfig.abilityParamsSchema, 'precheck', 'input');
                if ((0, typeGuards_1.isAbilityFailureResult)(parsedAbilityParams)) {
                    return parsedAbilityParams;
                }
                const result = await precheckFn({ abilityParams }, context);
                console.log('AbilityConfig precheck result', JSON.stringify(result, utils_1.bigintReplacer));
                const { schemaToUse } = (0, zod_2.getSchemaForAbilityResult)({
                    value: result,
                    successResultSchema: precheckSuccessSchema,
                    failureResultSchema: precheckFailSchema,
                });
                const resultOrFailure = (0, zod_2.validateOrFail)(result.result, schemaToUse, 'precheck', 'output');
                if ((0, typeGuards_1.isAbilityFailureResult)(resultOrFailure)) {
                    return resultOrFailure;
                }
                // We parsed the result successfully -- it may be a success or a failure, return appropriately
                if ((0, typeGuards_1.isAbilityFailureResult)(result)) {
                    return (0, resultCreators_1.wrapFailure)(resultOrFailure, result.runtimeError);
                }
                return (0, resultCreators_1.wrapSuccess)(resultOrFailure);
            }
            catch (err) {
                return (0, resultCreators_1.wrapNoResultFailure)(err instanceof Error ? err.message : 'Unknown error');
            }
        })
        : undefined;
    return {
        packageName: AbilityConfig.packageName,
        vincentAbilityApiVersion: constants_1.VINCENT_TOOL_API_VERSION,
        abilityDescription: AbilityConfig.abilityDescription,
        execute,
        precheck,
        supportedPolicies: AbilityConfig.supportedPolicies,
        policyByPackageName,
        abilityParamsSchema: AbilityConfig.abilityParamsSchema,
        /** @hidden */
        __schemaTypes: {
            precheckSuccessSchema: AbilityConfig.precheckSuccessSchema,
            precheckFailSchema: AbilityConfig.precheckFailSchema,
            executeSuccessSchema: AbilityConfig.executeSuccessSchema,
            executeFailSchema: AbilityConfig.executeFailSchema,
        },
    };
}
