// @lit-protocol/vincent-contracts-sdk/dist/src/fees/signOwnerAttestation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signOwnerAttestation = signOwnerAttestation;
exports.getBaseSepoliaFeeDiamondAddress = getBaseSepoliaFeeDiamondAddress;
const constants_1 = require("../constants");
// Chronicle Yellowstone chain ID
const CHRONICLE_YELLOWSTONE_CHAIN_ID = 175188;
/**
 * Signs an owner attestation using a Lit Action
 *
 * This function executes a Lit Action that:
 * 1. Validates the issuedAt timestamp is within ±30 seconds of current time
 * 2. Verifies the owner is actually the app manager by reading from the Vincent Diamond on Chronicle Yellowstone
 * 3. Creates an OwnerAttestation structure with the provided parameters
 * 4. Signs it using the specified PKP
 * 5. Returns the signature
 *
 * The signature can then be used to call `withdrawAppFees` on the Fee Diamond contract.
 * The Lit Action automatically retrieves the Chronicle Yellowstone RPC URL using Lit.Actions.getRpcUrl().
 *
 * IMPORTANT: This function generates the `issuedAt` timestamp and passes it to all Lit nodes.
 * This ensures all nodes sign the exact same message, which is required for signature combining.
 *
 * @param params - Parameters for signing the owner attestation
 * @returns Promise resolving to the signature and attestation data
 * @throws Error if the Lit Action execution fails or if ownership verification fails
 *
 * @example
 * ```typescript
 * const result = await signOwnerAttestation({
 *   litNodeClient,
 *   sessionSigs,
 *   appId: 12345,
 *   owner: '0x...',
 *   dstChainId: 84532, // Base Sepolia
 *   dstContract: '0x...', // Fee Diamond address
 *   litActionIpfsCid: 'Qm...',
 * });
 *
 * // Use the signature to withdraw fees
 * await feeAdminFacet.withdrawAppFees(
 *   result.attestation.appId,
 *   tokenAddress,
 *   result.attestation,
 *   result.signature,
 * );
 * ```
 */
async function signOwnerAttestation({ litNodeClient, sessionSigs, appId, owner, dstChainId, dstContract, litActionIpfsCid, srcChainId = CHRONICLE_YELLOWSTONE_CHAIN_ID, srcContract = constants_1.VINCENT_DIAMOND_CONTRACT_ADDRESS_PROD, }) {
    // Generate issuedAt timestamp here so all nodes use the same value
    // This is critical for signature combining across the 10 parallel Lit nodes
    const issuedAt = Math.floor(Date.now() / 1000);
    const result = await litNodeClient.executeJs({
        sessionSigs,
        ipfsId: litActionIpfsCid,
        jsParams: {
            srcChainId,
            srcContract,
            owner,
            appId,
            issuedAt,
            dstChainId,
            dstContract,
        },
    });
    // Check for errors in the response
    if (!result.response) {
        throw new Error('No response from Lit Action');
    }
    const response = result.response;
    // Check if the response indicates an error
    if (response.startsWith('Error:')) {
        throw new Error(response);
    }
    try {
        const parsedResult = JSON.parse(response);
        return parsedResult;
    }
    catch {
        throw new Error(`Failed to parse Lit Action response: ${response}`);
    }
}
/**
 * Gets the Base Sepolia Fee Diamond address from the contract address book
 */
function getBaseSepoliaFeeDiamondAddress() {
    return constants_1.VINCENT_CONTRACT_ADDRESS_BOOK.fee[84532].address;
}
