// @aa-sdk/core/dist/esm/actions/smartAccount/upgradeAccount.js
import { isBaseSmartAccountClient } from "../../client/isSmartAccountClient.js";
import { AccountNotFoundError } from "../../errors/account.js";
import { IncompatibleClientError } from "../../errors/client.js";
import { sendUserOperation } from "./sendUserOperation.js";
import { waitForUserOperationTransaction } from "./waitForUserOperationTransacation.js";
import { clientHeaderTrack } from "../../index.js";
export const upgradeAccount = async (client_, args) => {
    const client = clientHeaderTrack(client_, "upgradeAccount");
    const { account = client.account, upgradeTo, overrides, waitForTx, context, } = args;
    if (!account) {
        throw new AccountNotFoundError();
    }
    if (!isBaseSmartAccountClient(client)) {
        throw new IncompatibleClientError("BaseSmartAccountClient", "upgradeAccount", client);
    }
    const { implAddress: accountImplAddress, initializationData } = upgradeTo;
    const encodeUpgradeData = await account.encodeUpgradeToAndCall({
        upgradeToAddress: accountImplAddress,
        upgradeToInitData: initializationData,
    });
    const result = await sendUserOperation(client, {
        uo: {
            target: account.address,
            data: encodeUpgradeData,
        },
        account,
        overrides,
        context,
    });
    let hash = result.hash;
    if (waitForTx) {
        hash = await waitForUserOperationTransaction(client, result);
    }
    return hash;
};
