// @lit-protocol/vincent-ability-sdk/dist/src/lib/policyCore/policyParameters/getOnchainPolicyParams.js
"use strict";
// src/lib/policyCore/policyParameters/getOnchainPolicyParams.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPoliciesAndAppVersion = exports.getDecodedPolicyParams = void 0;
const ethers_1 = require("ethers");
const vincent_contracts_sdk_1 = require("@lit-protocol/vincent-contracts-sdk");
const utils_1 = require("../../utils");
const getDecodedPolicyParams = async ({ decodedPolicies, policyIpfsCid, }) => {
    console.log('All on-chain policy params:', JSON.stringify(decodedPolicies, utils_1.bigintReplacer));
    const policyParams = decodedPolicies[policyIpfsCid];
    if (policyParams) {
        return policyParams;
    }
    console.log('Found no on-chain parameters for policy IPFS CID:', policyIpfsCid);
    return undefined;
};
exports.getDecodedPolicyParams = getDecodedPolicyParams;
const getPoliciesAndAppVersion = async ({ registryRpcUrl, appDelegateeAddress, agentAddress, abilityIpfsCid, }) => {
    console.log('getPoliciesAndAppVersion', {
        registryRpcUrl,
        appDelegateeAddress,
        agentAddress,
        abilityIpfsCid,
    });
    try {
        // Create a signer using the registryRpcUrl
        const signer = ethers_1.ethers.Wallet.createRandom().connect(new ethers_1.ethers.providers.StaticJsonRpcProvider(registryRpcUrl));
        const contractClient = (0, vincent_contracts_sdk_1.getClient)({
            signer,
        });
        // Use the contracts-sdk to validate ability execution and get policies
        const validationResult = await contractClient.validateAbilityExecutionAndGetPolicies({
            delegateeAddress: appDelegateeAddress,
            agentAddress,
            abilityIpfsCid: abilityIpfsCid,
        });
        // We exit early here because !validationResult.isPermitted means appDelegateeAddress
        // is not permitted to execute abilityIpfsCid for the Vincent App on behalf of the agentWalletPkpTokenId
        // and no further processing is needed
        if (!validationResult.isPermitted) {
            throw new Error(`App Delegatee: ${appDelegateeAddress} is not permitted to execute Vincent Ability: ${abilityIpfsCid} for App ID: ${validationResult.appId} App Version: ${validationResult.appVersion} using Agent Address: ${agentAddress}`);
        }
        return {
            appId: ethers_1.ethers.BigNumber.from(validationResult.appId),
            appVersion: ethers_1.ethers.BigNumber.from(validationResult.appVersion),
            decodedPolicies: validationResult.decodedPolicies,
        };
    }
    catch (error) {
        throw new Error(`Error getting on-chain policy parameters from Vincent contract using App Delegatee: ${appDelegateeAddress} and Agent Address: ${agentAddress} and Vincent Ability: ${abilityIpfsCid}: ${error instanceof Error ? error.message : String(error)}`);
    }
};
exports.getPoliciesAndAppVersion = getPoliciesAndAppVersion;
