// @lit-protocol/vincent-contracts-sdk/dist/src/internal/user/UserView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllRegisteredAgentAddressesForUser = getAllRegisteredAgentAddressesForUser;
exports.getUserAddressForAgent = getUserAddressForAgent;
exports.getPermittedAppForAgents = getPermittedAppForAgents;
exports.getUnpermittedAppForAgents = getUnpermittedAppForAgents;
exports.getAllAbilitiesAndPoliciesForApp = getAllAbilitiesAndPoliciesForApp;
exports.validateAbilityExecutionAndGetPolicies = validateAbilityExecutionAndGetPolicies;
exports.isDelegateePermitted = isDelegateePermitted;
const ethers_1 = require("ethers");
const utils_1 = require("../../utils");
const policyParams_1 = require("../../utils/policyParams");
// Matches 0x-prefixed hex strings that are all zeros (e.g., empty bytes32).
const isZeroHex = (value) => /^0x0*$/.test(value);
// Normalizes and checks for the zero address sentinel (0x000...000).
const isZeroAddress = (value) => value.toLowerCase() === ethers_1.ethers.constants.AddressZero.toLowerCase();
// Helper to check if a value is zero (works with both BigNumber and plain number from uint40/uint256)
const isZeroValue = (value) => {
    if (typeof value === 'number')
        return value === 0;
    if (value?.isZero)
        return value.isZero();
    return Number(value) === 0;
};
// Treat all-zero structs from the contract as "no permitted app" sentinel.
const isEmptyPermittedApp = (permittedApp) => isZeroValue(permittedApp.appId) &&
    isZeroValue(permittedApp.version) &&
    isZeroAddress(permittedApp.pkpSigner) &&
    isZeroHex(permittedApp.pkpSignerPubKey);
// Treat all-zero structs from the contract as "no unpermitted app" sentinel.
const isEmptyUnpermittedApp = (unpermittedApp) => isZeroValue(unpermittedApp.appId) &&
    isZeroValue(unpermittedApp.previousPermittedVersion) &&
    isZeroAddress(unpermittedApp.pkpSigner) &&
    isZeroHex(unpermittedApp.pkpSignerPubKey);
async function getAllRegisteredAgentAddressesForUser(params) {
    const { contract, args: { userAddress, offset }, } = params;
    try {
        return await contract.getAllRegisteredAgentAddressesForUser(userAddress, offset);
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        if (decodedError.includes('NoRegisteredAgentsFound')) {
            return [];
        }
        throw new Error(`Failed to Get All Registered Agent Addresses: ${decodedError}`);
    }
}
async function getUserAddressForAgent(params) {
    const { contract, args: { agentAddress }, } = params;
    try {
        return await contract.getUserAddressForAgent(agentAddress);
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        if (decodedError.includes('AgentNotRegistered')) {
            return null;
        }
        throw new Error(`Failed to Get User Address For Agent: ${decodedError}`);
    }
}
async function getPermittedAppForAgents(params) {
    const { contract, args: { agentAddresses }, } = params;
    try {
        const results = await contract.getPermittedAppForAgents(agentAddresses);
        return results.map((result) => {
            const permittedApp = isEmptyPermittedApp(result.permittedApp)
                ? null
                : {
                    appId: Number(result.permittedApp.appId),
                    version: Number(result.permittedApp.version),
                    pkpSigner: result.permittedApp.pkpSigner,
                    pkpSignerPubKey: result.permittedApp.pkpSignerPubKey,
                    versionEnabled: result.permittedApp.versionEnabled,
                    isDeleted: result.permittedApp.isDeleted,
                };
            return {
                agentAddress: result.agentAddress,
                permittedApp,
            };
        });
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Get Permitted App For Agents: ${decodedError}`);
    }
}
async function getUnpermittedAppForAgents(params) {
    const { contract, args: { agentAddresses }, } = params;
    try {
        const results = await contract.getUnpermittedAppForAgents(agentAddresses);
        return results.map((result) => {
            const unpermittedApp = isEmptyUnpermittedApp(result.unpermittedApp)
                ? null
                : {
                    appId: Number(result.unpermittedApp.appId),
                    previousPermittedVersion: Number(result.unpermittedApp.previousPermittedVersion),
                    pkpSigner: result.unpermittedApp.pkpSigner,
                    pkpSignerPubKey: result.unpermittedApp.pkpSignerPubKey,
                    versionEnabled: result.unpermittedApp.versionEnabled,
                    isDeleted: result.unpermittedApp.isDeleted,
                };
            return {
                agentAddress: result.agentAddress,
                unpermittedApp,
            };
        });
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Get Unpermitted App For Agents: ${decodedError}`);
    }
}
async function getAllAbilitiesAndPoliciesForApp(params) {
    const { contract, args: { agentAddress, appId }, } = params;
    try {
        const abilities = await contract.getAllAbilitiesAndPoliciesForApp(agentAddress, appId);
        return (0, policyParams_1.decodePermissionDataFromChain)(abilities);
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Get All Abilities And Policies For App: ${decodedError}`);
    }
}
async function validateAbilityExecutionAndGetPolicies(params) {
    const { contract, args: { delegateeAddress, agentAddress, abilityIpfsCid }, } = params;
    try {
        const validationResult = await contract.validateAbilityExecutionAndGetPolicies(delegateeAddress, agentAddress, abilityIpfsCid);
        const { policies } = validationResult;
        const decodedPolicies = (0, policyParams_1.decodePolicyParametersForOneAbility)({ policies });
        return {
            ...validationResult,
            appId: validationResult.appId,
            appVersion: validationResult.appVersion,
            decodedPolicies,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Validate Ability Execution And Get Policies: ${decodedError}`);
    }
}
async function isDelegateePermitted(params) {
    const { contract, args: { delegateeAddress, agentAddress, abilityIpfsCid }, } = params;
    try {
        // We reuse validateAbilityExecutionAndGetPolicies because the registry view currently exposes
        // only this selector for permission checks, so this keeps the read path aligned with on-chain behavior.
        const validationResult = await contract.validateAbilityExecutionAndGetPolicies(delegateeAddress, agentAddress, abilityIpfsCid);
        return validationResult.isPermitted;
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Check If Delegatee Is Permitted: ${decodedError}`);
    }
}
