// @lit-protocol/vincent-ability-sdk/dist/src/lib/policyCore/vincentPolicy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createVincentPolicy = createVincentPolicy;
exports.createVincentAbilityPolicy = createVincentAbilityPolicy;
// src/lib/policyCore/vincentPolicy.ts
const zod_1 = require("zod");
const assertSupportedAbilityVersion_1 = require("../assertSupportedAbilityVersion");
const utils_1 = require("../utils");
const helpers_1 = require("./helpers");
const resultCreators_1 = require("./helpers/resultCreators");
const policyConfigContext_1 = require("./policyConfig/context/policyConfigContext");
/**
 * The `createVincentPolicy()` method is used to define a policy's lifecycle methods and ensure that arguments provided to the ability's
 * lifecycle methods, as well as their return values, are validated and fully type-safe by defining ZOD schemas for them.
 *
 * @typeParam PackageName {@removeTypeParameterCompletely}
 * @typeParam PolicyAbilityParams {@removeTypeParameterCompletely}
 * @typeParam UserParams {@removeTypeParameterCompletely}
 * @typeParam PrecheckAllowResult {@removeTypeParameterCompletely}
 * @typeParam PrecheckDenyResult {@removeTypeParameterCompletely}
 * @typeParam EvalAllowResult {@removeTypeParameterCompletely}
 * @typeParam EvalDenyResult {@removeTypeParameterCompletely}
 * @typeParam CommitParams {@removeTypeParameterCompletely}
 * @typeParam CommitAllowResult {@removeTypeParameterCompletely}
 * @typeParam CommitDenyResult {@removeTypeParameterCompletely}
 * @category API Methods
 */
function createVincentPolicy(PolicyConfig) {
    if (PolicyConfig.commitParamsSchema && !PolicyConfig.commit) {
        throw new Error('Policy defines commitParamsSchema but is missing commit function');
    }
    const userParamsSchema = (PolicyConfig.userParamsSchema ?? zod_1.z.undefined());
    const evalAllowSchema = (PolicyConfig.evalAllowResultSchema ?? zod_1.z.undefined());
    const evalDenySchema = (PolicyConfig.evalDenyResultSchema ?? zod_1.z.undefined());
    const evaluate = async (args, baseContext) => {
        try {
            const context = (0, policyConfigContext_1.createPolicyContext)({
                baseContext,
            });
            const paramsOrDeny = (0, helpers_1.getValidatedParamsOrDeny)({
                rawAbilityParams: args.abilityParams,
                rawUserParams: args.userParams,
                abilityParamsSchema: PolicyConfig.abilityParamsSchema,
                userParamsSchema: userParamsSchema,
                phase: 'evaluate',
            });
            if ((0, helpers_1.isPolicyDenyResponse)(paramsOrDeny)) {
                return paramsOrDeny;
            }
            const { abilityParams, userParams } = paramsOrDeny;
            const result = await PolicyConfig.evaluate({ abilityParams, userParams }, context);
            const { schemaToUse } = (0, helpers_1.getSchemaForPolicyResponseResult)({
                value: result,
                allowResultSchema: evalAllowSchema,
                denyResultSchema: evalDenySchema,
            });
            const resultOrDeny = (0, helpers_1.validateOrDeny)(result.result, schemaToUse, 'evaluate', 'output');
            if ((0, helpers_1.isPolicyDenyResponse)(resultOrDeny)) {
                return resultOrDeny;
            }
            // We parsed the result -- it may be a success or a failure; return appropriately.
            if ((0, helpers_1.isPolicyDenyResponse)(result)) {
                return (0, helpers_1.createDenyResult)({
                    result: resultOrDeny,
                });
            }
            return (0, resultCreators_1.wrapAllow)(resultOrDeny);
        }
        catch (err) {
            return (0, resultCreators_1.returnNoResultDeny)(err instanceof Error ? err.message : 'Unknown error');
        }
    };
    const precheckAllowSchema = (PolicyConfig.precheckAllowResultSchema ??
        zod_1.z.undefined());
    const precheckDenySchema = (PolicyConfig.precheckDenyResultSchema ??
        zod_1.z.undefined());
    const precheck = PolicyConfig.precheck
        ? (async (args, baseContext) => {
            try {
                const context = (0, policyConfigContext_1.createPolicyContext)({
                    baseContext,
                });
                const { precheck: precheckFn } = PolicyConfig;
                if (!precheckFn) {
                    throw new Error('precheck function unexpectedly missing');
                }
                const paramsOrDeny = (0, helpers_1.getValidatedParamsOrDeny)({
                    rawAbilityParams: args.abilityParams,
                    rawUserParams: args.userParams,
                    abilityParamsSchema: PolicyConfig.abilityParamsSchema,
                    userParamsSchema,
                    phase: 'precheck',
                });
                if ((0, helpers_1.isPolicyDenyResponse)(paramsOrDeny)) {
                    return paramsOrDeny;
                }
                const result = await precheckFn(args, context);
                const { schemaToUse } = (0, helpers_1.getSchemaForPolicyResponseResult)({
                    value: result,
                    allowResultSchema: precheckAllowSchema,
                    denyResultSchema: precheckDenySchema,
                });
                const resultOrDeny = (0, helpers_1.validateOrDeny)(result.result, schemaToUse, 'precheck', 'output');
                if ((0, helpers_1.isPolicyDenyResponse)(resultOrDeny)) {
                    return resultOrDeny;
                }
                // We parsed the result -- it may be a success or a failure; return appropriately.
                if ((0, helpers_1.isPolicyDenyResponse)(result)) {
                    return (0, helpers_1.createDenyResult)({
                        result: resultOrDeny,
                    });
                }
                return (0, resultCreators_1.createAllowResult)({ result: resultOrDeny });
            }
            catch (err) {
                return (0, resultCreators_1.createDenyNoResult)(err instanceof Error ? err.message : 'Unknown error');
            }
        })
        : undefined;
    const commitAllowSchema = (PolicyConfig.commitAllowResultSchema ??
        zod_1.z.undefined());
    const commitDenySchema = (PolicyConfig.commitDenyResultSchema ??
        zod_1.z.undefined());
    const commitParamsSchema = (PolicyConfig.commitParamsSchema ?? zod_1.z.undefined());
    const commit = PolicyConfig.commit
        ? (async (args, baseContext) => {
            try {
                const context = (0, policyConfigContext_1.createPolicyContext)({
                    baseContext,
                });
                const { commit: commitFn } = PolicyConfig;
                if (!commitFn) {
                    throw new Error('commit function unexpectedly missing');
                }
                console.log('commit', JSON.stringify({ args, context }, utils_1.bigintReplacer, 2));
                const paramsOrDeny = (0, helpers_1.validateOrDeny)(args, commitParamsSchema, 'commit', 'input');
                if ((0, helpers_1.isPolicyDenyResponse)(paramsOrDeny)) {
                    return paramsOrDeny;
                }
                const result = await commitFn(args, context);
                const { schemaToUse } = (0, helpers_1.getSchemaForPolicyResponseResult)({
                    value: result,
                    allowResultSchema: commitAllowSchema,
                    denyResultSchema: commitDenySchema,
                });
                const resultOrDeny = (0, helpers_1.validateOrDeny)(result.result, schemaToUse, 'commit', 'output');
                if ((0, helpers_1.isPolicyDenyResponse)(resultOrDeny)) {
                    return resultOrDeny;
                }
                // We parsed the result -- it may be a success or a failure; return appropriately.
                if ((0, helpers_1.isPolicyDenyResponse)(result)) {
                    return (0, helpers_1.createDenyResult)({
                        result: resultOrDeny,
                    });
                }
                return (0, resultCreators_1.createAllowResult)({ result: resultOrDeny });
            }
            catch (err) {
                return (0, resultCreators_1.createDenyNoResult)(err instanceof Error ? err.message : 'Unknown error');
            }
        })
        : undefined;
    const vincentPolicy = {
        ...PolicyConfig,
        evaluate,
        precheck,
        commit,
    };
    return vincentPolicy;
}
/**
 * `createVincentAbilityPolicy()` is used to bind a policy to a specific ability. You must provide a `abilityParameterMappings` argument
 * which instructs the ability which of its abilityParams should be passed to the Vincent Policy during evaluation, and
 * defines what the argument passed to the ability should be.
 *
 * For example, a Ability might receive an argument called `tokenInAmount`, but it may need to pass that as `buyAmount` to a
 * policy that uses the `tokenInAmount` for its own purposes.
 *
 * ```typescript
 * import { bundledVincentPolicy } from '@lit-protocol/vincent-policy-spending-limit';
 *
 * const SpendingLimitPolicy = createVincentAbilityPolicy({
 *   abilityParamsSchema,
 *   bundledVincentPolicy,
 *   abilityParameterMappings: {
 *     rpcUrlForUniswap: 'rpcUrlForUniswap',
 *     chainIdForUniswap: 'chainIdForUniswap',
 *     ethRpcUrl: 'ethRpcUrl',
 *     tokenInAddress: 'tokenAddress',
 *     tokenInDecimals: 'tokenDecimals',
 *     tokenInAmount: 'buyAmount',
 *   },
 * });
 * ```
 *
 * @typeParam PackageName {@removeTypeParameterCompletely}
 * @typeParam IpfsCid {@removeTypeParameterCompletely}
 * @typeParam AbilityParamsSchema {@removeTypeParameterCompletely}
 * @typeParam PolicyAbilityParams {@removeTypeParameterCompletely}
 * @typeParam UserParams {@removeTypeParameterCompletely}
 * @typeParam PrecheckAllowResult {@removeTypeParameterCompletely}
 * @typeParam PrecheckDenyResult {@removeTypeParameterCompletely}
 * @typeParam EvalAllowResult {@removeTypeParameterCompletely}
 * @typeParam EvalDenyResult {@removeTypeParameterCompletely}
 * @typeParam CommitParams {@removeTypeParameterCompletely}
 * @typeParam CommitAllowResult {@removeTypeParameterCompletely}
 * @typeParam CommitDenyResult {@removeTypeParameterCompletely}
 *
 * @returns A Vincent Policy that is configured to work with the provided ability
 * {@displayType A Vincent Policy that is configured to work with the provided ability }
 * @category API Methods
 */
function createVincentAbilityPolicy(config) {
    const { bundledVincentPolicy: { vincentPolicy, ipfsCid, vincentAbilityApiVersion }, } = config;
    (0, assertSupportedAbilityVersion_1.assertSupportedAbilityVersion)(vincentAbilityApiVersion);
    const result = {
        vincentPolicy: vincentPolicy,
        ipfsCid,
        vincentAbilityApiVersion,
        abilityParameterMappings: config.abilityParameterMappings,
        // Explicitly include schema types in the returned object for type inference
        /** @hidden */
        __schemaTypes: {
            policyAbilityParamsSchema: vincentPolicy.abilityParamsSchema,
            userParamsSchema: vincentPolicy.userParamsSchema,
            evalAllowResultSchema: vincentPolicy.evalAllowResultSchema,
            evalDenyResultSchema: vincentPolicy.evalDenyResultSchema,
            commitParamsSchema: vincentPolicy.commitParamsSchema,
            precheckAllowResultSchema: vincentPolicy.precheckAllowResultSchema,
            precheckDenyResultSchema: vincentPolicy.precheckDenyResultSchema,
            commitAllowResultSchema: vincentPolicy.commitAllowResultSchema,
            commitDenyResultSchema: vincentPolicy.commitDenyResultSchema,
            // Explicit function types
            evaluate: vincentPolicy.evaluate,
            precheck: vincentPolicy.precheck,
            commit: vincentPolicy.commit,
        },
    };
    // Use the same type assertion -- but include __schemaTypes to fix generic inference issues
    return result;
}
