// @lit-protocol/vincent-ability-sdk/dist/src/lib/abilityCore/helpers/getPkpInfo.js
"use strict";
// src/lib/abilityCore/helpers/getPkpInfo.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPkpInfo = void 0;
const ethers_1 = require("ethers");
const getPkpInfo = async ({ litPubkeyRouterAddress, yellowstoneRpcUrl, pkpEthAddress, }) => {
    try {
        const PUBKEY_ROUTER_ABI = [
            'function ethAddressToPkpId(address ethAddress) public view returns (uint256)',
            'function getPubkey(uint256 tokenId) public view returns (bytes memory)',
        ];
        const pubkeyRouter = new ethers_1.ethers.Contract(litPubkeyRouterAddress, PUBKEY_ROUTER_ABI, new ethers_1.ethers.providers.StaticJsonRpcProvider(yellowstoneRpcUrl));
        const pkpTokenId = await pubkeyRouter.ethAddressToPkpId(pkpEthAddress);
        const publicKey = await pubkeyRouter.getPubkey(pkpTokenId);
        return {
            tokenId: pkpTokenId.toString(),
            ethAddress: pkpEthAddress,
            publicKey: publicKey.toString('hex'),
        };
    }
    catch (error) {
        throw new Error(`Error getting PKP info for PKP Eth Address: ${pkpEthAddress} using Lit Pubkey Router: ${litPubkeyRouterAddress} and Yellowstone RPC URL: ${yellowstoneRpcUrl}: ${error instanceof Error ? error.message : String(error)}`);
    }
};
exports.getPkpInfo = getPkpInfo;
