// @lit-protocol/vincent-ability-sdk/dist/src/lib/abilityCore/helpers/resultCreators.js
"use strict";
// src/lib/abilityCore/helpers/resultCreators.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAbilitySuccessResult = createAbilitySuccessResult;
exports.createAbilityFailureResult = createAbilityFailureResult;
exports.createAbilityFailureNoResult = createAbilityFailureNoResult;
exports.wrapFailure = wrapFailure;
exports.wrapNoResultFailure = wrapNoResultFailure;
exports.wrapSuccess = wrapSuccess;
exports.wrapNoResultSuccess = wrapNoResultSuccess;
function createAbilitySuccessResult(args) {
    if (!args || args.result === undefined) {
        return { success: true };
    }
    return { success: true, result: args.result };
}
function createAbilityFailureResult({ runtimeError, result, schemaValidationError, }) {
    if (result === undefined) {
        return {
            success: false,
            runtimeError: runtimeError,
            result: undefined,
            ...(schemaValidationError ? { schemaValidationError } : {}),
        };
    }
    return {
        success: false,
        runtimeError: runtimeError,
        result,
        ...(schemaValidationError ? { schemaValidationError } : {}),
    };
}
function createAbilityFailureNoResult(runtimeError, schemaValidationError) {
    return createAbilityFailureResult({ runtimeError, schemaValidationError });
}
function wrapFailure(value, runtimeError, schemaValidationError) {
    return createAbilityFailureResult({ result: value, runtimeError, schemaValidationError });
}
function wrapNoResultFailure(runtimeError, schemaValidationError) {
    return createAbilityFailureNoResult(runtimeError, schemaValidationError);
}
function wrapSuccess(value) {
    return createAbilitySuccessResult({ result: value });
}
function wrapNoResultSuccess() {
    return createAbilitySuccessResult();
}
