// @lit-protocol/vincent-ability-sdk/dist/src/lib/abilityCore/helpers/zod.js
"use strict";
// src/lib/abilityCore/helpers/zod.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbilityResultShape = void 0;
exports.validateOrFail = validateOrFail;
exports.getSchemaForAbilityResult = getSchemaForAbilityResult;
const zod_1 = require("zod");
const resultCreators_1 = require("./resultCreators");
const typeGuards_1 = require("./typeGuards");
/**
 * Matches the minimum structure of a AbilityResult.
 * This is useful when validating that a response shape is at least plausible.
 */
exports.AbilityResultShape = zod_1.z.object({
    success: zod_1.z.boolean(),
    result: zod_1.z.unknown(),
});
/**
 * Used as the default fallback schema when one is missing and the result must be undefined.
 */
const mustBeUndefinedSchema = zod_1.z.undefined();
/**
 * Validates a value using a Zod schema (or requires undefined if none given).
 * Returns parsed result or a standardized failure object.
 *
 * @param value - The raw value to validate
 * @param schema - A Zod schema to apply
 * @param stage - Whether this is input or output validation
 * @param phase - Whether this is 'precheck' or 'execute'
 */
function validateOrFail(value, schema, phase, stage) {
    const effectiveSchema = schema ?? mustBeUndefinedSchema;
    const parsed = effectiveSchema.safeParse(value);
    if (!parsed.success) {
        const descriptor = stage === 'input' ? 'parameters' : 'result';
        const message = `Invalid ${phase} ${descriptor}.`;
        return (0, resultCreators_1.createAbilityFailureResult)({
            runtimeError: message,
            schemaValidationError: {
                zodError: parsed.error,
                phase,
                stage,
            },
        });
    }
    return parsed.data;
}
/**
 * Given an unknown ability response result and the known success/failure schemas,
 * this function returns the appropriate Zod schema to use when validating `.result`.
 *
 * - If the response shape is invalid, returns a Zod schema matching the AbilityResult structure.
 * - If the shape is valid, returns either the success or failure result schema.
 */
function getSchemaForAbilityResult({ value, successResultSchema, failureResultSchema, }) {
    if (!(0, typeGuards_1.isAbilityResult)(value)) {
        return {
            schemaToUse: exports.AbilityResultShape,
            parsedType: 'unknown',
        };
    }
    const schemaToUse = value.success
        ? (successResultSchema ?? zod_1.z.undefined())
        : (failureResultSchema ?? zod_1.z.undefined());
    return {
        schemaToUse,
        parsedType: value.success ? 'success' : 'failure',
    };
}
