// @lit-protocol/vincent-ability-sdk/dist/src/lib/policyCore/helpers/resultCreators.js
"use strict";
// src/lib/policyCore/helpers/resultCreators.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDenyResult = createDenyResult;
exports.createDenyNoResult = createDenyNoResult;
exports.createAllowResult = createAllowResult;
exports.createAllowEvaluationResult = createAllowEvaluationResult;
exports.createDenyEvaluationResult = createDenyEvaluationResult;
exports.wrapAllow = wrapAllow;
exports.wrapDeny = wrapDeny;
exports.returnNoResultDeny = returnNoResultDeny;
exports.isTypedAllowResponse = isTypedAllowResponse;
/**
 * Implementation
 */
function createDenyResult(params) {
    if (params.result === undefined) {
        return {
            allow: false,
            runtimeError: params.runtimeError,
            result: undefined,
            ...(params.schemaValidationError
                ? { schemaValidationError: params.schemaValidationError }
                : {}),
        };
    }
    return {
        allow: false,
        runtimeError: params.runtimeError,
        result: params.result,
        ...(params.schemaValidationError
            ? { schemaValidationError: params.schemaValidationError }
            : {}),
    };
}
function createDenyNoResult(runtimeError, schemaValidationError) {
    return createDenyResult({ runtimeError, schemaValidationError });
}
/**
 * Overload: return an allow response with no result
 */
/**
 * Implementation
 */
function createAllowResult(params) {
    if (params.result === undefined) {
        return {
            allow: true,
            result: undefined,
        };
    }
    return {
        allow: true,
        result: params.result,
    };
}
function createAllowEvaluationResult(params) {
    return {
        allow: true,
        evaluatedPolicies: params.evaluatedPolicies,
        allowedPolicies: params.allowedPolicies,
        deniedPolicy: undefined, // important for union discrimination
    };
}
function createDenyEvaluationResult(params) {
    return {
        allow: false,
        evaluatedPolicies: params.evaluatedPolicies,
        allowedPolicies: params.allowedPolicies,
        deniedPolicy: params.deniedPolicy,
    };
}
// Wraps a validated value as a typed allow result
function wrapAllow(value) {
    return createAllowResult({ result: value });
}
// Wraps a deny result as fully typed (for schema-defined denials)
function wrapDeny(runtimeError, result, schemaValidationError) {
    return createDenyResult({ runtimeError, result, schemaValidationError });
}
// Wraps a schema-less denial into a conditionally valid deny return
function returnNoResultDeny(runtimeError, schemaValidationError) {
    return createDenyNoResult(runtimeError, schemaValidationError);
}
// Optionally: type guard if needed
function isTypedAllowResponse(val) {
    return typeof val === 'object' && val !== null && val.allow === true;
}
