// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/helpers/smartAccounts/safe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeEip712Params = void 0;
exports.decodeInternalTransactions = decodeInternalTransactions;
exports.tryDecodeSafeCalldataToLowLevelCalls = tryDecodeSafeCalldataToLowLevelCalls;
exports.formatSafeSignature = formatSafeSignature;
const viem_1 = require("viem");
const safeAccountsAbi = [
    {
        inputs: [
            { name: 'to', internalType: 'address', type: 'address' },
            { name: 'value', internalType: 'uint256', type: 'uint256' },
            { name: 'data', internalType: 'bytes', type: 'bytes' },
            { name: 'operation', internalType: 'uint8', type: 'uint8' },
        ],
        name: 'executeUserOp',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        type: 'function',
        name: 'executeUserOpWithErrorString',
        inputs: [
            { name: 'to', internalType: 'address', type: 'address' },
            { name: 'value', internalType: 'uint256', type: 'uint256' },
            { name: 'data', internalType: 'bytes', type: 'bytes' },
            { name: 'operation', internalType: 'uint8', type: 'uint8' },
        ],
        outputs: [],
        stateMutability: 'nonpayable',
    },
];
const multiSendAbi = [
    {
        inputs: [
            {
                internalType: 'bytes',
                name: 'transactions',
                type: 'bytes',
            },
        ],
        name: 'multiSend',
        outputs: [],
        stateMutability: 'payable',
        type: 'function',
    },
];
/**
 * Decodes a stream of packed internal transactions.
 * Opposite of encodeInternalTransaction.
 * https://github.com/pimlicolabs/permissionless.js/blob/main/packages/permissionless/accounts/safe/toSafeSmartAccount.ts#L566
 * * Format:
 * [operation (1b)][to (20b)][value (32b)][dataLength (32b)][data (variable)]
 */
function decodeInternalTransactions(data) {
    // Ensure data is a valid Hex string with 0x prefix for viem utils
    const hex = data.startsWith('0x') ? data : `0x${data}`;
    // If empty, return empty array
    if (hex === '0x' || hex.length === 0)
        return [];
    const txs = [];
    let cursor = 0;
    const totalBytes = (hex.length - 2) / 2; // Calculate byte length from hex string
    while (cursor < totalBytes) {
        // 1. Decode Operation (uint8 = 1 byte)
        const opBytes = (0, viem_1.slice)(hex, cursor, cursor + 1);
        const operation = (0, viem_1.hexToNumber)(opBytes);
        cursor += 1;
        // 2. Decode To Address (address = 20 bytes)
        const toBytes = (0, viem_1.slice)(hex, cursor, cursor + 20);
        const to = (0, viem_1.getAddress)(toBytes); // Checksums the address
        cursor += 20;
        // 3. Decode Value (uint256 = 32 bytes)
        const valueBytes = (0, viem_1.slice)(hex, cursor, cursor + 32);
        const value = (0, viem_1.hexToBigInt)(valueBytes);
        cursor += 32;
        // 4. Decode Data Length (uint256 = 32 bytes)
        const lengthBytes = (0, viem_1.slice)(hex, cursor, cursor + 32);
        const dataLength = (0, viem_1.hexToNumber)(lengthBytes);
        cursor += 32;
        // 5. Decode Data (bytes = dynamic length)
        // Note: encodePacked simply appends the bytes, no padding
        const dataBytes = (0, viem_1.slice)(hex, cursor, cursor + dataLength);
        cursor += dataLength;
        txs.push({
            operation,
            to,
            value,
            data: dataBytes,
        });
    }
    return txs;
}
function tryDecodeSafeMultiSendCalldata(df) {
    try {
        const { args } = (0, viem_1.decodeFunctionData)({
            data: df.args[2],
            abi: multiSendAbi,
        });
        const [internalTxs] = (args || ['0x']);
        const decodedInternalTxs = decodeInternalTransactions(internalTxs);
        return decodedInternalTxs.map((t) => ({ data: t.data, to: t.to, value: t.value }));
    }
    catch (e) {
        console.error('tryDecodeSafeMultiSendCalldata', e);
        return null;
    }
}
function tryDecodeSafeSendCalldata(df) {
    try {
        const lowLevelCall = {
            data: df.args[2],
            to: df.args[0],
            value: df.args[1],
        };
        return [lowLevelCall];
    }
    catch (e) {
        console.error('tryDecodeSafeSendCalldata', e);
        return null;
    }
}
function tryDecodeSafeCalldataToLowLevelCalls(callData) {
    let df;
    try {
        df = (0, viem_1.decodeFunctionData)({ abi: safeAccountsAbi, data: callData });
    }
    catch {
        console.log('Calldata is not a known Safe calldata. Continuing...');
        return null;
    }
    console.log('Decoded calldata for Safe smart account.');
    if (!['executeUserOp', 'executeUserOpWithErrorString'].includes(df.functionName)) {
        console.log('Not a known Safe execute/executor call');
        return null;
    }
    const safeShapes = [
        tryDecodeSafeMultiSendCalldata,
        tryDecodeSafeSendCalldata,
    ];
    for (const dec of safeShapes) {
        const res = dec(df);
        if (res)
            return res;
    }
    return null;
}
exports.safeEip712Params = {
    domain: {
        chainId: '$chainId',
        verifyingContract: '$safe4337ModuleAddress',
    },
    types: {
        SafeOp: [
            { type: 'address', name: 'safe' },
            { type: 'uint256', name: 'nonce' },
            { type: 'bytes', name: 'initCode' },
            { type: 'bytes', name: 'callData' },
            { type: 'uint128', name: 'verificationGasLimit' },
            { type: 'uint128', name: 'callGasLimit' },
            { type: 'uint256', name: 'preVerificationGas' },
            { type: 'uint128', name: 'maxPriorityFeePerGas' },
            { type: 'uint128', name: 'maxFeePerGas' },
            { type: 'bytes', name: 'paymasterAndData' },
            { type: 'uint48', name: 'validAfter' },
            { type: 'uint48', name: 'validUntil' },
            { type: 'address', name: 'entryPoint' },
        ],
    },
    primaryType: 'SafeOp',
    message: {
        safe: '$userOp.sender',
        nonce: '$userOp.nonce',
        initCode: '$userOp.initCode',
        callData: '$userOp.callData',
        callGasLimit: '$userOp.callGasLimit',
        verificationGasLimit: '$userOp.verificationGasLimit',
        preVerificationGas: '$userOp.preVerificationGas',
        maxFeePerGas: '$userOp.maxFeePerGas',
        maxPriorityFeePerGas: '$userOp.maxPriorityFeePerGas',
        paymasterAndData: '$userOp.paymasterAndData',
        validAfter: '$validAfter',
        validUntil: '$validUntil',
        entryPoint: '$entryPointAddress',
    },
};
function formatSafeSignature(params) {
    return (0, viem_1.concat)([
        (0, viem_1.toHex)(params.validAfter, { size: 6 }),
        (0, viem_1.toHex)(params.validUntil, { size: 6 }),
        params.signature,
    ]);
}
