// @account-kit/smart-contracts/dist/esm/src/msca/plugins/multisig/extension.js
import {} from "@aa-sdk/core";
import {} from "viem";
import { getThreshold } from "./actions/getThreshold.js";
import { isOwnerOf } from "./actions/isOwnerOf.js";
import { proposeUserOperation } from "./actions/proposeUserOperation.js";
import { readOwners } from "./actions/readOwners.js";
import { signMultisigUserOperation } from "./actions/signMultisigUserOperation.js";
import { multisigPluginActions as multisigPluginActions_, } from "./plugin.js";
import {} from "./types.js";
/**
 * Provides actions for managing a multisig plugin within the specified client, including reading owners, checking ownership, getting the threshold, proposing user operations, and signing multisig user operations.
 *
 * @example
 * ```ts
 * import { createModularAccountAlchemyClient, multisigPluginActions } from "@account-kit/smart-contracts";
 *
 * const client = createModularAccountAlchemyClient(...).extend(multisigPluginActions);
 * ```
 *
 * @param {Client<TTransport, TChain, TAccount>} client The client instance configured with transport, chain, and account information
 * @returns {MultisigPluginActions<TAccount>} An object containing methods to perform actions related to the multisig plugin
 */
export const multisigPluginActions = (client) => ({
    ...multisigPluginActions_(client),
    readOwners: (args) => readOwners(client, args),
    isOwnerOf: (args) => isOwnerOf(client, args),
    getThreshold: (args) => getThreshold(client, args),
    proposeUserOperation: (args) => proposeUserOperation(client, args),
    signMultisigUserOperation: (params) => signMultisigUserOperation(client, params),
});
