// @lit-protocol/vincent-app-sdk/dist/src/abilityClient/precheck/resultCreators.js
"use strict";
// src/lib/abilityClient/precheck/resultCreators.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAllowPrecheckResult = createAllowPrecheckResult;
exports.createDenyPrecheckResult = createDenyPrecheckResult;
exports.createAbilityPrecheckResponseSuccess = createAbilityPrecheckResponseSuccess;
exports.createAbilityPrecheckResponseSuccessNoResult = createAbilityPrecheckResponseSuccessNoResult;
exports.createAbilityPrecheckResponseFailure = createAbilityPrecheckResponseFailure;
exports.createAbilityPrecheckResponseFailureNoResult = createAbilityPrecheckResponseFailureNoResult;
function createAllowPrecheckResult(evaluatedPolicies, allowedPolicies) {
    return {
        allow: true,
        evaluatedPolicies,
        allowedPolicies,
    };
}
function createDenyPrecheckResult(evaluatedPolicies, allowedPolicies, deniedPolicy) {
    return {
        allow: false,
        evaluatedPolicies,
        allowedPolicies,
        deniedPolicy,
    };
}
function createAbilityPrecheckResponseSuccess(params) {
    return {
        success: true,
        result: params.result,
        context: params.context,
    };
}
function createAbilityPrecheckResponseSuccessNoResult(params) {
    return {
        success: true,
        result: undefined,
        context: params === null || params === void 0 ? void 0 : params.context,
    };
}
function createAbilityPrecheckResponseFailure(params) {
    return {
        success: false,
        schemaValidationError: params.schemaValidationError,
        runtimeError: params.runtimeError,
        result: params.result,
        context: params.context,
    };
}
function createAbilityPrecheckResponseFailureNoResult(params) {
    return {
        success: false,
        runtimeError: params.runtimeError,
        schemaValidationError: params.schemaValidationError,
        result: undefined,
        context: params.context,
    };
}
