// @lit-protocol/vincent-app-sdk/dist/src/abilityClient/precheck/runPolicyPrechecks.js
"use strict";
// src/lib/abilityClient/precheck/runPolicyPrechecks.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.runAbilityPolicyPrechecks = runAbilityPolicyPrechecks;
const zod_1 = require("zod");
const internal_1 = require("@lit-protocol/vincent-ability-sdk/internal");
const resultCreators_1 = require("./resultCreators");
const bigintReplacer = (key, value) => {
    return typeof value === 'bigint' ? value.toString() : value;
};
async function runAbilityPolicyPrechecks(params) {
    var _a, _b;
    const { bundledVincentAbility: { vincentAbility, ipfsCid }, abilityParams, context, decodedPolicies, } = params;
    console.log('Executing runAbilityPolicyPrechecks()', Object.keys(params.bundledVincentAbility.vincentAbility.supportedPolicies.policyByPackageName));
    const validatedPolicies = await (0, internal_1.validatePolicies)({
        decodedPolicies,
        vincentAbility,
        abilityIpfsCid: ipfsCid,
        parsedAbilityParams: abilityParams,
    });
    const decodedPoliciesByPackageName = {};
    for (const { policyPackageName, parameters } of validatedPolicies) {
        decodedPoliciesByPackageName[policyPackageName] = parameters;
    }
    const evaluatedPolicies = [];
    const allowedPolicies = {};
    let deniedPolicy = undefined;
    const policyByName = vincentAbility.supportedPolicies.policyByPackageName;
    for (const { policyPackageName, abilityPolicyParams } of validatedPolicies) {
        const key = policyPackageName;
        const abilityPolicy = policyByName[key];
        evaluatedPolicies.push(key);
        const vincentPolicy = abilityPolicy.vincentPolicy;
        if (!vincentPolicy.precheck) {
            console.log('No precheck() defined policy', key, 'skipping...');
            continue;
        }
        try {
            console.log('Executing precheck() for policy', key);
            const result = await vincentPolicy.precheck({
                abilityParams: abilityPolicyParams,
                userParams: decodedPoliciesByPackageName[key],
            }, context);
            console.log('vincentPolicy.precheck() result', JSON.stringify(result, bigintReplacer, 2));
            // precheck() might have thrown a runtimeError or failed to parse the input
            if (((0, internal_1.isPolicyDenyResponse)(result) && result.runtimeError) || result.schemaValidationError) {
                deniedPolicy = {
                    result: undefined,
                    runtimeError: result.runtimeError,
                    schemaValidationError: result.schemaValidationError,
                    packageName: key,
                };
                break;
            }
            const { schemaToUse } = (0, internal_1.getSchemaForPolicyResponseResult)({
                value: result,
                allowResultSchema: (_a = vincentPolicy.precheckAllowResultSchema) !== null && _a !== void 0 ? _a : zod_1.z.undefined(),
                denyResultSchema: (_b = vincentPolicy.precheckDenyResultSchema) !== null && _b !== void 0 ? _b : zod_1.z.undefined(),
            });
            const validated = (0, internal_1.validateOrDeny)(result.result, schemaToUse, 'precheck', 'output');
            if ((0, internal_1.isPolicyDenyResponse)(result)) {
                // Return value from precheck was invalid in this case
                deniedPolicy = {
                    runtimeError: result.runtimeError,
                    schemaValidationError: result.schemaValidationError,
                    result: validated,
                    packageName: key,
                };
                break;
            }
            allowedPolicies[key] = {
                result: validated,
            };
        }
        catch (err) {
            deniedPolicy = {
                packageName: key,
                ...(0, internal_1.createDenyResult)({
                    runtimeError: err instanceof Error ? err.message : 'Unknown error in precheck()',
                }),
            };
            break;
        }
    }
    if (deniedPolicy) {
        const policiesContext = (0, resultCreators_1.createDenyPrecheckResult)(evaluatedPolicies, allowedPolicies, deniedPolicy);
        return {
            ...context,
            policiesContext,
        };
    }
    const policiesContext = (0, resultCreators_1.createAllowPrecheckResult)(evaluatedPolicies, allowedPolicies);
    return {
        ...context,
        policiesContext,
    };
}
