// @account-kit/smart-contracts/dist/esm/src/ma-v2/actions/install-validation/installValidation.js
import { AccountNotFoundError, NotAModularAccountV2Error, EntityIdOverrideError, IncompatibleClientError, isSmartAccountClient, isSmartAccountWithSigner, } from "@aa-sdk/core";
import { concatHex, encodeFunctionData, zeroAddress, } from "viem";
import { semiModularAccountBytecodeAbi } from "../../abis/semiModularAccountBytecodeAbi.js";
import { serializeHookConfig, serializeModuleEntity, serializeValidationConfig, } from "../common/utils.js";
import { isModularAccountV2, } from "../../account/common/modularAccountV2Base.js";
import { DEFAULT_OWNER_ENTITY_ID } from "../../utils.js";
/**
 * Provides validation installation and uninstallation functionalities for a MA v2 client, ensuring compatibility with `SmartAccountClient`.
 *
 * @example
 * ```ts
 * import { createModularAccountV2Client, installValidationActions, getDefaultSingleSignerValidationModuleAddress, SingleSignerValidationModule } from "@account-kit/smart-contracts";
 * import { Address } from "viem";
 *
 * const client = (await createModularAccountV2Client({ ... })).extend(installValidationActions);
 * const sessionKeyAddress: Address = "0x1234";
 * const sessionKeyEntityId: number = 1;
 *
 * await client.installValidation({
 *   validationConfig: {
 *     moduleAddress: getDefaultSingleSignerValidationModuleAddress(
 *       client.chain
 *     ),
 *     entityId: sessionKeyEntityId,
 *     isGlobal: true,
 *     isSignatureValidation: false,
 *     isUserOpValidation: true,
 *   },
 *   selectors: [],
 *   installData: SingleSignerValidationModule.encodeOnInstallData({
 *     entityId: sessionKeyEntityId,
 *     signer: sessionKeyAddress,
 *   }),
 *   hooks: [],
 * });
 *
 * await client.uninstallValidation({
 *   moduleAddress: sessionKeyAddress,
 *   entityId: sessionKeyEntityId,
 *   uninstallData: SingleSignerValidationModule.encodeOnUninstallData({
 *     entityId: sessionKeyEntityId,
 *   }),
 *   hookUninstallDatas: [],
 * });
 *
 * ```
 *
 * @param {object} client - The client instance which provides account and sendUserOperation functionality.
 * @returns {object} - An object containing two methods, `installValidation` and `uninstallValidation`.
 */
export function installValidationActions(client) {
    const encodeInstallValidation = async ({ validationConfig, selectors, installData, hooks, account = client.account, }) => {
        if (!account) {
            throw new AccountNotFoundError();
        }
        if (!isModularAccountV2(account)) {
            throw new NotAModularAccountV2Error();
        }
        if (isSmartAccountWithSigner(account) && !isSmartAccountClient(client)) {
            // if we don't differentiate between WebauthnModularAccountV2Client and ModularAccountV2Client, passing client to isSmartAccountClient complains
            throw new IncompatibleClientError("SmartAccountClient", "installValidation", client);
        }
        // an entityId of zero is only allowed if we're installing or uninstalling hooks on the fallback validation
        if (validationConfig.entityId === DEFAULT_OWNER_ENTITY_ID &&
            validationConfig.moduleAddress !== zeroAddress) {
            throw new EntityIdOverrideError();
        }
        return account.encodeCallData(encodeFunctionData({
            abi: semiModularAccountBytecodeAbi,
            functionName: "installValidation",
            args: [
                serializeValidationConfig(validationConfig),
                selectors,
                installData,
                hooks.map((hook) => concatHex([serializeHookConfig(hook.hookConfig), hook.initData])),
            ],
        }));
    };
    const encodeUninstallValidation = async ({ moduleAddress, entityId, uninstallData, hookUninstallDatas, account = client.account, }) => {
        if (!account) {
            throw new AccountNotFoundError();
        }
        if (!isModularAccountV2(account)) {
            throw new NotAModularAccountV2Error();
        }
        if (isSmartAccountWithSigner(account) && !isSmartAccountClient(client)) {
            // if we don't differentiate between WebauthnModularAccountV2Client and ModularAccountV2Client, passing client to isSmartAccountClient complains
            throw new IncompatibleClientError("SmartAccountClient", "uninstallValidation", client);
        }
        return account.encodeCallData(encodeFunctionData({
            abi: semiModularAccountBytecodeAbi,
            functionName: "uninstallValidation",
            args: [
                serializeModuleEntity({
                    moduleAddress,
                    entityId,
                }),
                uninstallData,
                hookUninstallDatas,
            ],
        }));
    };
    return {
        encodeInstallValidation,
        encodeUninstallValidation,
        installValidation: async ({ validationConfig, selectors, installData, hooks, account = client.account, overrides, }) => {
            if (!isSmartAccountClient(client)) {
                throw new IncompatibleClientError("SmartAccountClient", "installValidation", client);
            }
            if (!account) {
                throw new AccountNotFoundError();
            }
            const callData = await encodeInstallValidation({
                validationConfig,
                selectors,
                installData,
                hooks,
                account,
            });
            return client.sendUserOperation({
                uo: callData,
                account,
                overrides,
            });
        },
        uninstallValidation: async ({ moduleAddress, entityId, uninstallData, hookUninstallDatas, account = client.account, overrides, }) => {
            if (!account) {
                throw new AccountNotFoundError();
            }
            if (!isSmartAccountClient(client)) {
                throw new IncompatibleClientError("SmartAccountClient", "uninstallValidation", client);
            }
            const callData = await encodeUninstallValidation({
                moduleAddress,
                entityId,
                uninstallData,
                hookUninstallDatas,
                account,
            });
            return client.sendUserOperation({
                uo: callData,
                account,
                overrides,
            });
        },
    };
}
