// @lit-protocol/vincent-ability-sdk/dist/src/type-inference-verification/ability-lifecycle-succeed-fail-tests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ability_none_p = exports.ability_none_pf = exports.ability_f_none = exports.ability_f_p = exports.ability_f_pf = exports.ability_s_none = exports.ability_s_p = void 0;
// src/type-inference-verification/ability-lifecycle-succeed-fail-tests.ts
const zod_1 = require("zod");
const helpers_1 = require("../lib/abilityCore/helpers");
const vincentAbility_1 = require("../lib/abilityCore/vincentAbility");
const bundledPolicy_1 = require("../lib/policyCore/bundledPolicy/bundledPolicy");
const vincentPolicy_1 = require("../lib/policyCore/vincentPolicy");
const abilityParams = zod_1.z.object({ x: zod_1.z.string() });
const dummyPolicy = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema: abilityParams,
    bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)((0, vincentPolicy_1.createVincentPolicy)({
        packageName: 'test',
        abilityParamsSchema: zod_1.z.object({ foo: zod_1.z.string() }),
        evalAllowResultSchema: zod_1.z.string(),
        evaluate: async (_, ctx) => ctx.allow('ok'),
    }), 'cid-test'),
    abilityParameterMappings: { x: 'foo' },
});
const S = zod_1.z.object({ ok: zod_1.z.boolean() });
const F = zod_1.z.object({ err: zod_1.z.string() });
exports.ability_s_p = (0, vincentAbility_1.createVincentAbility)({
    packageName: '@lit-protocol/yesability3@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema: abilityParams,
    supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([dummyPolicy]),
    executeSuccessSchema: S,
    precheckSuccessSchema: S,
    execute: async (_, { succeed, fail }) => {
        // @ts-expect-error - succeed requires argument matching success schema
        succeed();
        succeed({ ok: true });
        fail();
        // @ts-expect-error - fail should not take object when no fail schema
        fail({ err: 'fail' });
        return succeed({ ok: true });
    },
    precheck: async (_, { succeed, fail }) => {
        // @ts-expect-error - succeed requires argument matching success schema
        succeed();
        succeed({ ok: true });
        fail();
        // @ts-expect-error - fail should not take object when no fail schema
        fail({ err: 'fail' });
        return succeed({ ok: true });
    },
});
exports.ability_s_none = (0, vincentAbility_1.createVincentAbility)({
    packageName: '@lit-protocol/yesability3@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema: abilityParams,
    supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([dummyPolicy]),
    executeSuccessSchema: S,
    execute: async (_, { succeed, fail }) => {
        // @ts-expect-error - succeed requires argument matching success schema
        succeed();
        succeed({ ok: true });
        fail();
        // @ts-expect-error - fail should not take object when no fail schema
        fail({ err: 'fail' });
        return succeed({ ok: true });
    },
    precheck: async (_, { succeed, fail }) => {
        succeed();
        // @ts-expect-error - succeed should not take argument without schema
        succeed({ ok: true });
        fail();
        // @ts-expect-error - fail should not take object when no fail schema
        fail({ err: 'fail' });
        return succeed();
    },
});
exports.ability_f_pf = (0, vincentAbility_1.createVincentAbility)({
    packageName: '@lit-protocol/yesability3@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema: abilityParams,
    supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([dummyPolicy]),
    executeFailSchema: F,
    precheckFailSchema: F,
    execute: async (_, { succeed, fail }) => {
        succeed();
        // @ts-expect-error - succeed should not take argument without schema
        succeed({ ok: true });
        // @ts-expect-error - fail requires argument matching fail schema
        fail();
        return fail({ err: 'fail' });
    },
    precheck: async (_, { succeed, fail }) => {
        succeed();
        // @ts-expect-error - succeed should not take argument without schema
        succeed({ ok: true });
        // @ts-expect-error - fail requires argument matching fail schema
        fail();
        return fail({ err: 'fail' });
    },
});
exports.ability_f_p = (0, vincentAbility_1.createVincentAbility)({
    packageName: '@lit-protocol/yesability3@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema: abilityParams,
    supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([dummyPolicy]),
    executeFailSchema: F,
    precheckSuccessSchema: S,
    execute: async (_, { succeed, fail }) => {
        succeed();
        // @ts-expect-error - succeed should not take argument without schema
        succeed({ ok: true });
        // @ts-expect-error - fail requires argument matching fail schema
        fail();
        return fail({ err: 'fail' });
    },
    precheck: async (_, { succeed, fail }) => {
        // @ts-expect-error - succeed requires argument matching success schema
        succeed();
        succeed({ ok: true });
        fail();
        // @ts-expect-error - fail should not take object when no fail schema
        fail({ err: 'fail' });
        return succeed({ ok: true });
    },
});
exports.ability_f_none = (0, vincentAbility_1.createVincentAbility)({
    packageName: '@lit-protocol/yesability3@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema: abilityParams,
    supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([dummyPolicy]),
    executeFailSchema: F,
    execute: async (_, { succeed, fail }) => {
        succeed();
        // @ts-expect-error - succeed should not take argument without schema
        succeed({ ok: true });
        // @ts-expect-error - fail requires argument matching fail schema
        fail();
        return fail({ err: 'fail' });
    },
    precheck: async (_, { succeed, fail }) => {
        succeed();
        // @ts-expect-error - succeed should not take argument without schema
        succeed({ ok: true });
        fail();
        // @ts-expect-error - fail should not take object when no fail schema
        fail({ err: 'fail' });
        return succeed();
    },
});
exports.ability_none_pf = (0, vincentAbility_1.createVincentAbility)({
    packageName: '@lit-protocol/yesability3@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema: abilityParams,
    supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([dummyPolicy]),
    precheckFailSchema: F,
    execute: async (_, { succeed, fail }) => {
        succeed();
        // @ts-expect-error - succeed should not take argument without schema
        succeed({ ok: true });
        fail();
        // @ts-expect-error - fail should not take object when no fail schema
        fail({ err: 'fail' });
        return succeed();
    },
    precheck: async (_, { succeed, fail }) => {
        succeed();
        // @ts-expect-error - succeed should not take argument without schema
        succeed({ ok: true });
        // @ts-expect-error - fail requires argument matching fail schema
        fail();
        return fail({ err: 'fail' });
    },
});
exports.ability_none_p = (0, vincentAbility_1.createVincentAbility)({
    packageName: '@lit-protocol/yesability3@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema: abilityParams,
    supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([dummyPolicy]),
    precheckSuccessSchema: S,
    execute: async (_, { succeed, fail }) => {
        succeed();
        // @ts-expect-error - succeed should not take argument without schema
        succeed({ ok: true });
        fail();
        // @ts-expect-error - fail should not take object when no fail schema
        fail({ err: 'fail' });
        return succeed();
    },
    precheck: async (_, { succeed, fail }) => {
        // @ts-expect-error - succeed requires argument matching success schema
        succeed();
        succeed({ ok: true });
        fail();
        // @ts-expect-error - fail should not take object when no fail schema
        fail({ err: 'fail' });
        return succeed({ ok: true });
    },
});
