// @lit-protocol/vincent-ability-sdk/dist/src/type-inference-verification/context-switching-tests.js
"use strict";
// src/type-inference-verification/context-switching-tests.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.testPrecheckEvaluateContextSwitching = testPrecheckEvaluateContextSwitching;
exports.testEvaluateCommitContextSwitching = testEvaluateCommitContextSwitching;
exports.testFullPolicyContextSwitching = testFullPolicyContextSwitching;
/**
 * Context Switching Tests
 *
 * This file contains test cases for verifying that the type constraints are
 * properly enforced when switching between different policy function contexts
 * (precheck, evaluate, commit) which may have different schemas.
 */
const zod_1 = require("zod");
const bundledPolicy_1 = require("../lib/policyCore/bundledPolicy/bundledPolicy");
const vincentPolicy_1 = require("../lib/policyCore/vincentPolicy");
// Base ability schema for all tests
const baseAbilitySchema = zod_1.z.object({
    action: zod_1.z.string(),
    target: zod_1.z.string(),
    amount: zod_1.z.number(),
});
/**
 * Test Case 1: Verify different schemas between precheck and evaluate
 */
function testPrecheckEvaluateContextSwitching() {
    // Different schemas for precheck and evaluate
    const precheckAllowSchema = zod_1.z.object({ prelimStatus: zod_1.z.string() });
    const precheckDenySchema = zod_1.z.object({ prelimReason: zod_1.z.string() });
    const evalAllowSchema = zod_1.z.object({ finalStatus: zod_1.z.string() });
    const evalDenySchema = zod_1.z.object({ finalReason: zod_1.z.string() });
    const precheckEvalContextSwitchPolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/test-policy@1.2.3',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        // Different schemas for each context
        precheckAllowResultSchema: precheckAllowSchema,
        precheckDenyResultSchema: precheckDenySchema,
        evalAllowResultSchema: evalAllowSchema,
        evalDenyResultSchema: evalDenySchema,
        precheck: async (params, context) => {
            // Test TypeScript errors first
            // @ts-expect-error - Using eval schema in precheck context
            context.allow({ finalStatus: 'wrong-context' });
            // @ts-expect-error - Using eval schema in precheck context
            context.deny({ finalReason: 'wrong-context' });
            // Valid for precheck context
            if (Math.random() > 0.5) {
                return context.allow({ prelimStatus: 'valid' });
            }
            else {
                return context.deny({ prelimReason: 'invalid-prelim' });
            }
        },
        evaluate: async (params, context) => {
            // Test TypeScript errors first
            // @ts-expect-error - Using precheck schema in evaluate context
            context.allow({ prelimStatus: 'wrong-context' });
            // @ts-expect-error - Using precheck schema in evaluate context
            context.deny({ prelimReason: 'wrong-context' });
            // Valid for evaluate context
            if (Math.random() > 0.5) {
                return context.allow({ finalStatus: 'approved' });
            }
            else {
                return context.deny({ finalReason: 'denied-final' });
            }
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(precheckEvalContextSwitchPolicy, '02u3kjwdsnfs2890hf'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
}
/**
 * Test Case 2: Verify different schemas between evaluate and commit
 */
function testEvaluateCommitContextSwitching() {
    // Different schemas for evaluate and commit
    const evalAllowSchema = zod_1.z.object({ approvalId: zod_1.z.string() });
    const evalDenySchema = zod_1.z.object({ rejectionReason: zod_1.z.string() });
    const commitParamsSchema = zod_1.z.object({ confirmationId: zod_1.z.string() });
    const commitAllowSchema = zod_1.z.object({ transactionHash: zod_1.z.string() });
    const commitDenySchema = zod_1.z.object({ failureCode: zod_1.z.number() });
    const evalCommitContextSwitchPolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/test-policy@1.23.1',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        // Different schemas for each context
        evalAllowResultSchema: evalAllowSchema,
        evalDenyResultSchema: evalDenySchema,
        commitParamsSchema: commitParamsSchema,
        commitAllowResultSchema: commitAllowSchema,
        commitDenyResultSchema: commitDenySchema,
        evaluate: async (params, context) => {
            // Test TypeScript errors first
            // @ts-expect-error - Using commit schema in evaluate context
            context.allow({ transactionHash: 'wrong-context' });
            // @ts-expect-error - Using commit schema in evaluate context
            context.deny({ failureCode: 400 });
            // Valid for evaluate context
            if (Math.random() > 0.5) {
                return context.allow({ approvalId: 'approved-123' });
            }
            else {
                return context.deny({ rejectionReason: 'policy-violation' });
            }
        },
        commit: async (params, context) => {
            // Confirm params inference works
            const confId = params.confirmationId;
            // Test TypeScript errors first
            // @ts-expect-error - Using evaluate schema in commit context
            context.allow({ approvalId: 'wrong-context' });
            // @ts-expect-error - Using evaluate schema in commit context
            context.deny({ rejectionReason: 'wrong-context' });
            // Valid for commit context
            if (confId.startsWith('valid')) {
                return context.allow({ transactionHash: `hash-${confId}` });
            }
            else {
                return context.deny({ failureCode: 500 }, 'Transaction failed');
            }
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(evalCommitContextSwitchPolicy, '048oifjnwfoij3208h'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
}
/**
 * Test Case 3: Full policy with all three contexts having different schemas
 */
function testFullPolicyContextSwitching() {
    // Different schemas for each context
    const precheckAllowSchema = zod_1.z.object({ check: zod_1.z.boolean() });
    const precheckDenySchema = zod_1.z.object({ checkFailed: zod_1.z.boolean() });
    const evalAllowSchema = zod_1.z.object({ granted: zod_1.z.boolean() });
    const evalDenySchema = zod_1.z.object({ denied: zod_1.z.boolean() });
    const commitParamsSchema = zod_1.z.object({ txId: zod_1.z.string() });
    const commitAllowSchema = zod_1.z.object({ completed: zod_1.z.boolean() });
    const commitDenySchema = zod_1.z.object({ aborted: zod_1.z.boolean() });
    const fullPolicyContext = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/testpolicyawesome@12.1.10',
        abilityParamsSchema: zod_1.z.object({
            actionType: zod_1.z.string(),
            amount: zod_1.z.number(),
        }),
        userParamsSchema: zod_1.z.object({
            limit: zod_1.z.number(),
        }),
        // All different schemas
        precheckAllowResultSchema: precheckAllowSchema,
        precheckDenyResultSchema: precheckDenySchema,
        evalAllowResultSchema: evalAllowSchema,
        evalDenyResultSchema: evalDenySchema,
        commitParamsSchema: commitParamsSchema,
        commitAllowResultSchema: commitAllowSchema,
        commitDenyResultSchema: commitDenySchema,
        precheck: async ({ abilityParams, userParams }, context) => {
            // Test TypeScript errors first
            // @ts-expect-error - Using eval schema in precheck
            context.allow({ granted: true });
            // @ts-expect-error - Using commit schema in precheck
            context.allow({ completed: true });
            // @ts-expect-error - Using eval schema in precheck
            context.deny({ denied: true });
            // @ts-expect-error - Using commit schema in precheck
            context.deny({ aborted: true });
            // Valid for precheck
            if (abilityParams.amount <= userParams.limit) {
                return context.allow({ check: true });
            }
            else {
                return context.deny({ checkFailed: true });
            }
        },
        evaluate: async ({ abilityParams, userParams }, context) => {
            // Test TypeScript errors first
            // @ts-expect-error - Using precheck schema in eval
            context.allow({ check: true });
            // @ts-expect-error - Using commit schema in eval
            context.allow({ completed: true });
            // @ts-expect-error - Using precheck schema in eval
            context.deny({ checkFailed: true });
            // @ts-expect-error - Using commit schema in eval
            context.deny({ aborted: true });
            // Valid for eval
            if (abilityParams.amount <= userParams.limit) {
                return context.allow({ granted: true });
            }
            else {
                return context.deny({ denied: true });
            }
        },
        commit: async (params, context) => {
            const id = params.txId;
            // Test TypeScript errors first
            // @ts-expect-error - Using precheck schema in commit
            context.allow({ check: true });
            // @ts-expect-error - Using eval schema in commit
            context.allow({ granted: true });
            // @ts-expect-error - Using precheck schema in commit
            context.deny({ checkFailed: true });
            // @ts-expect-error - Using eval schema in commit
            context.deny({ denied: true });
            // Valid for commit
            if (id.startsWith('valid')) {
                return context.allow({ completed: true });
            }
            else {
                return context.deny({ aborted: true });
            }
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(fullPolicyContext, '908joiun2f928q1hjbfn'),
        abilityParameterMappings: {
            action: 'actionType',
            amount: 'amount',
        },
    });
}
