// @lit-protocol/vincent-ability-sdk/dist/src/type-inference-verification/create-vincent-ability-policy.js
"use strict";
// src/type-inference-verification/create-vincent-ability-policy.ts
Object.defineProperty(exports, "__esModule", { value: true });
const zod_1 = require("zod");
const bundledPolicy_1 = require("../lib/policyCore/bundledPolicy/bundledPolicy");
const vincentPolicy_1 = require("../lib/policyCore/vincentPolicy");
const abilityParamsSchema = zod_1.z.object({
    x: zod_1.z.string(),
});
const policyParamsSchema = zod_1.z.object({
    message: zod_1.z.string(),
});
const evalAllow = zod_1.z.object({ allowed: zod_1.z.boolean() });
const evalDeny = zod_1.z.object({ reason: zod_1.z.string() });
const PolicyConfig = (0, vincentPolicy_1.createVincentPolicy)({
    packageName: 'my-policy',
    abilityParamsSchema: policyParamsSchema,
    evalAllowResultSchema: evalAllow,
    evalDenyResultSchema: evalDeny,
    evaluate: async ({ abilityParams }, ctx) => {
        return ctx.allow({ allowed: true });
    },
});
const bundled = (0, bundledPolicy_1.asBundledVincentPolicy)(PolicyConfig, 'QmCID12345');
const abilityPolicy = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema,
    bundledVincentPolicy: bundled,
    abilityParameterMappings: {
        x: 'message',
    },
});
const cid = true;
console.log(cid);
// ✅ Infere nce: ipfsCid should be literal
// type Cid = typeof abilityPolicy.ipfsCid;
//    ^? "QmCID12345"
// ✅ Inference: packageName should be literal
// type Package = typeof abilityPolicy.vincentPolicy.packageName;
//    ^? "my-policy"
// ✅ Inference: schema accessors should be accurate
// type AllowSchema = typeof abilityPolicy.__schemaTypes.evalAllowResultSchema;
//    ^? z.ZodObject<{ allowed: z.ZodBoolean }>
// ✅ Function types inferred
// type EvaluateFn = typeof abilityPolicy.__schemaTypes.evaluate;
//    ^? (input: ..., context: ...) => Promise<...>
// ❌ Invalid schema field — should error
// @ts-expect-error invalid schema key
console.log(abilityPolicy.__schemaTypes.notARealSchema);
