// @lit-protocol/vincent-ability-sdk/dist/src/type-inference-verification/playground.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gogo = exports.gogoPolicy = exports.myAbility = void 0;
const zod_1 = require("zod");
const helpers_1 = require("../lib/abilityCore/helpers");
const vincentAbility_1 = require("../lib/abilityCore/vincentAbility");
const bundledPolicy_1 = require("../lib/policyCore/bundledPolicy/bundledPolicy");
const helpers_2 = require("../lib/policyCore/helpers");
const resultCreators_1 = require("../lib/policyCore/helpers/resultCreators");
const vincentPolicy_1 = require("../lib/policyCore/vincentPolicy");
// Define your ability schema
const myAbilitySchema = zod_1.z.object({
    action: zod_1.z.string(),
    target: zod_1.z.string(),
    amount: zod_1.z.number(),
});
/* eslint-disable @typescript-eslint/no-unused-vars */
// Define policy schemas
const policy1Schema = zod_1.z.object({
    targetAllowed: zod_1.z.boolean(),
    actionType: zod_1.z.string(),
});
const userParams1Schema = zod_1.z.object({ allowedTargets: zod_1.z.array(zod_1.z.string()) });
const commitParams1Schema = zod_1.z.object({ confirmation: zod_1.z.boolean() });
// Define the result schemas for policy1
const policy1EvalAllowResult = zod_1.z.string(); // Result when allowed
const policy1EvalDenyResult = zod_1.z.object({
    reason: zod_1.z.string(),
    numTries: zod_1.z.number(),
}); // Result when denied
const policy1CommitAllowResult = zod_1.z.object({ success: zod_1.z.boolean() });
const policy1CommitDenyResult = zod_1.z.object({ errorCode: zod_1.z.number() });
// Create policies with full type inference
const PolicyConfig1 = (0, vincentPolicy_1.createVincentPolicy)({
    packageName: 'extra-rate-limit',
    abilityParamsSchema: policy1Schema,
    userParamsSchema: userParams1Schema,
    commitParamsSchema: commitParams1Schema,
    // Define schema types explicitly (optional but helpful for validation)
    evalAllowResultSchema: policy1EvalAllowResult,
    evalDenyResultSchema: policy1EvalDenyResult,
    commitAllowResultSchema: policy1CommitAllowResult,
    commitDenyResultSchema: policy1CommitDenyResult,
    precheck: async ({ abilityParams, userParams }, context) => {
        const wat = userParams.allowedTargets;
        console.log(wat);
        if (abilityParams.targetAllowed) {
            return context.allow();
        }
        else {
            return context.deny();
        }
    },
    evaluate: async ({ abilityParams, userParams }, context) => {
        const targetIsAllowed = userParams.allowedTargets.includes('example');
        if (targetIsAllowed && abilityParams.targetAllowed) {
            return context.allow('APPROVED');
        }
        else {
            return context.deny({
                reason: 'Target not in allowed list',
                numTries: 1,
            });
        }
    },
    commit: async (params, context) => {
        if (params.confirmation) {
            return context.allow({ success: true });
        }
        else {
            return context.deny({ errorCode: 400 });
        }
    },
});
const policy1 = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema: myAbilitySchema,
    bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(PolicyConfig1, 'QmX7Dqn4zYhJVvXYwKr8cFX5Xp7gVpqK5r8QHwvF8zYjXa'),
    abilityParameterMappings: {
        target: 'targetAllowed',
        action: 'actionType',
    },
});
// Define the result schemas for policy2
const policy2Schema = zod_1.z.object({
    maxAmount: zod_1.z.number(),
    currency: zod_1.z.string(),
});
const userParams2Schema = zod_1.z.object({ limit: zod_1.z.number() });
const commitParams2Schema = zod_1.z.object({ transactionId: zod_1.z.string() });
// For policy2, let's use different result types
const policy2PrecheckAllowResult = zod_1.z.object({ validatedAmount: zod_1.z.number() });
const policy2PrecheckDenyResult = zod_1.z.object({ limitExceeded: zod_1.z.boolean() });
const policy2EvalAllowResult = zod_1.z.object({ approvedCurrency: zod_1.z.string() });
const policy2EvalDenyResult = zod_1.z.object({
    reason: zod_1.z.string(),
    code: zod_1.z.number(),
});
const policy2CommitAllowResult = zod_1.z.object({
    transaction: zod_1.z.string(),
    timestamp: zod_1.z.number(),
});
const policy2CommitDenyResult = zod_1.z.object({ failureReason: zod_1.z.string() });
const PolicyConfig2 = (0, vincentPolicy_1.createVincentPolicy)({
    packageName: 'rate-limit',
    abilityParamsSchema: policy2Schema,
    userParamsSchema: userParams2Schema,
    commitParamsSchema: commitParams2Schema,
    // Define schema types explicitly
    precheckAllowResultSchema: policy2PrecheckAllowResult,
    precheckDenyResultSchema: policy2PrecheckDenyResult,
    evalAllowResultSchema: policy2EvalAllowResult,
    evalDenyResultSchema: policy2EvalDenyResult,
    commitAllowResultSchema: policy2CommitAllowResult,
    commitDenyResultSchema: policy2CommitDenyResult,
    precheck: async ({ abilityParams, userParams }, context) => {
        const isWithinLimit = abilityParams.maxAmount <= userParams.limit;
        if (isWithinLimit) {
            return context.allow({ validatedAmount: 1983 });
        }
        else {
            return context.deny({ limitExceeded: true });
        }
    },
    evaluate: async ({ abilityParams, userParams }, context) => {
        const isValidCurrency = ['USD', 'EUR', 'GBP'].includes(abilityParams.currency);
        const isWithinLimit = abilityParams.maxAmount <= userParams.limit;
        if (isValidCurrency && isWithinLimit) {
            return context.allow({ approvedCurrency: abilityParams.currency });
        }
        else {
            const reason = !isValidCurrency
                ? `Currency ${abilityParams.currency} is not supported`
                : `Amount ${abilityParams.maxAmount} exceeds limit ${userParams.limit}`;
            return context.deny({
                reason: reason,
                code: !isValidCurrency ? 415 : 400,
            });
        }
    },
    commit: async (params, context) => {
        // For demo purposes, let's say transactions with "fail" in the ID will be denied
        if (!params.transactionId.includes('fail')) {
            return context.allow({
                transaction: `completed-${params.transactionId}`,
                timestamp: Date.now(),
            });
        }
        else {
            return context.deny({
                failureReason: 'Invalid transaction ID format',
            });
        }
    },
});
const policy2 = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema: myAbilitySchema,
    bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(PolicyConfig2, 'QmYwAPJzv5CZsnA625s3Xf2nemtYgPpHdWEz79ojWnPbdG'),
    abilityParameterMappings: {
        amount: 'maxAmount',
        action: 'currency',
    },
});
// Define schemas for policy3
const policy3AbilityParams = zod_1.z.object({
    abilityName: zod_1.z.string(),
    maxUsage: zod_1.z.number(),
});
const policy3UserParams = zod_1.z.object({
    userAddress: zod_1.z.string(),
    accessLevel: zod_1.z.enum(['basic', 'premium', 'admin']),
});
const policy3EvalAllowResult = zod_1.z.object({
    message: zod_1.z.string(),
    timedate: zod_1.z.date(),
});
const policy3EvalDenyResult = zod_1.z.object({
    reason: zod_1.z.string(),
    numberOfTries: zod_1.z.number(),
    suggestedAccessLevel: zod_1.z.enum(['basic', 'premium', 'admin']).optional(),
});
// Create policy3 without a commit function
const PolicyConfig3 = (0, vincentPolicy_1.createVincentPolicy)({
    packageName: 'vincent-ability-sdk',
    abilityParamsSchema: policy3AbilityParams,
    userParamsSchema: policy3UserParams,
    evalAllowResultSchema: policy3EvalAllowResult,
    evalDenyResultSchema: policy3EvalDenyResult,
    commit: async (_, { deny, allow }) => {
        return allow();
    },
    // Only has evaluate, no precheck or commit
    evaluate: async ({ abilityParams, userParams }, context) => {
        // Policy logic: Allow only premium and admin users
        if (userParams.accessLevel === 'basic') {
            return context.deny({
                reason: 'Basic users cannot access this ability',
                suggestedAccessLevel: 'premium',
                numberOfTries: 1,
            });
        }
        return context.allow({
            message: `Access granted to ${abilityParams.abilityName}`,
            timedate: new Date(),
        });
    },
});
const policy3 = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema: myAbilitySchema,
    bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(PolicyConfig3, 'QmYwAasdPJzv5CZA625s3Xf2nemtYgPpHdWEz79ojWnPbdG'),
    abilityParameterMappings: {
        amount: 'abilityName',
    },
});
const abilityExecuteSuccessSchema = zod_1.z.object({
    executedAction: zod_1.z.string(),
    targetAddress: zod_1.z.string(),
    transactionAmount: zod_1.z.number(),
    timestamp: zod_1.z.number(),
    txHash: zod_1.z.string(),
});
const abilityExecuteFailSchema = zod_1.z.object({
    errorCode: zod_1.z.number(),
    errorMessage: zod_1.z.string(),
    suggestion: zod_1.z.string().optional(),
});
const abilityPrecheckSuccessSchema = zod_1.z.object({
    validatedParams: zod_1.z.boolean(),
    message: zod_1.z.string(),
});
const abilityPrecheckFailSchema = zod_1.z.object({
    invalidField: zod_1.z.string(),
    reason: zod_1.z.string(),
});
// Create your ability with fully typed policies
exports.myAbility = (0, vincentAbility_1.createVincentAbility)({
    packageName: '@lit-protocol/awesome-ability@1.0.2',
    abilityDescription: 'Awesome Ability',
    abilityParamsSchema: myAbilitySchema,
    supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([policy1, policy2, policy3]),
    // Add schemas for ability results
    executeSuccessSchema: abilityExecuteSuccessSchema,
    executeFailSchema: abilityExecuteFailSchema,
    precheckSuccessSchema: abilityPrecheckSuccessSchema,
    precheckFailSchema: abilityPrecheckFailSchema,
    precheck: async ({ abilityParams }, { fail, policiesContext: { allow, allowedPolicies, deniedPolicy }, succeed }) => {
        // Basic validation
        if (!abilityParams.action || !abilityParams.target) {
            return fail({
                invalidField: !abilityParams.action ? 'action' : 'target',
                reason: 'Required field is empty or missing',
            });
        }
        // Amount validation
        if (abilityParams.amount <= 0) {
            return fail({
                invalidField: 'amount',
                reason: 'Amount must be greater than zero',
            });
        }
        // Process policy results if needed
        if (allow) {
            // Type-safe access to policies
            const extraRateLimitResults = allowedPolicies['extra-rate-limit'];
            if (extraRateLimitResults) {
                const policy1Result = extraRateLimitResults.result;
                console.log('policy1Result', policy1Result);
            }
            const rateLimitResults = allowedPolicies['rate-limit'];
            if (rateLimitResults) {
                const policy2Result = rateLimitResults.result;
                console.log(policy2Result);
            }
            const abilitySdkResults = allowedPolicies['vincent-ability-sdk'];
            if (abilitySdkResults) {
                const policy3Result = abilitySdkResults.result;
                console.log(policy3Result);
            }
            // Return success result
            return succeed({
                validatedParams: true,
                message: `All parameters validated for ${abilityParams.action} on ${abilityParams.target}`,
            });
        }
        else {
            // Handle the denial case
            const denyReason = deniedPolicy.runtimeError || 'Policy check failed';
            return fail({
                invalidField: 'policy',
                reason: denyReason,
            });
        }
    },
    execute: async ({ abilityParams }, { policiesContext, fail, succeed }) => {
        try {
            // Simulate successful execution
            if (abilityParams.action === 'transfer' && abilityParams.amount > 0) {
                // Execute the transaction
                const txHash = `0x${Math.random().toString(16).substring(2, 10)}`;
                // Use commit functions from policies if available
                const extraRateLimitPolicyContext = policiesContext.allowedPolicies['extra-rate-limit'];
                if (extraRateLimitPolicyContext) {
                    const commitResult = await extraRateLimitPolicyContext.commit({
                        confirmation: true,
                    });
                    console.log(commitResult);
                }
                const rateLimitPolicyContext = policiesContext.allowedPolicies['rate-limit'];
                if (rateLimitPolicyContext) {
                    const commitResult = await rateLimitPolicyContext.commit({
                        transactionId: txHash,
                    });
                    console.log(commitResult);
                }
                const abilitySdkPolicyContext = policiesContext.allowedPolicies['vincent-ability-sdk'];
                if (abilitySdkPolicyContext) {
                    const commitResult = await abilitySdkPolicyContext.commit();
                    const wat = commitResult.allow;
                    console.log(wat);
                    if (commitResult.allow) {
                        console.log(commitResult.result);
                    }
                }
                // Return success with typed result
                return succeed({
                    executedAction: abilityParams.action,
                    targetAddress: abilityParams.target,
                    transactionAmount: abilityParams.amount,
                    timestamp: Date.now(),
                    txHash,
                });
            }
            else {
                // Simulate a failure case
                return fail({
                    errorCode: 400,
                    errorMessage: `Cannot perform ${abilityParams.action} operation`,
                    suggestion: 'Try a "transfer" action instead',
                });
            }
        }
        catch (error) {
            // Handle any unexpected errors
            return fail({
                errorCode: 500,
                errorMessage: 'Internal execution error',
            });
        }
    },
});
const gogoPolicy = async function () {
    await policy3.vincentPolicy.evaluate({
        abilityParams: {
            abilityName: 'wat',
            maxUsage: 2383,
        },
        userParams: { userAddress: 'meow', accessLevel: 'basic' },
    }, {
        delegation: {
            delegateeAddress: 'meow',
            delegatorPkpInfo: {
                tokenId: '90128301832',
                ethAddress: '0x102398103981032',
                publicKey: '0398103810938ef987ef978fe987ef',
            },
        },
        abilityIpfsCid: 'oijskljfdj',
        appId: 123123,
        appVersion: 123,
    });
    if (policy2.vincentPolicy.commit) {
        return policy2.vincentPolicy.commit({
            transactionId: '10981328981279487',
        }, {
            abilityIpfsCid: 'oijskljfdj',
            appId: 123123,
            appVersion: 123,
            delegation: {
                delegateeAddress: 'meow',
                delegatorPkpInfo: {
                    tokenId: '90128301832',
                    ethAddress: '0x102398103981032',
                    publicKey: '0398103810938ef987ef978fe987ef',
                },
            },
        });
    }
    return true;
};
exports.gogoPolicy = gogoPolicy;
const gogo = async function () {
    const abilityExecuteResult = await exports.myAbility.execute({ abilityParams: { action: 'wat', target: 'meow', amount: 23098123 } }, {
        abilityIpfsCid: 'oijskljfdj',
        appId: 123123,
        appVersion: 123,
        delegation: {
            delegateeAddress: 'meow',
            delegatorPkpInfo: {
                tokenId: '90128301832',
                ethAddress: '0x102398103981032',
                publicKey: '0398103810938ef987ef978fe987ef',
            },
        },
        policiesContext: {
            allow: true,
            evaluatedPolicies: ['extra-rate-limit', 'rate-limit'],
            allowedPolicies: {
                'extra-rate-limit': {
                    // @ts-expect-error result for this policy is a string
                    result: { this_is_wrong: true },
                },
                'rate-limit': {
                    result: {
                        approvedCurrency: 'USD',
                    },
                    commit: async (params) => {
                        if (!params.transactionId.includes('fail')) {
                            return (0, resultCreators_1.createAllowResult)({
                                result: {
                                    transaction: `completed-${params.transactionId}`,
                                    timestamp: 1700000000000,
                                },
                            });
                        }
                        else {
                            return (0, helpers_2.createDenyResult)({
                                result: {
                                    failureReason: 'Invalid transaction ID format',
                                },
                            });
                        }
                    },
                },
            },
        },
    });
    console.log(abilityExecuteResult);
    if (exports.myAbility.precheck) {
        await exports.myAbility.precheck({ abilityParams: { action: 'wat', target: 'meow', amount: 23098123 } }, {
            abilityIpfsCid: 'oijskljfdj',
            appId: 123123,
            appVersion: 123,
            delegation: {
                delegateeAddress: 'meow',
                delegatorPkpInfo: {
                    tokenId: '90128301832',
                    ethAddress: '0x102398103981032',
                    publicKey: '0398103810938ef987ef978fe987ef',
                },
            },
            policiesContext: {
                allow: false,
                deniedPolicy: {
                    packageName: 'extra-rate-limit',
                    result: { reason: 'wth', numTries: 1 },
                },
                evaluatedPolicies: ['extra-rate-limit', 'rate-limit'],
                allowedPolicies: {
                    'extra-rate-limit': {
                        result: 'wat',
                    },
                    'rate-limit': {
                        result: {
                            approvedCurrency: 'USD',
                        },
                    },
                },
            },
        });
    }
    return true;
};
exports.gogo = gogo;
