// @account-kit/smart-contracts/dist/esm/src/light-account/accounts/account.js
import { createBundlerClient, getEntryPoint, } from "@aa-sdk/core";
import { concatHex, encodeFunctionData, } from "viem";
import { LightAccountAbi_v1 } from "../abis/LightAccountAbi_v1.js";
import { LightAccountAbi_v2 } from "../abis/LightAccountAbi_v2.js";
import { LightAccountFactoryAbi_v1 } from "../abis/LightAccountFactoryAbi_v1.js";
import { LightAccountFactoryAbi_v2 } from "../abis/LightAccountFactoryAbi_v2.js";
import { AccountVersionRegistry, LightAccountUnsupported1271Factories, defaultLightAccountVersion, getDefaultLightAccountFactoryAddress, } from "../utils.js";
import { createLightAccountBase, } from "./base.js";
import { predictLightAccountAddress } from "./predictAddress.js";
/**
 * Creates a light account based on the provided parameters such as transport, chain, signer, init code, and more. Ensures that an account is configured and returned with various capabilities, such as transferring ownership and retrieving the owner's address.
 *
 * @example
 * ```ts
 * import { createLightAccount } from "@account-kit/smart-contracts";
 * import { LocalAccountSigner } from "@aa-sdk/core";
 * import { sepolia } from "viem/chains";
 * import { http, generatePrivateKey } from "viem"
 *
 * const account = await createLightAccount({
 *  chain: sepolia,
 *  transport: http("RPC_URL"),
 *  signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey())
 * });
 * ```
 *
 * @param {CreateLightAccountParams} config The parameters for creating a light account
 * @returns {Promise<LightAccount>} A promise that resolves to a `LightAccount` object containing the created account information and methods
 */
export async function createLightAccount({ transport, chain, signer, initCode, version = defaultLightAccountVersion(), entryPoint = getEntryPoint(chain, {
    version: AccountVersionRegistry["LightAccount"][version]
        .entryPointVersion,
}), accountAddress, factoryAddress = getDefaultLightAccountFactoryAddress(chain, version), salt: salt_ = 0n, }) {
    const client = createBundlerClient({
        transport,
        chain,
    });
    const accountAbi = version === "v2.0.0" ? LightAccountAbi_v2 : LightAccountAbi_v1;
    const factoryAbi = version === "v2.0.0"
        ? LightAccountFactoryAbi_v1
        : LightAccountFactoryAbi_v2;
    const signerAddress = await signer.getAddress();
    const salt = LightAccountUnsupported1271Factories.has(factoryAddress.toLowerCase())
        ? 0n
        : salt_;
    const getAccountInitCode = async () => {
        if (initCode)
            return initCode;
        return concatHex([
            factoryAddress,
            encodeFunctionData({
                abi: factoryAbi,
                functionName: "createAccount",
                args: [signerAddress, salt],
            }),
        ]);
    };
    const address = accountAddress ??
        predictLightAccountAddress({
            factoryAddress,
            salt,
            ownerAddress: signerAddress,
            version,
        });
    const account = await createLightAccountBase({
        transport,
        chain,
        signer,
        abi: accountAbi,
        type: "LightAccount",
        version,
        entryPoint,
        accountAddress: address,
        getAccountInitCode,
    });
    return {
        ...account,
        encodeTransferOwnership: (newOwner) => {
            return encodeFunctionData({
                abi: accountAbi,
                functionName: "transferOwnership",
                args: [newOwner],
            });
        },
        async getOwnerAddress() {
            const callResult = await client.readContract({
                address,
                abi: accountAbi,
                functionName: "owner",
            });
            if (callResult == null) {
                throw new Error("could not get on-chain owner");
            }
            return callResult;
        },
    };
}
