// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/helpers/smartAccounts/kernel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tryDecodeKernelCalldataToLowLevelCalls = tryDecodeKernelCalldataToLowLevelCalls;
exports.formatSecondaryValidatorKernelSignature = formatSecondaryValidatorKernelSignature;
const viem_1 = require("viem");
const kernelAccountsAbi = [
    // Kernel v3.3/v3.1/v3.0
    {
        type: 'function',
        name: 'execute',
        inputs: [
            { name: 'execMode', type: 'bytes32', internalType: 'ExecMode' },
            { name: 'executionCalldata', type: 'bytes', internalType: 'bytes' },
        ],
        outputs: [],
        stateMutability: 'payable',
    },
    {
        type: 'function',
        name: 'executeFromExecutor',
        inputs: [
            { name: 'execMode', type: 'bytes32', internalType: 'ExecMode' },
            { name: 'executionCalldata', type: 'bytes', internalType: 'bytes' },
        ],
        outputs: [{ name: 'returnData', type: 'bytes[]', internalType: 'bytes[]' }],
        stateMutability: 'payable',
    },
    {
        type: 'function',
        name: 'executeUserOp',
        inputs: [
            {
                name: 'userOp',
                type: 'tuple',
                internalType: 'struct PackedUserOperation',
                components: [
                    {
                        name: 'sender',
                        type: 'address',
                        internalType: 'address',
                    },
                    { name: 'nonce', type: 'uint256', internalType: 'uint256' },
                    { name: 'initCode', type: 'bytes', internalType: 'bytes' },
                    { name: 'callData', type: 'bytes', internalType: 'bytes' },
                    {
                        name: 'accountGasLimits',
                        type: 'bytes32',
                        internalType: 'bytes32',
                    },
                    {
                        name: 'preVerificationGas',
                        type: 'uint256',
                        internalType: 'uint256',
                    },
                    {
                        name: 'gasFees',
                        type: 'bytes32',
                        internalType: 'bytes32',
                    },
                    {
                        name: 'paymasterAndData',
                        type: 'bytes',
                        internalType: 'bytes',
                    },
                    { name: 'signature', type: 'bytes', internalType: 'bytes' },
                ],
            },
            { name: 'userOpHash', type: 'bytes32', internalType: 'bytes32' },
        ],
        outputs: [],
        stateMutability: 'payable',
    },
];
function tryDecodeKernelExecuteBatchCall(df) {
    try {
        // https://github.com/zerodevapp/sdk/blob/main/packages/core/accounts/kernel/utils/ep0_7/encodeExecuteBatchCall.ts
        const [tuples] = (0, viem_1.decodeAbiParameters)([
            {
                name: 'executionBatch',
                type: 'tuple[]',
                components: [
                    {
                        name: 'target',
                        type: 'address',
                    },
                    {
                        name: 'value',
                        type: 'uint256',
                    },
                    {
                        name: 'callData',
                        type: 'bytes',
                    },
                ],
            },
        ], df.args[1]);
        if (!tuples.length)
            return null;
        return tuples.map((t) => ({
            to: (0, viem_1.getAddress)(t.target),
            value: t.value,
            data: t.callData,
        }));
    }
    catch (e) {
        console.error('tryDecodeKernelExecuteBatchCall', e);
        return null;
    }
}
function tryDecodeKernelExecuteCall(df) {
    // https://github.com/zerodevapp/sdk/blob/main/packages/core/accounts/kernel/utils/ep0_7/encodeExecuteSingleCall.ts
    // Because Execute with a single call is just a hex concatenation of the calldata,
    // this function might produce invalid decoding. It MUST be called last, only if
    // all other, stricter, decodings failed or if we know the calldata is a single call.
    // Also, the current implementation does not support delegate calls. It will produce
    // an invalid decoding for them.
    try {
        const data = df.args[1];
        const hex = data.slice(2);
        const MIN = 20 * 2 + 32 * 2; // address + value
        if (hex.length < MIN)
            return null;
        const to = (0, viem_1.getAddress)((0, viem_1.sliceHex)(data, 0, 20));
        const value = (0, viem_1.hexToBigInt)((0, viem_1.sliceHex)(data, 20, 52));
        const calldata = (0, viem_1.sliceHex)(data, 52);
        return [
            {
                data: calldata,
                to,
                value,
            },
        ];
    }
    catch (e) {
        console.error('tryDecodeKernelExecuteCall', e);
        return null;
    }
}
function isDelegatecallOrUnknown(execMode) {
    const lastByte = BigInt(execMode) & 0xffn;
    return lastByte !== 0n; // 0x00 == CALL; others => block
}
function tryDecodeKernelCalldataToLowLevelCalls(callData) {
    let df;
    try {
        df = (0, viem_1.decodeFunctionData)({ abi: kernelAccountsAbi, data: callData });
    }
    catch {
        console.log('Calldata is not a known Kernel calldata. Continuing...');
        return null;
    }
    console.log('Decoded calldata for Kernel smart account.');
    if (!['execute', 'executeFromExecutor'].includes(df.functionName)) {
        console.log('Not a known Kernel execute/executor call');
        return null;
    }
    const [execMode] = df.args;
    // BLOCK delegatecall/unknown modes up-front
    if (isDelegatecallOrUnknown(execMode)) {
        throw new Error(`Blocked by execMode (non-CALL): ${execMode}`);
    }
    // Try common inner shapes
    const shapes = [
        tryDecodeKernelExecuteBatchCall,
        tryDecodeKernelExecuteCall,
    ];
    for (const dec of shapes) {
        const res = dec(df);
        if (res)
            return res;
    }
    return null;
}
function formatSecondaryValidatorKernelSignature(params) {
    return (0, viem_1.concat)([params.secondValidatorId, params.signature]);
}
