// @lit-protocol/vincent-app-sdk/dist/src/type-inference-verification/ability-client-inference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = run;
exports.gogo = gogo;
const zod_1 = require("zod");
const vincent_ability_sdk_1 = require("@lit-protocol/vincent-ability-sdk");
const vincentAbilityClient_1 = require("../abilityClient/vincentAbilityClient");
const currencyPolicy = (0, vincent_ability_sdk_1.createVincentPolicy)({
    packageName: 'currency-policy',
    abilityParamsSchema: zod_1.z.object({ currency: zod_1.z.string() }),
    evaluate: async ({ abilityParams }, ctx) => ctx.allow({ approvedCurrency: abilityParams.currency }),
    evalAllowResultSchema: zod_1.z.object({ approvedCurrency: zod_1.z.string() }),
    evalDenyResultSchema: zod_1.z.object({ borked: zod_1.z.string() }),
});
const currencyAbilityPolicy = (0, vincent_ability_sdk_1.createVincentAbilityPolicy)({
    abilityParamsSchema: zod_1.z.object({
        currency: zod_1.z.string(),
    }),
    bundledVincentPolicy: (0, vincent_ability_sdk_1.asBundledVincentPolicy)(currencyPolicy, 'QmCurrency123'),
    abilityParameterMappings: {
        currency: 'currency',
    },
});
const rateLimitPolicy = (0, vincent_ability_sdk_1.createVincentPolicy)({
    packageName: 'rate-limit',
    abilityParamsSchema: zod_1.z.object({ userId: zod_1.z.string() }),
    evaluate: async (_, ctx) => ctx.allow({ allowed: true }),
    evalAllowResultSchema: zod_1.z.object({ allowed: zod_1.z.literal(true) }),
    commit: async (_, ctx) => ctx.allow({ committed: true }),
    commitAllowResultSchema: zod_1.z.object({ committed: zod_1.z.boolean() }),
});
const rateLimitAbilityPolicy = (0, vincent_ability_sdk_1.createVincentAbilityPolicy)({
    abilityParamsSchema: zod_1.z.object({
        userId: zod_1.z.string(),
    }),
    bundledVincentPolicy: (0, vincent_ability_sdk_1.asBundledVincentPolicy)(rateLimitPolicy, 'QmRateLimit123'),
    abilityParameterMappings: {
        userId: 'userId',
    },
});
const abilityParamsSchema = zod_1.z.object({
    currency: zod_1.z.string(),
    userId: zod_1.z.string(),
});
const ability = (0, vincent_ability_sdk_1.createVincentAbility)({
    packageName: '@lit-protocol/ability3@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema,
    supportedPolicies: (0, vincent_ability_sdk_1.supportedPoliciesForAbility)([currencyAbilityPolicy, rateLimitAbilityPolicy]),
    execute: async ({ abilityParams }, ctx) => ctx.succeed({ ok: true }),
    executeSuccessSchema: zod_1.z.object({ ok: zod_1.z.literal(true) }),
});
const client = (0, vincentAbilityClient_1.getVincentAbilityClient)({
    bundledVincentAbility: (0, vincent_ability_sdk_1.asBundledVincentAbility)(ability, 'QmFakeAbility123'),
    ethersSigner: {}, // stubbed
    debug: false,
});
async function run() {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    const abilityParams = {
        currency: 'USD',
        userId: 'user-123',
    };
    const precheckResult = await client.precheck(abilityParams, {
        delegatorPkpEthAddress: '0xabc',
        agentAddress: '0xagent',
    });
    if (precheckResult.success === true) {
        // ✅ Inferred success result shape
        const successResult = precheckResult.result;
        // Should be ok to access successResult with correct type
        if (successResult !== undefined) {
            // @ts-expect-error - no properties defined on undefined schema
            console.log(successResult.foo);
        }
        const policiesContext = (_a = precheckResult === null || precheckResult === void 0 ? void 0 : precheckResult.context) === null || _a === void 0 ? void 0 : _a.policiesContext;
        if (precheckResult && policiesContext && policiesContext.allow) {
            const p = policiesContext.allowedPolicies;
            // ✅ Should infer currency-policy result shape
            if (p['currency-policy']) {
                // @ts-expect-error There is no precheck result schema; result must be undefined
                p['currency-policy'].result.approvedCurrency.toUpperCase();
                // @ts-expect-error - no commit on currency-policy
                p['currency-policy'].commit();
            }
            // ✅ Should infer rate-limit result shape and commit
            if (p['rate-limit']) {
                // @ts-expect-error There is no precheck result schema; result must be undefined
                const val = p['rate-limit'].result.allowed;
                console.log(val);
            }
            // @ts-expect-error - policy not in map
            console.log(p['non-existent-policy']);
        }
    }
    else {
        // ✅ Inferred failure shape
        const fail = precheckResult.result;
        // @ts-expect-error - fail is z.undefined(), cannot have properties
        console.log(fail.reason);
        // Can still access error message
        (_b = precheckResult.runtimeError) === null || _b === void 0 ? void 0 : _b.toUpperCase();
        // Should still be optional policiesContext
        const deniedPolicy = (_d = (_c = precheckResult.context) === null || _c === void 0 ? void 0 : _c.policiesContext) === null || _d === void 0 ? void 0 : _d.deniedPolicy;
        if (deniedPolicy) {
            if (deniedPolicy.packageName === 'currency-policy' &&
                deniedPolicy.result &&
                // @ts-expect-error There's no precheck deny schema!
                deniedPolicy.result.borked) {
                // @ts-expect-error There's no precheck deny schema!
                console.log(deniedPolicy.result.borked);
            }
            console.log(deniedPolicy.result);
        }
    }
    // 🧪 Now check execute inference
    const executeResult = await client.execute(abilityParams, {
        delegatorPkpEthAddress: '0x09182301238',
        agentAddress: '0xagent',
    });
    if (executeResult.success === true) {
        if (executeResult.result) {
            const val = executeResult.result;
            // @ts-expect-error - invalid field on success result
            console.log(val.failureReason);
            const result = (_f = (_e = executeResult.context) === null || _e === void 0 ? void 0 : _e.policiesContext) === null || _f === void 0 ? void 0 : _f.allowedPolicies;
            if (result) {
                const currencyPolicyResult = result['currency-policy'];
                if (currencyPolicyResult) {
                    console.log(currencyPolicyResult.result.approvedCurrency);
                }
            }
        }
    }
    else {
        console.log((_j = (_h = (_g = executeResult.context) === null || _g === void 0 ? void 0 : _g.policiesContext) === null || _h === void 0 ? void 0 : _h.deniedPolicy) === null || _j === void 0 ? void 0 : _j.result);
    }
}
const fullSchemaPolicy = (0, vincent_ability_sdk_1.createVincentPolicy)({
    packageName: 'full-policy',
    abilityParamsSchema: zod_1.z.object({ count: zod_1.z.number() }),
    evaluate: async ({ abilityParams }, ctx) => {
        if (abilityParams.count > 0) {
            return ctx.allow({ result: 'ok' });
        }
        else {
            return ctx.deny({ reason: 'must be > 0' });
        }
    },
    evalAllowResultSchema: zod_1.z.object({ result: zod_1.z.literal('ok') }),
    evalDenyResultSchema: zod_1.z.object({ reason: zod_1.z.string() }),
});
const fullSchemaAbilityPolicy = (0, vincent_ability_sdk_1.createVincentAbilityPolicy)({
    abilityParamsSchema: zod_1.z.object({ count: zod_1.z.number() }),
    bundledVincentPolicy: (0, vincent_ability_sdk_1.asBundledVincentPolicy)(fullSchemaPolicy, 'QmFullSchema123'),
    abilityParameterMappings: {
        count: 'count',
    },
});
const fullAbility = (0, vincent_ability_sdk_1.createVincentAbility)({
    packageName: '@lit-protocol/ability3@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema: zod_1.z.object({ count: zod_1.z.number() }),
    supportedPolicies: (0, vincent_ability_sdk_1.supportedPoliciesForAbility)([fullSchemaAbilityPolicy]),
    precheck: async ({ abilityParams }, ctx) => {
        if (abilityParams.count > 10) {
            return ctx.succeed({ accepted: true });
        }
        else {
            return ctx.fail({ reason: 'too small' });
        }
    },
    precheckSuccessSchema: zod_1.z.object({ accepted: zod_1.z.literal(true) }),
    precheckFailSchema: zod_1.z.object({ reason: zod_1.z.string() }),
    execute: async ({ abilityParams }, ctx) => {
        if (abilityParams.count < 100) {
            return ctx.succeed({ ok: true });
        }
        else {
            return ctx.fail({ reason: 'too big' });
        }
    },
    executeSuccessSchema: zod_1.z.object({ ok: zod_1.z.literal(true) }),
    executeFailSchema: zod_1.z.object({ reason: zod_1.z.string() }),
});
const fullClient = (0, vincentAbilityClient_1.getVincentAbilityClient)({
    bundledVincentAbility: (0, vincent_ability_sdk_1.asBundledVincentAbility)(fullAbility, 'QmFullAbility123'),
    ethersSigner: {},
    debug: false,
});
const fullParams = { count: 5 };
async function gogo() {
    const precheck = await fullClient.precheck(fullParams, {
        delegatorPkpEthAddress: '0x1sfskjdfhf',
        agentAddress: '0xagent',
    });
    if (precheck.success === true) {
        const ok = precheck.result.accepted;
        console.log(ok);
        // @ts-expect-error foo not valid on precheck result
        console.log(precheck.result.foo);
    }
    else {
        const reason = precheck.result.reason;
        console.log(reason);
        // @ts-expect-error accepted not valid on failure!
        console.log(precheck.result.accepted);
    }
    const execute = await fullClient.execute(fullParams, {
        delegatorPkpEthAddress: '0x098479847928734',
        agentAddress: '0xagent',
    });
    if (execute.success === true) {
        const val = execute.result.ok;
        console.log(val);
        // @ts-expect-error reason not valid on success result
        console.log(execute.result.reason);
    }
    else {
        const msg = execute.result.reason;
        console.log(msg);
        // @ts-expect-error OK not valid on failure result
        console.log(execute.result.ok);
    }
}
