// @lit-protocol/vincent-ability-sdk/dist/src/lib/abilityCore/abilityConfig/context/abilityContext.js
"use strict";
// src/lib/abilityCore/abilityConfig/context/abilityContext.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.createExecutionAbilityContext = createExecutionAbilityContext;
exports.createPrecheckAbilityContext = createPrecheckAbilityContext;
const resultCreators_1 = require("./resultCreators");
/**
 * Builds an execution-time AbilityContext for use inside `execute()` lifecycle methods.
 * It upgrades the incoming external policy context with `commit()` methods derived
 * from the ability's supported policies and schema definitions. Ensures commit calls
 * are fully typed and validated internally.
 */
function createExecutionAbilityContext(params) {
    const { baseContext, policiesByPackageName } = params;
    const allowedPolicies = {};
    for (const key of Object.keys(policiesByPackageName)) {
        const k = key;
        const entry = baseContext.policiesContext.allowedPolicies[k];
        if (!entry)
            continue;
        allowedPolicies[k] = {
            ...entry,
        };
    }
    const upgradedPoliciesContext = {
        evaluatedPolicies: baseContext.policiesContext.evaluatedPolicies,
        allow: true,
        deniedPolicy: undefined,
        allowedPolicies,
    };
    return {
        ...baseContext,
        policiesContext: upgradedPoliciesContext,
        succeed: resultCreators_1.createSuccess,
        fail: resultCreators_1.createFailure,
    };
}
/**
 * Builds a precheck-time AbilityContext for use inside `precheck()` lifecycle methods.
 * It includes only the evaluated policy metadata and denies commit access,
 * ensuring developers don’t call commit prematurely. Enforces policy result typing.
 */
function createPrecheckAbilityContext(params) {
    const { baseContext } = params;
    return {
        ...baseContext,
        succeed: resultCreators_1.createSuccess,
        fail: resultCreators_1.createFailure,
    };
}
