// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/helpers/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeSuccessSchema = exports.precheckSuccessSchema = exports.executeFailSchema = exports.precheckFailSchema = exports.isTransactionAbilityParams = exports.isUserOpAbilityParams = exports.abilityParamsSchema = void 0;
const account_abstraction_1 = require("viem/account-abstraction");
const zod_1 = require("zod");
const eip712_1 = require("./eip712");
const hex_1 = require("./hex");
const simulation_1 = require("./simulation");
const transaction_1 = require("./transaction");
const userOperation_1 = require("./userOperation");
const baseAbilityParamsSchema = zod_1.z.object({
    alchemyRpcUrl: simulation_1.alchemyRpcUrlSchema,
});
/**
 * Ability params Schemas. Can be to sign either a user operation or a transaction
 */
const userOpAbilityParamsSchema = baseAbilityParamsSchema.extend({
    userOp: userOperation_1.userOpSchema.describe('User operation to simulate, validate and sign. This MUST be a valid UserOperation object as defined in the UserOperation schemas.'),
    entryPointAddress: hex_1.addressSchema
        .optional()
        .default(account_abstraction_1.entryPoint07Address)
        .describe('EntryPoint to use for the simulation. Currently only v0.7 is supported. Defaults to standard v0.7 entryPoint address.'),
    alchemyRpcUrl: simulation_1.alchemyRpcUrlSchema,
    validAfter: zod_1.z.number().default(0).describe('Valid after timestamp (for Safe smart accounts)'),
    validUntil: zod_1.z.number().default(0).describe('Valid until timestamp (for Safe smart accounts)'),
    safe4337ModuleAddress: hex_1.addressSchema
        .default('0x75cf11467937ce3F2f357CE24ffc3DBF8fD5c226')
        .describe('Safe 4337 Module address (for Safe smart accounts)'),
    eip712Params: eip712_1.eip712ParamsSchema
        .optional()
        .describe('EIP-712 signing parameters for non-standard signing (e.g. Safe smart accounts)'),
});
const transactionAbilityParamsSchema = baseAbilityParamsSchema.extend({
    alchemyRpcUrl: simulation_1.alchemyRpcUrlSchema,
    transaction: transaction_1.transactionSchema.describe('EOA transaction to simulate, validate and sign. This must be a valid and complete transaction.'),
});
exports.abilityParamsSchema = zod_1.z.union([
    userOpAbilityParamsSchema,
    transactionAbilityParamsSchema,
]);
const isUserOpAbilityParams = (params) => 'userOp' in params;
exports.isUserOpAbilityParams = isUserOpAbilityParams;
const isTransactionAbilityParams = (params) => 'transaction' in params;
exports.isTransactionAbilityParams = isTransactionAbilityParams;
/**
 * Failure result schema
 */
const baseFailSchema = zod_1.z.object({
    error: zod_1.z.string().describe('A string containing the error message if the precheck failed.'),
});
exports.precheckFailSchema = baseFailSchema;
exports.executeFailSchema = baseFailSchema;
/**
 * Success result schema
 */
exports.precheckSuccessSchema = zod_1.z.object({
    simulationChanges: zod_1.z
        .array(simulation_1.simulateAssetChangeSchema)
        .describe('Simulated changes user op will make to the blockchain.'),
});
exports.executeSuccessSchema = exports.precheckSuccessSchema.extend({
    signature: hex_1.hexSchema.describe('ECDSA signature over the received user operation or transaction.'),
});
