// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/helpers/signUserOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signUserOperation = signUserOperation;
const viem_1 = require("viem");
const account_abstraction_1 = require("viem/account-abstraction");
const toLitActionAccount_1 = require("./toLitActionAccount");
async function signUserOperation({ alchemyRpcUrl, entryPointAddress, pkpPublicKey, userOp, validAfter, validUntil, safe4337ModuleAddress, eip712Params, }) {
    const publicClient = (0, viem_1.createPublicClient)({
        transport: (0, viem_1.http)(alchemyRpcUrl),
    });
    const chainId = await publicClient.getChainId();
    const account = (0, toLitActionAccount_1.toLitActionAccount)(pkpPublicKey);
    const userOperationToHash = {
        ...userOp,
        callGasLimit: BigInt(userOp.callGasLimit),
        maxFeePerGas: BigInt(userOp.maxFeePerGas),
        maxPriorityFeePerGas: BigInt(userOp.maxPriorityFeePerGas),
        nonce: BigInt(userOp.nonce),
        paymasterPostOpGasLimit: BigInt(userOp.paymasterPostOpGasLimit),
        paymasterVerificationGasLimit: BigInt(userOp.paymasterVerificationGasLimit),
        preVerificationGas: BigInt(userOp.preVerificationGas),
        verificationGasLimit: BigInt(userOp.verificationGasLimit),
    };
    if (eip712Params) {
        const resolveValue = (value) => {
            if (typeof value === 'string' && value.startsWith('$')) {
                switch (value) {
                    case '$userOp.sender':
                        return userOperationToHash.sender;
                    case '$userOp.nonce':
                        return userOperationToHash.nonce;
                    case '$userOp.initCode':
                        if (userOperationToHash.factory && userOperationToHash.factoryData) {
                            return (0, viem_1.concat)([userOperationToHash.factory, userOperationToHash.factoryData]);
                        }
                        return '0x';
                    case '$userOp.callData':
                        return userOperationToHash.callData;
                    case '$userOp.callGasLimit':
                        return userOperationToHash.callGasLimit;
                    case '$userOp.verificationGasLimit':
                        return userOperationToHash.verificationGasLimit;
                    case '$userOp.preVerificationGas':
                        return userOperationToHash.preVerificationGas;
                    case '$userOp.maxFeePerGas':
                        return userOperationToHash.maxFeePerGas;
                    case '$userOp.maxPriorityFeePerGas':
                        return userOperationToHash.maxPriorityFeePerGas;
                    case '$userOp.paymasterAndData':
                        if (userOperationToHash.paymaster) {
                            return (0, viem_1.concat)([
                                userOperationToHash.paymaster,
                                (0, viem_1.pad)((0, viem_1.toHex)(userOperationToHash.paymasterVerificationGasLimit || 0n), { size: 16 }),
                                (0, viem_1.pad)((0, viem_1.toHex)(userOperationToHash.paymasterPostOpGasLimit || 0n), { size: 16 }),
                                userOperationToHash.paymasterData || '0x',
                            ]);
                        }
                        return '0x';
                    case '$userOp.signature':
                        return userOperationToHash.signature;
                    case '$userOp.factory':
                        return userOperationToHash.factory;
                    case '$userOp.factoryData':
                        return userOperationToHash.factoryData;
                    case '$userOp.paymaster':
                        return userOperationToHash.paymaster;
                    case '$userOp.paymasterData':
                        return userOperationToHash.paymasterData;
                    case '$userOp.paymasterVerificationGasLimit':
                        return userOperationToHash.paymasterVerificationGasLimit;
                    case '$userOp.paymasterPostOpGasLimit':
                        return userOperationToHash.paymasterPostOpGasLimit;
                    case '$entryPointAddress':
                        return entryPointAddress;
                    case '$validAfter':
                        return validAfter;
                    case '$validUntil':
                        return validUntil;
                    case '$safe4337ModuleAddress':
                        return safe4337ModuleAddress;
                    case '$chainId':
                        return chainId;
                    default:
                        throw new Error(`Unknown mapping reference: ${value}`);
                }
            }
            if (typeof value === 'object' && value !== null) {
                const result = {};
                for (const key in value) {
                    result[key] = resolveValue(value[key]);
                }
                return result;
            }
            return value;
        };
        const message = resolveValue(eip712Params.message);
        const domain = resolveValue(eip712Params.domain);
        return await account.signTypedData({
            domain,
            types: eip712Params.types,
            primaryType: eip712Params.primaryType,
            message,
        });
    }
    const userOpHash = (0, account_abstraction_1.getUserOperationHash)({
        chainId,
        entryPointAddress,
        entryPointVersion: '0.7',
        userOperation: userOperationToHash,
    });
    return await account.signMessage({
        message: { raw: userOpHash },
    });
}
