// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/helpers/signTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signTransaction = signTransaction;
const viem_1 = require("viem");
const toLitActionAccount_1 = require("./toLitActionAccount");
async function signTransaction({ pkpPublicKey, transaction }) {
    const account = (0, toLitActionAccount_1.toLitActionAccount)(pkpPublicKey);
    const signableTransaction = {
        ...transaction,
        accessList: transaction.accessList || [],
        gas: transaction.gas ? (0, viem_1.hexToBigInt)(transaction.gas) : undefined,
        gasLimit: transaction.gasLimit ? (0, viem_1.hexToBigInt)(transaction.gasLimit) : undefined,
        gasPrice: transaction.gasPrice ? (0, viem_1.hexToBigInt)(transaction.gasPrice) : undefined,
        maxFeePerGas: transaction.maxFeePerGas ? (0, viem_1.hexToBigInt)(transaction.maxFeePerGas) : undefined,
        maxPriorityFeePerGas: transaction.maxPriorityFeePerGas
            ? (0, viem_1.hexToBigInt)(transaction.maxPriorityFeePerGas)
            : undefined,
        nonce: (0, viem_1.hexToNumber)(transaction.nonce),
        value: (0, viem_1.hexToBigInt)(transaction.value),
    };
    // @ts-expect-error viem complains but the tx should be coherent besides generalities
    const signedTx = await account.signTransaction(signableTransaction);
    const parsed = (0, viem_1.parseTransaction)(signedTx);
    if (!parsed.r || !parsed.s || parsed.yParity === undefined)
        throw new Error('Signed tx missing signature fields');
    const yParityByte = parsed.yParity === 0 ? '0x00' : '0x01';
    const signature = (0, viem_1.concatHex)([parsed.r, parsed.s, yParityByte]);
    return signature;
}
