// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/helpers/simulation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.simulateUserOp = exports.simulateTransaction = exports.simulateAssetChangesResponseSchema = exports.simulateAssetChangesErrorSchema = exports.simulateAssetChangeSchema = exports.simulateChangeTypeSchema = exports.SimulateChangeType = exports.simulateAssetTypeSchema = exports.SimulateAssetType = exports.alchemyRpcUrlSchema = void 0;
const zod_1 = require("zod");
const hex_1 = require("./hex");
/**
 * Validates valid Alchemy RPC URLs, the simulation provider
 */
exports.alchemyRpcUrlSchema = zod_1.z
    .string()
    .regex(/^https:\/\/[a-z0-9-]+\.g\.alchemy\.com\/v2\/.+/, { message: 'Invalid Alchemy RPC URL' })
    .url()
    .describe('Alchemy RPC URL for the desired chain. Will be used to simulate the transaction or user op and define the target chain.');
// Types copied from @account-kit/infra
// Zod schemas are created for runtime validation; types are derived from them to ensure consistency
// Enums and their schemas
var SimulateAssetType;
(function (SimulateAssetType) {
    SimulateAssetType["NATIVE"] = "NATIVE";
    SimulateAssetType["ERC20"] = "ERC20";
    SimulateAssetType["ERC721"] = "ERC721";
    SimulateAssetType["ERC1155"] = "ERC1155";
    /**
     * Special contracts that don't follow ERC 721/1155. Currently limited to
     * CryptoKitties and CryptoPunks.
     */
    SimulateAssetType["SPECIAL_NFT"] = "SPECIAL_NFT";
})(SimulateAssetType || (exports.SimulateAssetType = SimulateAssetType = {}));
exports.simulateAssetTypeSchema = zod_1.z.nativeEnum(SimulateAssetType);
var SimulateChangeType;
(function (SimulateChangeType) {
    SimulateChangeType["APPROVE"] = "APPROVE";
    SimulateChangeType["TRANSFER"] = "TRANSFER";
})(SimulateChangeType || (exports.SimulateChangeType = SimulateChangeType = {}));
exports.simulateChangeTypeSchema = zod_1.z.nativeEnum(SimulateChangeType);
exports.simulateAssetChangeSchema = zod_1.z.object({
    assetType: exports.simulateAssetTypeSchema,
    changeType: exports.simulateChangeTypeSchema,
    from: hex_1.hexSchema,
    to: hex_1.hexSchema,
    rawAmount: zod_1.z.string().nullable().optional(),
    amount: zod_1.z.string().nullable().optional(),
    contractAddress: hex_1.hexSchema.nullable().optional(),
    tokenId: zod_1.z.string().nullable().optional(),
    decimals: zod_1.z.number(),
    symbol: zod_1.z.string(),
    name: zod_1.z.string().nullable().optional(),
    logo: zod_1.z.string().nullable().optional(),
});
exports.simulateAssetChangesErrorSchema = zod_1.z
    .object({
    message: zod_1.z.string(),
})
    .catchall(zod_1.z.unknown());
exports.simulateAssetChangesResponseSchema = zod_1.z.object({
    changes: zod_1.z.array(exports.simulateAssetChangeSchema),
    error: exports.simulateAssetChangesErrorSchema.nullable().optional(),
});
const simulateTransaction = async ({ publicClient, transaction, }) => {
    // Convert transaction to RPC-compatible format
    const rpcTransaction = {
        from: transaction.from,
        to: transaction.to,
        data: transaction.data,
        value: transaction.value,
        nonce: transaction.nonce,
        ...(transaction.gas && { gas: transaction.gas }),
        ...(transaction.gasLimit && !transaction.gas && { gas: transaction.gasLimit }),
        ...(transaction.gasPrice && { gasPrice: transaction.gasPrice }),
        ...(transaction.maxFeePerGas && { maxFeePerGas: transaction.maxFeePerGas }),
        ...(transaction.maxPriorityFeePerGas && {
            maxPriorityFeePerGas: transaction.maxPriorityFeePerGas,
        }),
        ...(transaction.accessList && { accessList: transaction.accessList }),
    };
    return (await publicClient.request({
        // @ts-expect-error viem types do not include this method
        method: 'alchemy_simulateAssetChanges',
        // @ts-expect-error viem types do not include this method
        params: [rpcTransaction],
    }));
};
exports.simulateTransaction = simulateTransaction;
const simulateUserOp = async ({ entryPointAddress, publicClient, userOp, }) => {
    return (await publicClient.request({
        // @ts-expect-error viem types do not include this method
        method: 'alchemy_simulateUserOperationAssetChanges',
        // @ts-expect-error viem types do not include this method
        params: [userOp, entryPointAddress],
    }));
};
exports.simulateUserOp = simulateUserOp;
