// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/helpers/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transactionSchema = void 0;
exports.toVincentTransaction = toVincentTransaction;
const viem_1 = require("viem");
const zod_1 = require("zod");
const hex_1 = require("./hex");
const quantityHexSchema = hex_1.hexSchema.describe('Hex-encoded quantity (for example 0x5208 for gas or 0xde0b6b3a7640000 for value)');
const chainIdSchema = zod_1.z
    .union([zod_1.z.number().int().nonnegative(), quantityHexSchema])
    .describe('Numeric chain id for the transaction. Can be decimal or hex encoded.');
const accessListSchema = zod_1.z
    .array(zod_1.z.object({
    address: hex_1.addressSchema,
    storageKeys: zod_1.z.array(hex_1.hexSchema),
}))
    .describe('EIP-2930/EIP-1559 access list entries');
exports.transactionSchema = zod_1.z
    .object({
    data: hex_1.hexSchema.describe('Calldata for the transaction.'),
    from: hex_1.addressSchema.describe('Sender of the transaction. Must be the delegator PKP.'),
    to: hex_1.addressSchema.describe('Recipient address for the transaction.'),
    value: hex_1.hexSchema.describe('Hex-encoded value to transfer in wei.'),
    nonce: quantityHexSchema.describe('Hex-encoded nonce for the transaction.'),
    gas: quantityHexSchema.optional().describe('Gas limit for the transaction.'),
    gasLimit: quantityHexSchema.optional().describe('Alias for gas limit.'),
    gasPrice: quantityHexSchema.optional().describe('Legacy gas price.'),
    maxFeePerGas: quantityHexSchema.optional().describe('EIP-1559 maxFeePerGas value.'),
    maxPriorityFeePerGas: quantityHexSchema
        .optional()
        .describe('EIP-1559 maxPriorityFeePerGas value.'),
    chainId: chainIdSchema.describe('Chain ID for the transaction. Required.'),
    type: zod_1.z
        .union([zod_1.z.string(), zod_1.z.number()])
        .optional()
        .describe('Optional transaction type identifier.'),
    accessList: accessListSchema.optional(),
})
    .refine((data) => data.gas !== undefined || data.gasLimit !== undefined, {
    message: 'Either gas or gasLimit must be provided',
})
    .refine((data) => {
    const hasLegacyGas = data.gasPrice !== undefined;
    const hasEIP1559Gas = data.maxFeePerGas !== undefined && data.maxPriorityFeePerGas !== undefined;
    return hasLegacyGas || hasEIP1559Gas;
}, {
    message: 'Either gasPrice (for legacy) or both maxFeePerGas and maxPriorityFeePerGas (for EIP-1559) must be provided',
});
const txHexValues = [
    'value',
    'nonce',
    'gas',
    'gasLimit',
    'gasPrice',
    'maxFeePerGas',
    'maxPriorityFeePerGas',
];
function toQuantityHexSafe(value) {
    if ((0, viem_1.isHex)(value))
        return value;
    if (typeof value === 'string' && /^\d+$/.test(value)) {
        return (0, viem_1.toHex)(BigInt(value));
    }
    return (0, viem_1.toHex)(value);
}
function toVincentTransaction(tx) {
    const _tx = { ...tx };
    // Convert quantity-like fields to hex when needed
    for (const key of txHexValues) {
        if (_tx[key] !== undefined) {
            _tx[key] = toQuantityHexSafe(_tx[key]);
        }
    }
    // Convert chainId if provided as a hex-like non-number and not already hex
    if (_tx.chainId !== undefined && typeof _tx.chainId !== 'number') {
        _tx.chainId = toQuantityHexSafe(_tx.chainId);
    }
    // Normalize accessList storage keys
    if (Array.isArray(_tx.accessList)) {
        _tx.accessList = _tx.accessList.map((entry) => ({
            address: entry.address,
            storageKeys: entry.storageKeys.map((k) => toQuantityHexSafe(k)),
        }));
    }
    return _tx;
}
