// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/helpers/validation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertIsValidUserOp = void 0;
exports.assertIsValidTransaction = assertIsValidTransaction;
const viem_1 = require("viem");
const lowLevelCall_1 = require("./lowLevelCall");
const simulation_1 = require("./simulation");
const assertIsValidUserOp = async ({ alchemyRpcUrl, decodeTransaction, entryPointAddress, userOp, validateSimulation, validateTransaction, }) => {
    const publicClient = (0, viem_1.createPublicClient)({
        transport: (0, viem_1.http)(alchemyRpcUrl),
    });
    const chainId = await publicClient.getChainId();
    const sender = (0, viem_1.getAddress)(userOp.sender, chainId);
    const userOpTransactions = (0, lowLevelCall_1.getUserOpCalls)(userOp);
    for (const transaction of userOpTransactions) {
        const decodedTransaction = decodeTransaction({
            transaction: {
                data: transaction.data,
                to: transaction.to,
                value: BigInt(transaction.value),
            },
        });
        await validateTransaction({
            chainId,
            decodedTransaction,
            sender,
        });
    }
    // Simulate userOp and validate changes
    const simulation = await (0, simulation_1.simulateUserOp)({
        entryPointAddress,
        publicClient,
        userOp,
    });
    if (simulation.error) {
        const { message, revertReason } = simulation.error;
        throw new Error(`Simulation failed - Reason: ${revertReason} - Message: ${message}`);
    }
    await validateSimulation({
        chainId,
        sender,
        simulation,
    });
    return simulation.changes;
};
exports.assertIsValidUserOp = assertIsValidUserOp;
async function assertIsValidTransaction({ alchemyRpcUrl, decodeTransaction, transaction, validateSimulation, validateTransaction, }) {
    const publicClient = (0, viem_1.createPublicClient)({
        transport: (0, viem_1.http)(alchemyRpcUrl),
    });
    const chainId = await publicClient.getChainId();
    const sender = (0, viem_1.getAddress)(transaction.from);
    const decodedTransaction = decodeTransaction({
        transaction: {
            data: transaction.data,
            to: transaction.to,
            value: BigInt(transaction.value),
        },
    });
    await validateTransaction({
        chainId,
        decodedTransaction,
        sender,
    });
    const simulation = await (0, simulation_1.simulateTransaction)({
        publicClient,
        transaction,
    });
    const { changes, error } = simulation;
    if (error) {
        const { message, revertReason } = error;
        throw new Error(`Simulation failed - Reason: ${revertReason} - Message: ${message}`);
    }
    await validateSimulation({
        chainId,
        sender,
        simulation,
    });
    return changes;
}
