// @lit-protocol/vincent-contracts-sdk/dist/src/contractClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clientFromContract = clientFromContract;
exports.getClient = getClient;
exports.getTestClient = getTestClient;
const constants_1 = require("./constants");
const App_1 = require("./internal/app/App");
const AppView_1 = require("./internal/app/AppView");
const User_1 = require("./internal/user/User");
const UserView_1 = require("./internal/user/UserView");
const utils_1 = require("./utils");
/** Client method for use in CI or localDev situations where you need to inject an instance of a contract with a custom address
 *
 * @category Internal
 * @internal
 * @private
 * @hidden
 * */
function clientFromContract({ contract }) {
    return {
        // App write methods
        registerApp: (params, overrides) => (0, App_1.registerApp)({ contract, args: params, overrides }),
        registerNextVersion: (params, overrides) => (0, App_1.registerNextVersion)({ contract, args: params, overrides }),
        enableAppVersion: (params, overrides) => (0, App_1.enableAppVersion)({ contract, args: params, overrides }),
        addDelegatee: (params, overrides) => (0, App_1.addDelegatee)({ contract, args: params, overrides }),
        removeDelegatee: (params, overrides) => (0, App_1.removeDelegatee)({ contract, args: params, overrides }),
        setDelegatee: (params, overrides) => (0, App_1.setDelegatee)({ contract, args: params, overrides }),
        deleteApp: (params, overrides) => (0, App_1.deleteApp)({ contract, args: params, overrides }),
        undeleteApp: (params, overrides) => (0, App_1.undeleteApp)({ contract, args: params, overrides }),
        // App view methods
        getAppById: (params) => (0, AppView_1.getAppById)({ contract, args: params }),
        getAppIdByDelegatee: (params) => (0, AppView_1.getAppIdByDelegatee)({ contract, args: params }),
        getAppVersion: (params) => (0, AppView_1.getAppVersion)({ contract, args: params }),
        getAppsByManagerAddress: (params) => (0, AppView_1.getAppsByManagerAddress)({ contract, args: params }),
        getAppByDelegateeAddress: (params) => (0, AppView_1.getAppByDelegateeAddress)({ contract, args: params }),
        getDelegatedAgentAddresses: (params) => (0, AppView_1.getDelegatedAgentAddresses)({ contract, args: params }),
        // User write methods
        permitApp: (params, overrides) => (0, User_1.permitApp)({ contract, args: params, overrides }),
        unPermitApp: (params, overrides) => (0, User_1.unPermitApp)({ contract, args: params, overrides }),
        rePermitApp: (params, overrides) => (0, User_1.rePermitApp)({ contract, args: params, overrides }),
        setAbilityPolicyParameters: (params, overrides) => (0, User_1.setAbilityPolicyParameters)({ contract, args: params, overrides }),
        // User view methods
        getAllRegisteredAgentAddressesForUser: (params) => (0, UserView_1.getAllRegisteredAgentAddressesForUser)({ contract, args: params }),
        getUserAddressForAgent: (params) => (0, UserView_1.getUserAddressForAgent)({ contract, args: params }),
        getPermittedAppForAgents: (params) => (0, UserView_1.getPermittedAppForAgents)({ contract, args: params }),
        getUnpermittedAppForAgents: (params) => (0, UserView_1.getUnpermittedAppForAgents)({ contract, args: params }),
        getAllAbilitiesAndPoliciesForApp: (params) => (0, UserView_1.getAllAbilitiesAndPoliciesForApp)({ contract, args: params }),
        validateAbilityExecutionAndGetPolicies: (params) => (0, UserView_1.validateAbilityExecutionAndGetPolicies)({ contract, args: params }),
        isDelegateePermitted: (params) => (0, UserView_1.isDelegateePermitted)({ contract, args: params }),
    };
}
/** Get an instance of the contract client that is configured to use a 'production' instance of the contract
 *
 * Please use {@link getTestClient} for temporary / development, or for CI / integration test usage
 *
 * @category API */
function getClient({ signer, contractAddress, }) {
    const contract = (0, utils_1.createContract)({
        signer,
        contractAddress: contractAddress ?? constants_1.VINCENT_DIAMOND_CONTRACT_ADDRESS_PROD,
    });
    return clientFromContract({ contract });
}
/** Get an instance of the contract client that is configured to use a test/dev instance of the contract
 *
 * @category API */
function getTestClient({ signer, contractAddress, }) {
    const contract = (0, utils_1.createContract)({
        signer,
        contractAddress: contractAddress ?? constants_1.VINCENT_DIAMOND_CONTRACT_ADDRESS_PROD,
    });
    return clientFromContract({ contract });
}
