// @lit-protocol/vincent-app-sdk/dist/src/react/jwtProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JwtProvider = exports.JwtContext = void 0;
exports.useJwtContext = useJwtContext;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importStar(require("react"));
const jwt_1 = require("../jwt");
const useVincentWebAuthClient_1 = require("./useVincentWebAuthClient");
function jwtContextNotInitialized() {
    throw new Error('JwtContext must be used within an JwtProvider');
}
/**
 * React Context that exposes JWT authentication state and actions for Vincent apps.
 *
 * This context carries a value of type `JwtContextType` with:
 * - `authInfo`: `AuthInfo | null` — Authenticated user/application information or `null` when not authenticated.
 * - `loading`: `boolean` — Indicates whether an authentication operation (login/logout/validation) is in progress.
 * - `connect(redirectUri: string)`: Redirects the user to the Vincent consent page to initiate authentication.
 * - `loginWithJwt()`: Attempts to retrieve and validate a JWT from Vincent dashboard.
 * - `logOut()`: Removes the stored JWT and resets the authentication state.
 *
 * The provider component `JwtProvider` must wrap your component tree for this context to be available and for
 * `useJwtContext` to work in your components.
 *
 * @example
 * ```tsx
 * <JwtProvider appId={123}>
 *   <App />
 * </JwtProvider>
 * ```
 *
 * @see JwtProvider
 * @see useJwtContext
 */
exports.JwtContext = (0, react_1.createContext)({
    authInfo: null,
    loading: false,
    connect: jwtContextNotInitialized,
    loginWithJwt: jwtContextNotInitialized,
    logOut: jwtContextNotInitialized,
});
/**
 * React hook to access the JWT authentication context.
 *
 * This hook provides access to authentication state and methods for managing JWT-based
 * authentication in Vincent applications. It must be used within a component that is a
 * descendant of JwtProvider.
 *
 * @example
 * ```tsx
 * import { useJwtContext } from '@lit-protocol/vincent-app-sdk/react';
 *
 * function AuthenticatedComponent() {
 *   const { authInfo, loading, loginWithJwt, logOut } = useJwtContext();
 *
 *   if (loading) {
 *     return <div>Loading authentication...</div>;
 *   }
 *
 *   if (!authInfo) {
 *     return (
 *       <button onClick={loginWithJwt}>
 *         Login
 *       </button>
 *     );
 *   }
 *
 *   return (
 *     <div>
 *       <p>Logged in with PKP: {authInfo.pkp.ethAddress}</p>
 *       <button onClick={logOut}>Logout</button>
 *     </div>
 *   );
 * }
 * ```
 *
 * @returns The JWT context containing authentication state and methods
 *
 * @see JwtContext
 * @see JwtProvider
 */
function useJwtContext() {
    return (0, react_1.useContext)(exports.JwtContext);
}
/**
 * React component that provides JWT authentication capabilities for Vincent applications.
 *
 * The JwtProvider handles JWT token management, including
 * - Retrieving and validating JWTs from the Vincent consent page
 * - Storing and retrieving JWTs from persistent storage
 * - Providing authentication state and methods to child components
 * - Managing login/logout flows
 *
 * It uses the Context API to make authentication information and methods available
 * throughout your application without prop drilling.
 *
 * @example
 * ```tsx
 * import { JwtProvider } from '@lit-protocol/vincent-app-sdk/react';
 *
 * function App() {
 *   return (
 *     <JwtProvider appId=<YOUR_VINCENT_APP_ID>>
 *       <YourApplication />
 *     </JwtProvider>
 *   );
 * }
 *
 * // In a child component:
 * function LoginButton() {
 *   const { authInfo, loading, connect, logOut } = useJwtContext();
 *
 *   if (loading) return <div>Loading...</div>;
 *
 *   if (authInfo) {
 *     return (
 *       <div>
 *         <p>Logged in as: {authInfo.pkp.ethAddress}</p>
 *         <button onClick={logOut}>Log out</button>
 *       </div>
 *     );
 *   }
 *
 *   return (
 *     <button
 *       onClick={() => connect(window.location.href)}
 *     >
 *       Login with Vincent
 *     </button>
 *   );
 * }
 * ```
 *
 * @param props - Props for the JwtProvider component
 * @param props.children - Child components that will have access to the JWT context
 * @param props.appId - Your Vincent App Id
 * @param props.storage - Optional custom storage implementation (defaults to localStorage)
 * @param props.storageKeyBuilder - Optional function to customize the storage key for JWT tokens
 *
 * @see JwtContext
 * @see useJwtContext
 */
const JwtProvider = ({ children, appId, storage = localStorage, storageKeyBuilder = (appId) => `vincent-${appId}-jwt`, }) => {
    const appJwtKey = storageKeyBuilder(appId);
    const vincentWebAppClient = (0, useVincentWebAuthClient_1.useVincentWebAuthClient)(appId);
    const [authInfo, setAuthInfo] = (0, react_1.useState)(null);
    const [loading, setLoading] = (0, react_1.useState)(true);
    const logOut = (0, react_1.useCallback)(async () => {
        try {
            setLoading(true);
            await storage.removeItem(appJwtKey);
            setAuthInfo(null);
        }
        finally {
            setLoading(false);
        }
    }, [appJwtKey, storage]);
    const connect = (0, react_1.useCallback)((redirectUri) => {
        // Redirect to Vincent Auth consent page with appId and version
        vincentWebAppClient.redirectToConnectPage({
            // connectPageUrl: `http://localhost:5173/`,
            redirectUri,
        });
    }, [vincentWebAppClient]);
    const getJwt = (0, react_1.useCallback)(async () => {
        if (vincentWebAppClient.uriContainsVincentJWT()) {
            const jwtResult = await vincentWebAppClient.decodeVincentJWTFromUri(window.location.origin);
            if (!jwtResult) {
                return null;
            }
            const { decodedJWT, jwtStr } = jwtResult;
            await storage.setItem(appJwtKey, jwtStr);
            vincentWebAppClient.removeVincentJWTFromURI();
            return { jwtStr, decodedJWT };
        }
        const existingJwtStr = await storage.getItem(appJwtKey);
        if (!existingJwtStr) {
            return null;
        }
        const decodedJWT = await (0, jwt_1.verifyVincentAppUserJWT)({
            expectedAudience: window.location.origin,
            jwt: existingJwtStr,
            requiredAppId: appId,
        });
        return { jwtStr: existingJwtStr, decodedJWT };
    }, [appJwtKey, storage, vincentWebAppClient]);
    const loginWithJwt = (0, react_1.useCallback)(async () => {
        try {
            setLoading(true);
            const jwtResult = await getJwt();
            if (!jwtResult) {
                throw new Error('Could not get JWT');
            }
            const { decodedJWT, jwtStr } = jwtResult;
            setAuthInfo({
                app: decodedJWT.payload.app,
                authentication: decodedJWT.payload.authentication,
                jwt: jwtStr,
                pkp: decodedJWT.payload.pkpInfo,
            });
        }
        catch (error) {
            console.error(`Error logging in with JWT. Need to relogin: ${error.message}`);
            await logOut();
        }
        finally {
            setLoading(false);
        }
    }, [getJwt, logOut]);
    const value = (0, react_1.useMemo)(() => ({
        authInfo,
        connect,
        loading,
        loginWithJwt,
        logOut,
    }), [authInfo, connect, loading, loginWithJwt, logOut]);
    (0, react_1.useEffect)(() => {
        void loginWithJwt();
    }, [loginWithJwt]);
    return react_1.default.createElement(exports.JwtContext.Provider, { value: value }, children);
};
exports.JwtProvider = JwtProvider;
