// @lit-protocol/vincent-contracts-sdk/dist/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FEE_DIAMOND_ABI = exports.VINCENT_LIT_ACTIONS_ADDRESS_BOOK = exports.VINCENT_CONTRACT_ADDRESS_BOOK = exports.DEFAULT_PAGE_SIZE = exports.GAS_ADJUSTMENT_PERCENT = exports.COMBINED_ABI = exports.VINCENT_DIAMOND_CONTRACT_ADDRESS_PROD = void 0;
const tslib_1 = require("tslib");
const FeeDiamond_abi_json_1 = tslib_1.__importDefault(require("../abis/FeeDiamond.abi.json"));
exports.FEE_DIAMOND_ABI = FeeDiamond_abi_json_1.default;
const VincentAppFacet_abi_json_1 = tslib_1.__importDefault(require("../abis/VincentAppFacet.abi.json"));
const VincentAppViewFacet_abi_json_1 = tslib_1.__importDefault(require("../abis/VincentAppViewFacet.abi.json"));
const VincentERC2771Facet_abi_json_1 = tslib_1.__importDefault(require("../abis/VincentERC2771Facet.abi.json"));
const VincentUserFacet_abi_json_1 = tslib_1.__importDefault(require("../abis/VincentUserFacet.abi.json"));
const VincentUserViewFacet_abi_json_1 = tslib_1.__importDefault(require("../abis/VincentUserViewFacet.abi.json"));
const buildDiamondInterface_1 = require("./buildDiamondInterface");
// TODO!: Pull from the ABI after re-publishing
exports.VINCENT_DIAMOND_CONTRACT_ADDRESS_PROD = '0x1599E2c248C34833A606EB1dFeA86b839170d49f';
exports.COMBINED_ABI = (0, buildDiamondInterface_1.buildDiamondInterface)([
    VincentAppFacet_abi_json_1.default,
    VincentAppViewFacet_abi_json_1.default,
    VincentUserFacet_abi_json_1.default,
    VincentUserViewFacet_abi_json_1.default,
    VincentERC2771Facet_abi_json_1.default,
]);
exports.GAS_ADJUSTMENT_PERCENT = 120;
exports.DEFAULT_PAGE_SIZE = '50';
exports.VINCENT_CONTRACT_ADDRESS_BOOK = {
    fee: {
        175188: {
            chainName: 'chronicleYellowstone',
            address: '0x1599E2c248C34833A606EB1dFeA86b839170d49f',
            salt: 'VincentCreate2Salt_2',
        },
        8453: {
            chainName: 'base',
            address: '0x1599E2c248C34833A606EB1dFeA86b839170d49f',
            salt: 'VincentCreate2Salt_2',
        },
        84532: {
            chainName: 'baseSepolia',
            address: '0x1599E2c248C34833A606EB1dFeA86b839170d49f',
            salt: 'VincentCreate2Salt_2',
        },
    },
};
exports.VINCENT_LIT_ACTIONS_ADDRESS_BOOK = {
    signOwnerAttestation: {
        ipfsCid: 'QmfBfiD5XFQAagwcYgCEncpssvQJYkosbbww53t7EmEKdE',
        derivedActionPubkey: '0x0457611cae6d4c24e0f6d963181f503e56b308a8df43fc71584f2f586e8322ae3063a78a494e8fa1fdbca34b9dd2e2778b19eba5e7d54987148621e768b25ca9d3',
        derivedActionAddress: '0x32a2F1243F2402166c95Ef64b2a6ED932B04b7dC',
    },
};
