// @account-kit/smart-contracts/dist/esm/src/ma-v2/permissionBuilderErrors.js
import { BaseError } from "@aa-sdk/core";
export class PermissionBuilderError extends BaseError {
}
export class RootPermissionOnlyError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating that an account could not be found to execute the specified action.
     *
     * @param {Permission} permission The permission trying to be added atop the root permission
     */
    constructor(permission) {
        super(`Adding ${permission.type}: Cannot add permissions with ROOT permission`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: RootPermissionOnlyError"
        });
    }
}
export class AccountAddressAsTargetError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating that account address is used as target.
     *
     * @param {Permission} permission The permission with account address as target
     */
    constructor(permission) {
        super(`${permission.type}: Account address as target, use ACCOUNT_FUNCTIONS for account address`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: AccountAddressAsTargetError"
        });
    }
}
export class DuplicateTargetAddressError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating duplicate target address in permissions.
     *
     * @param {Permission} permission The permission with duplicate target address
     * @param {Address} targetAddress The duplicate target address
     */
    constructor(permission, targetAddress) {
        super(`${permission.type}: Address ${targetAddress} already has a permission. Cannot add multiple CONTRACT_ACCESS or FUNCTIONS_ON_CONTRACT permissions for the same target address.`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: DuplicateTargetAddressError"
        });
    }
}
export class NoFunctionsProvidedError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating no functions were provided.
     *
     * @param {Permission} permission The permission missing functions
     */
    constructor(permission) {
        super(`${permission.type}: No functions provided`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: NoFunctionsProvidedError"
        });
    }
}
export class ExpiredDeadlineError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating the deadline has expired.
     *
     * @param {number} deadline The expired deadline timestamp
     * @param {number} currentTime The current timestamp
     */
    constructor(deadline, currentTime) {
        super(`compileDeferred(): deadline ${deadline} cannot be before now (${currentTime})`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: ExpiredDeadlineError"
        });
    }
}
export class DeadlineOverLimitError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating the deadline has expired.
     *
     * @param {number} deadline The expired deadline timestamp
     */
    constructor(deadline) {
        super(`compileDeferred(): deadline ${deadline} cannot be > max uint48 (2^48 - 1)`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: DeadlineOverLimitError"
        });
    }
}
export class ValidationConfigUnsetError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating the validation config is unset.
     */
    constructor() {
        super("Missing permission among: functions on contract, functions on all contracts, account functions, contract access, or erc20 token transfer");
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: ValidationConfigUnsetError"
        });
    }
}
export class MultipleNativeTokenTransferError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating multiple native token transfer permissions.
     *
     * @param {Permission} permission The duplicate native token transfer permission
     */
    constructor(permission) {
        super(`${permission.type}: Must have at most ONE native token transfer permission`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: MultipleNativeTokenTransferError"
        });
    }
}
export class ZeroAddressError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating zero address was provided.
     *
     * @param {Permission} permission The permission with zero address
     */
    constructor(permission) {
        super(`${permission.type}: Zero address provided`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: ZeroAddressError"
        });
    }
}
export class MultipleGasLimitError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating multiple gas limit permissions.
     *
     * @param {Permission} permission The duplicate gas limit permission
     */
    constructor(permission) {
        super(`${permission.type}: Must have at most ONE gas limit permission`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: MultipleGasLimitError"
        });
    }
}
export class UnsupportedPermissionTypeError extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating an unsupported permission type.
     */
    constructor() {
        super(`Unsupported permission type`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "PermissionBuilder: UnsupportedPermissionTypeError"
        });
    }
}
export class SelectorNotAllowed extends PermissionBuilderError {
    /**
     * Constructor for initializing an error message indicating that the selector being added is not allowed.
     *
     * @param {string} functionName The function name of the selector that is being added.
     */
    constructor(functionName) {
        super(`Cannot add ${functionName} on the account`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "SelectorNotAllowed"
        });
    }
}
