// @lit-protocol/vincent-contracts-sdk/dist/src/internal/app/App.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerApp = registerApp;
exports.registerNextVersion = registerNextVersion;
exports.enableAppVersion = enableAppVersion;
exports.addDelegatee = addDelegatee;
exports.removeDelegatee = removeDelegatee;
exports.setDelegatee = setDelegatee;
exports.deleteApp = deleteApp;
exports.undeleteApp = undeleteApp;
const utils_1 = require("../../utils");
async function registerApp(params) {
    const { contract, args: { delegateeAddresses, versionAbilities }, overrides, } = params;
    try {
        const adjustedOverrides = await (0, utils_1.gasAdjustedOverrides)(contract, 'registerApp', [delegateeAddresses, versionAbilities], overrides);
        const tx = await contract.registerApp(delegateeAddresses, versionAbilities, {
            ...adjustedOverrides,
        });
        const receipt = await tx.wait();
        const appEvent = (0, utils_1.findEventByName)(contract, receipt.logs, 'NewAppRegistered');
        if (!appEvent) {
            throw new Error('NewAppRegistered event not found');
        }
        const parsedAppEvent = contract.interface.parseLog(appEvent);
        const appId = parsedAppEvent?.args?.appId;
        const accountIndexHash = parsedAppEvent?.args?.accountIndexHash;
        if (appId === undefined || accountIndexHash === undefined) {
            throw new Error('NewAppRegistered event missing appId or accountIndexHash');
        }
        const versionEvent = (0, utils_1.findEventByName)(contract, receipt.logs, 'NewAppVersionRegistered');
        if (!versionEvent) {
            throw new Error('NewAppVersionRegistered event not found');
        }
        const parsedVersionEvent = contract.interface.parseLog(versionEvent);
        const newAppVersion = parsedVersionEvent?.args?.appVersion;
        if (newAppVersion === undefined) {
            throw new Error('NewAppVersionRegistered event missing appVersion');
        }
        const appIdNumber = Number(appId);
        const newAppVersionNumber = Number(newAppVersion);
        const accountIndexHashValue = accountIndexHash.toString();
        const isZeroAccountIndexHash = /^0x0{64}$/i.test(accountIndexHashValue);
        if (appIdNumber <= 0) {
            throw new Error(`Invalid appId parsed from NewAppRegistered: ${appIdNumber}`);
        }
        if (newAppVersionNumber <= 0) {
            throw new Error(`Invalid appVersion parsed from NewAppVersionRegistered: ${newAppVersionNumber}`);
        }
        if (isZeroAccountIndexHash) {
            throw new Error('Invalid accountIndexHash parsed from NewAppRegistered');
        }
        return {
            txHash: tx.hash,
            appId: appIdNumber,
            newAppVersion: newAppVersionNumber,
            accountIndexHash: accountIndexHashValue,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Register App: ${decodedError}`);
    }
}
async function registerNextVersion(params) {
    const { contract, args: { appId, versionAbilities }, overrides, } = params;
    try {
        const adjustedOverrides = await (0, utils_1.gasAdjustedOverrides)(contract, 'registerNextAppVersion', [appId, versionAbilities], overrides);
        const tx = await contract.registerNextAppVersion(appId, versionAbilities, {
            ...adjustedOverrides,
        });
        const receipt = await tx.wait();
        const event = (0, utils_1.findEventByName)(contract, receipt.logs, 'NewAppVersionRegistered');
        if (!event) {
            throw new Error('NewAppVersionRegistered event not found');
        }
        const newAppVersion = contract.interface.parseLog(event)?.args?.appVersion;
        if (!newAppVersion) {
            throw new Error('NewAppVersionRegistered event does not contain appVersion argument');
        }
        return {
            txHash: tx.hash,
            newAppVersion,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Register Next Version: ${decodedError}`);
    }
}
async function enableAppVersion(params) {
    const { contract, args: { appId, appVersion, enabled }, overrides, } = params;
    try {
        const adjustedOverrides = await (0, utils_1.gasAdjustedOverrides)(contract, 'enableAppVersion', [appId, appVersion, enabled], overrides);
        const tx = await contract.enableAppVersion(appId, appVersion, enabled, {
            ...adjustedOverrides,
        });
        await tx.wait();
        return {
            txHash: tx.hash,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Enable App Version: ${decodedError}`);
    }
}
async function addDelegatee(params) {
    const { contract, args: { appId, delegateeAddress }, overrides, } = params;
    try {
        const adjustedOverrides = await (0, utils_1.gasAdjustedOverrides)(contract, 'addDelegatee', [appId, delegateeAddress], overrides);
        const tx = await contract.addDelegatee(appId, delegateeAddress, {
            ...adjustedOverrides,
        });
        await tx.wait();
        return {
            txHash: tx.hash,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Add Delegatee: ${decodedError}`);
    }
}
async function removeDelegatee(params) {
    const { contract, args: { appId, delegateeAddress }, overrides, } = params;
    try {
        const adjustedOverrides = await (0, utils_1.gasAdjustedOverrides)(contract, 'removeDelegatee', [appId, delegateeAddress], overrides);
        const tx = await contract.removeDelegatee(appId, delegateeAddress, {
            ...adjustedOverrides,
        });
        await tx.wait();
        return {
            txHash: tx.hash,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Remove Delegatee: ${decodedError}`);
    }
}
async function setDelegatee(params) {
    const { contract, args: { appId, delegateeAddresses }, overrides, } = params;
    try {
        const adjustedOverrides = await (0, utils_1.gasAdjustedOverrides)(contract, 'setDelegatee', [appId, delegateeAddresses], overrides);
        const tx = await contract.setDelegatee(appId, delegateeAddresses, {
            ...adjustedOverrides,
        });
        await tx.wait();
        return {
            txHash: tx.hash,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Set Delegatee: ${decodedError}`);
    }
}
async function deleteApp(params) {
    const { contract, args: { appId }, overrides, } = params;
    try {
        const adjustedOverrides = await (0, utils_1.gasAdjustedOverrides)(contract, 'deleteApp', [appId], overrides);
        const tx = await contract.deleteApp(appId, {
            ...adjustedOverrides,
        });
        await tx.wait();
        return {
            txHash: tx.hash,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Delete App: ${decodedError}`);
    }
}
async function undeleteApp(params) {
    const { contract, args: { appId }, overrides, } = params;
    try {
        const adjustedOverrides = await (0, utils_1.gasAdjustedOverrides)(contract, 'undeleteApp', [appId], overrides);
        const tx = await contract.undeleteApp(appId, {
            ...adjustedOverrides,
        });
        await tx.wait();
        return {
            txHash: tx.hash,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Undelete App: ${decodedError}`);
    }
}
