// @lit-protocol/vincent-contracts-sdk/dist/src/internal/app/AppView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAppById = getAppById;
exports.getAppIdByDelegatee = getAppIdByDelegatee;
exports.getAppVersion = getAppVersion;
exports.getAppsByManagerAddress = getAppsByManagerAddress;
exports.getAppByDelegateeAddress = getAppByDelegateeAddress;
exports.getDelegatedAgentAddresses = getDelegatedAgentAddresses;
const utils_1 = require("../../utils");
async function getAppById(params) {
    const { args: { appId }, contract, } = params;
    try {
        const chainApp = await contract.getAppById(appId);
        const { delegatees, ...app } = chainApp;
        return {
            ...app,
            id: app.id,
            latestVersion: app.latestVersion,
            delegateeAddresses: delegatees,
            accountIndexHash: app.accountIndexHash,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        // Check if the error is due to AppNotRegistered
        if (decodedError.includes('AppNotRegistered')) {
            return null;
        }
        throw new Error(`Failed to Get App By ID: ${decodedError}`);
    }
}
async function getAppIdByDelegatee(params) {
    const { args: { delegateeAddress }, contract, } = params;
    try {
        const app = await contract.getAppByDelegatee(delegateeAddress);
        return app.id;
    }
    catch (error) {
        const decodedError = error instanceof Error ? error.message : String(error);
        if (decodedError.includes('DelegateeNotRegistered')) {
            return null;
        }
        throw new Error(`Failed to Get App ID By Delegatee: ${decodedError}`);
    }
}
async function getAppVersion(params) {
    const { args: { appId, version }, contract, } = params;
    try {
        const appVersion = await contract.getAppVersion(appId, version);
        const convertedAppVersion = {
            version: appVersion.version,
            accountIndexHash: appVersion.accountIndexHash,
            enabled: appVersion.enabled,
            delegatedAgentAddresses: appVersion.delegatedAgents,
            abilities: appVersion.abilities.map((ability) => ({
                abilityIpfsCid: ability.abilityIpfsCid,
                policyIpfsCids: ability.policyIpfsCids,
            })),
        };
        return {
            appVersion: convertedAppVersion,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        // Check if the error is due to AppVersionNotRegistered
        if (decodedError.includes('AppVersionNotRegistered') ||
            decodedError.includes('AppNotRegistered')) {
            return null;
        }
        throw new Error(`Failed to Get App Version: ${decodedError}`);
    }
}
async function getAppsByManagerAddress(params) {
    const { args: { managerAddress, offset }, contract, } = params;
    try {
        const [appIds, appVersionCounts] = await contract.getAppsByManager(managerAddress, offset);
        return appIds.map((id, idx) => ({
            id,
            versionCount: appVersionCounts[idx],
        }));
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        // Check if the error is due to NoAppsFoundForManager or ZeroAddressNotAllowed
        if (decodedError.includes('NoAppsFoundForManager')) {
            return [];
        }
        throw new Error(`Failed to Get Apps By Manager: ${decodedError}`);
    }
}
async function getAppByDelegateeAddress(params) {
    const { args: { delegateeAddress }, contract, } = params;
    try {
        const chainApp = await contract.getAppByDelegatee(delegateeAddress);
        const { delegatees, ...app } = chainApp;
        return {
            ...app,
            delegateeAddresses: delegatees,
            id: app.id,
            latestVersion: app.latestVersion,
            accountIndexHash: app.accountIndexHash,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        // Check if the error is due to DelegateeNotRegistered or ZeroAddressNotAllowed
        if (decodedError.includes('DelegateeNotRegistered')) {
            return null;
        }
        throw new Error(`Failed to Get App By Delegatee: ${decodedError}`);
    }
}
async function getDelegatedAgentAddresses(params) {
    const { args: { appId, offset, version }, contract, } = params;
    try {
        const delegatedAgentAddresses = await contract.getDelegatedAgentAddresses(appId, version, offset);
        return delegatedAgentAddresses;
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Get Delegated Agent Addresses: ${decodedError}`);
    }
}
