// @lit-protocol/vincent-ability-sdk/dist/src/lib/policyCore/helpers/zod.js
"use strict";
// src/lib/policyCore/helpers/zod.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolicyResponseShape = void 0;
exports.validateOrDeny = validateOrDeny;
exports.getValidatedParamsOrDeny = getValidatedParamsOrDeny;
exports.getSchemaForPolicyResponseResult = getSchemaForPolicyResponseResult;
const zod_1 = require("zod");
const utils_1 = require("../../utils");
const resultCreators_1 = require("./resultCreators");
const typeGuards_1 = require("./typeGuards");
/**
 * Matches the minimum structure of a PolicyResponse.
 * This is useful when validating that a response shape is at least plausible.
 */
exports.PolicyResponseShape = zod_1.z.object({
    allow: zod_1.z.boolean(),
    result: zod_1.z.unknown(),
});
/**
 * Validates a value using a Zod schema (or requires undefined if none given).
 * Returns parsed result or a standardized deny object.
 *
 * @param value - The raw value to validate
 * @param schema - A Zod schema to apply, or undefined (meaning: must be undefined)
 * @param phase - One of 'evaluate' | 'precheck' | 'commit'
 * @param stage - 'parameters' (before calling the lifecycle phase) or 'result' (after receiving the result)
 *
 * @returns The parsed result, or a PolicyResponseDeny with a ZodValidationDenyResult
 */
function validateOrDeny(value, schema, phase, stage) {
    const parsed = schema.safeParse(value);
    if (!parsed.success) {
        const descriptor = stage === 'input' ? 'parameters' : 'result';
        const message = `Invalid ${phase} ${descriptor}.`;
        return (0, resultCreators_1.createDenyResult)({
            runtimeError: message,
            schemaValidationError: {
                zodError: parsed.error,
                phase,
                stage,
            },
        });
    }
    return parsed.data;
}
function getValidatedParamsOrDeny({ rawAbilityParams, rawUserParams, abilityParamsSchema, userParamsSchema, phase, }) {
    const abilityParams = validateOrDeny(rawAbilityParams, abilityParamsSchema, phase, 'input');
    if ((0, typeGuards_1.isPolicyDenyResponse)(abilityParams)) {
        return abilityParams;
    }
    const userParams = validateOrDeny(rawUserParams, userParamsSchema, phase, 'input');
    if ((0, typeGuards_1.isPolicyDenyResponse)(userParams)) {
        return userParams;
    }
    return {
        abilityParams,
        userParams,
    };
}
/**
 * Given an unknown policy response result and the known allow/deny schemas,
 * this function returns the appropriate Zod schema to use when validating `.result`.
 *
 * - If the response shape is invalid, returns a Zod schema matching the PolicyResponse structure.
 * - If the shape is valid, returns either the allow or deny result schema.
 */
function getSchemaForPolicyResponseResult({ value, allowResultSchema, denyResultSchema, }) {
    if (!(0, typeGuards_1.isPolicyResponse)(value)) {
        console.log('getSchemaForPolicyResponseResult !isPolicyResponse', JSON.stringify(value, utils_1.bigintReplacer));
        return {
            schemaToUse: exports.PolicyResponseShape,
            parsedType: 'unknown',
        };
    }
    console.log('getSchemaForPolicyResponseResult value is', JSON.stringify(value, utils_1.bigintReplacer));
    return {
        schemaToUse: value.allow ? allowResultSchema : denyResultSchema,
        parsedType: value.allow ? 'allow' : 'deny',
    };
}
