// @lit-protocol/vincent-app-sdk/dist/src/utils/vincentViemPkpSigner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createVincentViemPkpSigner = createVincentViemPkpSigner;
const accounts_1 = require("viem/accounts");
function createVincentViemPkpSigner(params) {
    const { pkpAddress, onSignUserOpHash } = params;
    let currentUserOp = null;
    const account = (0, accounts_1.toAccount)({
        address: pkpAddress,
        async signMessage({ message }) {
            console.log('[createVincentViemPkpSigner] signMessage called by ZeroDev');
            console.log('message:', message);
            console.log('message type:', typeof message);
            console.log('currentUserOp status:', currentUserOp ? 'SET' : 'NOT SET');
            let userOpHash;
            if (typeof message === 'object' && 'raw' in message) {
                userOpHash = message.raw;
                console.log('message.raw (UserOp hash):', userOpHash);
                if (!currentUserOp) {
                    throw new Error('[createVincentViemPkpSigner] currentUserOp is null, wrapKernelAccountWithUserOpCapture did not intercept correctly');
                }
                console.log('Full UserOp available for Lit Action:', JSON.stringify(currentUserOp, (key, value) => (typeof value === 'bigint' ? value.toString() : value), 2));
                // Delegate signing to Lit Actions via callback
                const signature = await onSignUserOpHash({
                    userOpHash,
                    userOp: currentUserOp,
                });
                // Clear the current UserOp after signing
                currentUserOp = null;
                return signature;
            }
            throw new Error('[createVincentViemPkpSigner] signMessage called without UserOp hash');
        },
        async signTransaction(transaction, options) {
            console.log('[createVincentViemPkpSigner] signTransaction called by ZeroDev');
            console.log('transaction:', JSON.stringify(transaction, null, 2));
            console.log('options:', options);
            throw new Error('[createVincentViemPkpSigner] signTransaction not supported - use signMessage for UserOps');
        },
        async signTypedData(parameters) {
            console.log('[createVincentViemPkpSigner] signTypedData called by ZeroDev');
            console.log('parameters:', JSON.stringify(parameters, null, 2));
            throw new Error('[createVincentViemPkpSigner] signTypedData not supported - use signMessage for UserOps');
        },
    });
    // Add method to set the current UserOp before signing
    account.setCurrentUserOp = (userOp) => {
        currentUserOp = userOp;
    };
    return account;
}
