// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/vincentGatedSignerAbility.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createVincentGatedSignerAbility = createVincentGatedSignerAbility;
const helpers_1 = require("../abilityCore/helpers");
const vincentAbility_1 = require("../abilityCore/vincentAbility");
const lifecycleFunctions_1 = require("./helpers/lifecycleFunctions");
const schemas_1 = require("./helpers/schemas");
// https://www.notion.so/litprotocol/Vincent-2-0-abilities-development-proposal-2b16f449b04180f4a01fe5088fd0a2b1
/**
 * Creates a Vincent Gated Signer Ability.
 *
 * @param AbilityConfig - The configuration for the ability.
 * @returns A Vincent Ability instance configured as a gated signer.
 *
 * @example
 * ```ts
 * export const myAbility = createVincentGatedSignerAbility({
 *   packageName: '@my-org/my-ability',
 *   abilityDescription: 'My ability description',
 *   decodeTransaction,
 *   validateSimulation,
 *   validateTransaction,
 * });
 * ```
 */
function createVincentGatedSignerAbility(AbilityConfig) {
    // TODO No policies for this type of abilities yet
    const supportedPolicies = (0, helpers_1.supportedPoliciesForAbility)([]);
    const { abilityDescription, decodeTransaction, packageName, validateSimulation, validateTransaction, } = AbilityConfig;
    const { precheck, execute } = (0, lifecycleFunctions_1.buildLifecycleFunctions)({
        decodeTransaction,
        validateSimulation,
        validateTransaction,
    });
    return (0, vincentAbility_1.createVincentAbility)({
        packageName,
        abilityDescription,
        abilityParamsSchema: schemas_1.abilityParamsSchema,
        executeSuccessSchema: schemas_1.executeSuccessSchema,
        executeFailSchema: schemas_1.executeFailSchema,
        precheckSuccessSchema: schemas_1.precheckSuccessSchema,
        precheckFailSchema: schemas_1.precheckFailSchema,
        supportedPolicies,
        precheck,
        execute,
    });
}
