// @account-kit/smart-contracts/dist/esm/src/msca/client/client.js
import { createSmartAccountClient, smartAccountClientActions, } from "@aa-sdk/core";
import {} from "viem";
import { accountLoupeActions, } from "../account-loupe/decorator.js";
import { createMultiOwnerModularAccount, } from "../account/multiOwnerAccount.js";
import { createMultisigModularAccount, } from "../account/multisigAccount.js";
import { pluginManagerActions, } from "../plugin-manager/decorator.js";
import { multiOwnerPluginActions, } from "../plugins/multi-owner/index.js";
import { multisigPluginActions, } from "../plugins/multisig/index.js";
import { multisigSignatureMiddleware } from "../plugins/multisig/middleware.js";
import { createAlchemySmartAccountClient, isAlchemyTransport, } from "@account-kit/infra";
/**
 * Creates a multi-owner modular account client with the provided parameters including account, transport, chain, and additional client configuration. This function uses a modular account and extends it with various plugin actions.
 *
 * @example
 * ```ts
 * import { createMultiOwnerModularAccountClient } from "@account-kit/smart-contracts";
 * import { LocalAccountSigner } from "@aa-sdk/core";
 * import { sepolia } from "viem/chains";
 * import { http } from "viem";
 * import { generatePrivateKey } from "viem/accounts";
 *
 * const accountClient = await createMultiOwnerModularAccountClient({
 *  chain: sepolia,
 *  transport: http("RPC_URL"),
 *  signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey())
 * });
 * ```
 * @example
 * ```ts
 * import { createMultiOwnerModularAccountClient } from "@account-kit/smart-contracts";
 * import { sepolia, alchemy } from "@account-kit/infra";
 * import { LocalAccountSigner } from "@aa-sdk/core";
 * import { generatePrivateKey } from "viem"
 *
 * const alchemyAccountClient = await createMultiOwnerModularAccountClient({
 *  transport: alchemy({ apiKey: "your-api-key" }),
 *  chain: sepolia,
 *  signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey())
 * });
 * ```
 *
 * @param {CreateMultiOwnerModularAccountClientParams} config The parameters for creating the multi-owner modular account client
 * @returns {Promise<SmartAccountClient>} A promise that resolves to a `SmartAccountClient` instance with extended plugin actions
 */
export async function createMultiOwnerModularAccountClient({ transport, chain, ...params }) {
    const modularAccount = await createMultiOwnerModularAccount({
        ...params,
        transport,
        chain,
    });
    if (isAlchemyTransport(transport, chain)) {
        const { opts } = params;
        return createAlchemySmartAccountClient({
            ...params,
            account: modularAccount,
            transport,
            chain,
            opts,
        })
            .extend(multiOwnerPluginActions)
            .extend(pluginManagerActions)
            .extend(accountLoupeActions);
    }
    return createSmartAccountClient({
        ...params,
        transport,
        chain,
        account: modularAccount,
    })
        .extend(pluginManagerActions)
        .extend(multiOwnerPluginActions)
        .extend(accountLoupeActions);
}
/**
 * Creates a multisig modular account client using the provided parameters including account details, transport, chain, and additional client configuration. This function constructs the multisig modular account and extends it with various actions to create a comprehensive client.
 *
 * @example
 * ```ts
 * import { createMultisigModularAccountClient } from "@account-kit/smart-contracts";
 * import { LocalAccountSigner } from "@aa-sdk/core";
 * import { sepolia } from "viem/chains";
 * import { http } from "viem"
 * import { generatePrivateKey } from "viem/accounts";
 *
 * const accountClient = await createMultisigModularAccountClient({
 *  chain: sepolia,
 *  transport: http("RPC_URL"),
 *  signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey()),
 *  owners: [], // other owners on the account
 *  threshold: 2, // 2 of N signatures
 * });
 * ```
 * @example
 * ```ts
 * import { createMultisigModularAccountClient } from "@account-kit/smart-contracts";
 * import { sepolia } from "@account-kit/infra";
 * import { LocalAccountSigner } from "@aa-sdk/core";
 * import { generatePrivateKey } from "viem"
 *
 * const alchemyAccountClient = await createMultisigModularAccountClient({
 *  transport: alchemy({ apiKey: "your-api-key" }),
 *  chain: sepolia,
 *  signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey()),
 *  owners: [...], // other owners on the account
 *  threshold: 2, // 2 of N signatures
 * });
 * ```
 *
 * @param {CreateMultisigModularAccountClientParams} config the parameters for configuring the multisig modular account client
 * @returns {Promise<SmartAccountClient<Transport, Chain, MultisigModularAccount<SmartAccountSigner>, {}, SmartAccountClientRpcSchema, MultisigUserOperationContext>>} a promise that resolves to a `SmartAccountClient` object extended with the multisig modular account and additional actions
 */
export async function createMultisigModularAccountClient({ transport, chain, ...params }) {
    const modularAccount = await createMultisigModularAccount({
        ...params,
        transport,
        chain,
    });
    if (isAlchemyTransport(transport, chain)) {
        // Need to fit the type into this since the previous multiSigAlchemyClient had it at this point, but without an Value as Type should be safe
        // And the createAlchemySmartAccountClient signUserOperation could not infer without this
        let config = {
            ...params,
            chain,
            transport,
        };
        const { opts } = config;
        return createAlchemySmartAccountClient({
            ...config,
            account: modularAccount,
            opts,
            signUserOperation: multisigSignatureMiddleware,
        })
            .extend(smartAccountClientActions)
            .extend(multisigPluginActions)
            .extend(pluginManagerActions)
            .extend(accountLoupeActions);
    }
    const client = createSmartAccountClient({
        ...params,
        transport,
        chain,
        account: modularAccount,
        signUserOperation: multisigSignatureMiddleware,
    })
        .extend(smartAccountClientActions)
        .extend(pluginManagerActions)
        .extend(multisigPluginActions)
        .extend(accountLoupeActions);
    return client;
}
