// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/helpers/toLitActionAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toLitActionAccount = toLitActionAccount;
const viem_1 = require("viem");
const accounts_1 = require("viem/accounts");
const utils_1 = require("viem/utils");
function toLitActionAccount(pkpPublicKey) {
    const pkpEthAddress = (0, utils_1.publicKeyToAddress)(pkpPublicKey);
    const pkpPublicKeyForLit = pkpPublicKey.replace('0x', '');
    return (0, accounts_1.toAccount)({
        address: pkpEthAddress,
        async signMessage({ message }) {
            const hashedMessage = (0, viem_1.hashMessage)(message);
            const signature = await Lit.Actions.signAndCombineEcdsa({
                toSign: (0, viem_1.hexToBytes)(hashedMessage),
                publicKey: pkpPublicKeyForLit,
                sigName: 'signMessage',
            });
            const structuredSignature = JSON.parse(signature);
            return (0, viem_1.serializeSignature)({
                r: `0x${structuredSignature.r.substring(2)}`,
                s: `0x${structuredSignature.s}`,
                yParity: structuredSignature.v,
            });
        },
        async signTransaction(transaction, { serializer } = {}) {
            const signableTransaction = {
                ...transaction,
                chainId: transaction.chainId ? Number(transaction.chainId) : undefined,
            };
            const serializerToUse = serializer || viem_1.serializeTransaction;
            const serializedTx = (await serializerToUse(signableTransaction));
            const txHash = (0, viem_1.keccak256)(serializedTx);
            const toSign = (0, viem_1.hexToBytes)(txHash);
            const signature = await Lit.Actions.signAndCombineEcdsa({
                toSign,
                publicKey: pkpPublicKeyForLit,
                sigName: 'signTransaction',
            });
            const structuredSignature = JSON.parse(signature);
            return serializerToUse(signableTransaction, {
                r: `0x${structuredSignature.r.substring(2)}`,
                s: `0x${structuredSignature.s}`,
                yParity: structuredSignature.v,
            });
        },
        async signTypedData(typedData) {
            const { domain, types, primaryType, message } = typedData;
            const hashedTypedData = (0, viem_1.hashTypedData)({
                domain,
                types,
                primaryType,
                message,
            });
            const signature = await Lit.Actions.signAndCombineEcdsa({
                toSign: (0, viem_1.hexToBytes)(hashedTypedData),
                publicKey: pkpPublicKeyForLit,
                sigName: 'signTypedData',
            });
            const structuredSignature = JSON.parse(signature);
            return (0, viem_1.serializeSignature)({
                r: `0x${structuredSignature.r.substring(2)}`,
                s: `0x${structuredSignature.s}`,
                yParity: structuredSignature.v,
            });
        },
    });
}
