// @lit-protocol/vincent-ability-sdk/dist/src/lib/gatedSigner/helpers/userOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.userOpSchema = exports.userOpV070Schema = void 0;
exports.toVincentUserOp = toVincentUserOp;
const viem_1 = require("viem");
const zod_1 = require("zod");
const hex_1 = require("./hex");
const userOpBaseSchema = zod_1.z.object({
    sender: hex_1.addressSchema.describe('The account making the operation'),
    nonce: zod_1.z
        .string()
        .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]*|0)$/)
        .describe('Account nonce or creation salt'),
    callData: hex_1.hexSchema.describe('Data for operation call'),
    callGasLimit: zod_1.z
        .string()
        .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,15})|0$/)
        .describe('Gas allocated for call'),
    verificationGasLimit: zod_1.z
        .string()
        .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,15})|0$/)
        .describe('Gas allocated for verification'),
    preVerificationGas: zod_1.z
        .string()
        .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,15})|0$/)
        .describe('Gas for pre-verification execution and calldata'),
    maxFeePerGas: zod_1.z
        .string()
        .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,15})|0$/)
        .describe('Maximum fee per gas (EIP-1559)'),
    maxPriorityFeePerGas: zod_1.z
        .string()
        .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,15})|0$/)
        .describe('Max priority fee per gas (EIP-1559)'),
    signature: hex_1.hexSchema.default('0x').describe('Data passed during verification.'),
    eip7702Auth: zod_1.z
        .object({
        chain_id: zod_1.z
            .string()
            .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,15})|0$/)
            .optional()
            .describe('The chain Id of the authorization'),
        address: hex_1.addressSchema.optional().describe('The address of the authorization'),
        nonce: zod_1.z
            .string()
            .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,15})|0$/)
            .optional()
            .describe('The nonce for the authorization'),
        y_parity: zod_1.z
            .string()
            .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]*|0)$/)
            .optional()
            .describe('Y parity of signed authorization tuple'),
        r: zod_1.z
            .string()
            .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,31})|0$/)
            .optional()
            .describe('R of signed authorization tuple'),
        s: zod_1.z
            .string()
            .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,31})|0$/)
            .optional()
            .describe('S of signed authorization tuple'),
    })
        .optional(),
});
exports.userOpV070Schema = userOpBaseSchema.extend({
    paymaster: hex_1.addressSchema.optional().describe('Paymaster contract address'),
    paymasterData: hex_1.hexSchema.optional().describe('Data for paymaster'),
    paymasterVerificationGasLimit: zod_1.z
        .string()
        .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,15})|0$/)
        .describe('The gas limit for paymaster verification.'),
    factory: hex_1.addressSchema
        .optional()
        .describe('The account factory address (needed if and only if the account is not yet on-chain and needs to be created)'),
    factoryData: hex_1.hexSchema
        .optional()
        .describe('Data for the account factory (only if the account factory exists)'),
    paymasterPostOpGasLimit: zod_1.z
        .string()
        .regex(/^0x([1-9a-fA-F]+[0-9a-fA-F]{0,15})|0$/)
        .describe('The amount of gas to allocate for the paymaster post-op code (only if a paymaster exists)'),
});
// Use z.union when adding new userOp versions
exports.userOpSchema = exports.userOpV070Schema;
const hexValues = [
    'callGasLimit',
    'maxFeePerGas',
    'maxPriorityFeePerGas',
    'nonce',
    'paymasterPostOpGasLimit',
    'paymasterVerificationGasLimit',
    'preVerificationGas',
    'verificationGasLimit',
];
function toVincentUserOp(userOp) {
    const _userOp = { ...userOp };
    for (const key of hexValues) {
        if (hexValues.includes(key) && _userOp[key] && !(0, viem_1.isHex)(_userOp[key])) {
            _userOp[key] = (0, viem_1.toHex)(_userOp[key]);
        }
    }
    return _userOp;
}
