// @account-kit/logging/dist/esm/index.js
import { noopLogger } from "./noop.js";
export function createLogger(_context) {
    // We are currently not logging events from the SDK (as Sept 2, 2025), so
    // a no-op logger is always used. This will be further explored in v5.
    const innerLogger = noopLogger;
    const logger = {
        ...innerLogger,
        profiled(name, func) {
            return function (...args) {
                const start = Date.now();
                const result = func.apply(this, args);
                if (result instanceof Promise) {
                    return result.then((res) => {
                        innerLogger.trackEvent({
                            name: "performance",
                            data: {
                                executionTimeMs: Date.now() - start,
                                functionName: name,
                            },
                        });
                        return res;
                    });
                }
                innerLogger.trackEvent({
                    name: "performance",
                    data: {
                        executionTimeMs: Date.now() - start,
                        functionName: name,
                    },
                });
                return result;
            };
        },
    };
    return logger;
}
