// @lit-protocol/vincent-ability-sdk/dist/src/lib/abilityHelpers/populateTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.populateTransaction = void 0;
const ethers_v6_1 = require("ethers-v6");
const populateTransaction = async ({ to, from, data, value, rpcUrl, chainId, gasBufferPercentage, baseFeePerGasBufferPercentage, }) => {
    if (gasBufferPercentage !== undefined && !Number.isInteger(gasBufferPercentage)) {
        throw new Error('[populateTransaction] gasBufferPercentage must be an integer');
    }
    if (baseFeePerGasBufferPercentage !== undefined &&
        !Number.isInteger(baseFeePerGasBufferPercentage)) {
        throw new Error('[populateTransaction] baseFeePerGasBufferPercentage must be an integer');
    }
    const provider = new ethers_v6_1.JsonRpcProvider(rpcUrl);
    const signer = new ethers_v6_1.VoidSigner(from, provider);
    const populatedTx = await signer.populateTransaction({
        to,
        from,
        data,
        value,
        chainId,
    });
    if (!populatedTx.gasLimit) {
        throw new Error(`[estimateGas] Unable to estimate gas for transaction: ${JSON.stringify({
            to,
            from,
            data,
            value,
        })}`);
    }
    if (gasBufferPercentage !== undefined) {
        populatedTx.gasLimit =
            (BigInt(populatedTx.gasLimit) * BigInt(100 + gasBufferPercentage)) / 100n;
    }
    const partialUnsignedTx = {
        to: populatedTx.to ?? undefined,
        nonce: populatedTx.nonce ?? undefined,
        gasLimit: (0, ethers_v6_1.toQuantity)(populatedTx.gasLimit),
        data: populatedTx.data ?? undefined,
        value: populatedTx.value ? (0, ethers_v6_1.toQuantity)(populatedTx.value) : '0x0',
        chainId: populatedTx.chainId ? (0, ethers_v6_1.toNumber)(populatedTx.chainId) : undefined,
        // Typed-Transaction features
        type: populatedTx.type ?? undefined,
        // EIP-2930; Type 1 & EIP-1559; Type 2
        accessList: populatedTx.accessList ?? undefined,
    };
    // Legacy fee path
    if (populatedTx.gasPrice != null) {
        return {
            ...partialUnsignedTx,
            gasPrice: (0, ethers_v6_1.toQuantity)(populatedTx.gasPrice),
        };
    }
    // EIP-1559 path
    if (populatedTx.maxFeePerGas == null) {
        throw new Error('[estimateGas] maxFeePerGas is missing from populated transaction');
    }
    if (populatedTx.maxPriorityFeePerGas == null) {
        throw new Error('[estimateGas] maxPriorityFeePerGas is missing from populated transaction');
    }
    if (baseFeePerGasBufferPercentage !== undefined) {
        // Apply baseFeePerGas buffer (e.g. 20% -> multiply by 1.2)
        populatedTx.maxFeePerGas =
            (BigInt(populatedTx.maxFeePerGas) * BigInt(100 + baseFeePerGasBufferPercentage)) / 100n;
    }
    return {
        ...partialUnsignedTx,
        maxFeePerGas: (0, ethers_v6_1.toQuantity)(populatedTx.maxFeePerGas),
        maxPriorityFeePerGas: (0, ethers_v6_1.toQuantity)(populatedTx.maxPriorityFeePerGas),
    };
};
exports.populateTransaction = populateTransaction;
