// @lit-protocol/vincent-app-sdk/dist/src/internal/LitNodeClient/LitNodeClientInstance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LitNodeClientInstance = void 0;
const lit_node_client_1 = require("@lit-protocol/lit-node-client");
class LitNodeClientInstance {
    constructor({ network, debug }) {
        this.isConnected = false;
        this.connectHandle = null;
        this.litNodeClientInstance = new lit_node_client_1.LitNodeClient({
            debug,
            litNetwork: network,
        });
    }
    async connect() {
        if (!this.isConnected) {
            // Coalesce concurrent calls
            if (this.connectHandle) {
                return this.connectHandle;
            }
            // Stash a handle so concurrent calls to connect are coaelesced into 1
            this.connectHandle = this.litNodeClientInstance.connect().then(() => true);
            try {
                // Don't return until we know the result of pending connect attempt
                await this.connectHandle;
                this.isConnected = true;
            }
            catch (e) {
                // We allow multiple calls to (retries!) to `connect()` even in case where one succeeded
                // if `isConnected` is false (e.g. a prior attempt failed)
                this.isConnected = false;
                throw e;
            }
            finally {
                this.connectHandle = null;
            }
            return this.isConnected;
        }
        return true;
    }
    async disconnect() {
        if (!this.isConnected && this.connectHandle) {
            // Wait for connect attempt to finish so we don't end up just re-adding the listener when it finishes.
            try {
                await this.connectHandle;
                // eslint-disable-next-line @typescript-eslint/no-unused-vars
            }
            catch (e) {
                /* Ignore errors on connect; we're disconnecting! */
            }
        }
        try {
            await this.litNodeClientInstance.disconnect();
        }
        catch (e) {
            /* We did our best */
            console.log('Failed to disconnect from LitNodeClientInstance.', e.message);
        }
        finally {
            this.isConnected = false;
        }
    }
    get litNodeClient() {
        return this.litNodeClientInstance;
    }
}
exports.LitNodeClientInstance = LitNodeClientInstance;
