// @lit-protocol/vincent-app-sdk/dist/src/webAuthClient/internal/uriHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeSearchParam = exports.uriContainsVincentJWT = exports.decodeVincentJWTFromUri = void 0;
exports.composeConnectUrl = composeConnectUrl;
const verify_1 = require("../../jwt/core/verify");
const constants_1 = require("../constants");
const decodeVincentJWTFromUri = async ({ uri, expectedAudience, requiredAppId, }) => {
    const url = new URL(uri);
    const jwt = url.searchParams.get(constants_1.JWT_URL_KEY);
    if (!jwt) {
        return null;
    }
    try {
        return {
            decodedJWT: await (0, verify_1.verifyVincentAppUserJWT)({ jwt, expectedAudience, requiredAppId }),
            jwtStr: jwt,
        };
    }
    catch (error) {
        // Explicitly throw if the JWT doesn't contain the required appId
        throw new Error(`Failed to decode JWT: ${error.message}`);
    }
};
exports.decodeVincentJWTFromUri = decodeVincentJWTFromUri;
const uriContainsVincentJWT = (uri) => {
    const url = new URL(uri);
    const connectJwt = url.searchParams.get(constants_1.JWT_URL_KEY);
    return !!connectJwt;
};
exports.uriContainsVincentJWT = uriContainsVincentJWT;
function composeConnectUrl(appId, redirectUri, connectPageUrl) {
    return new URL(`/user/appId/${String(appId)}/connect?redirectUri=${redirectUri}`, connectPageUrl || constants_1.PRODUCTION_VINCENT_DASHBOARD_URL);
}
const removeSearchParam = ({ paramName, uri, }) => {
    const url = new URL(uri);
    url.searchParams.delete(paramName);
    return url.toString();
};
exports.removeSearchParam = removeSearchParam;
