// @lit-protocol/vincent-app-sdk/dist/src/webAuthClient/app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWebAuthClient = void 0;
const constants_1 = require("./constants");
const internal_1 = require("./internal");
const { uriContainsVincentJWT, composeConnectUrl, removeSearchParam, decodeVincentJWTFromUri } = internal_1.uriHelpers;
const redirectToConnectPage = ({ appId, redirectUri, connectPageUrl, }) => window.open(composeConnectUrl(appId, redirectUri, connectPageUrl).toString(), '_self');
/** Create a new {@link WebAuthClient} instance.
 *
 * - `appId` is the numeric app ID displayed for your app on the Vincent Dashboard
 * */
const getWebAuthClient = (appClientConfig) => {
    const { appId } = appClientConfig;
    return {
        redirectToConnectPage: (redirectConnectPageConfig) => {
            const { connectPageUrl, redirectUri } = redirectConnectPageConfig;
            redirectToConnectPage({
                appId,
                connectPageUrl,
                redirectUri,
            });
        },
        uriContainsVincentJWT: () => uriContainsVincentJWT(window.location.href),
        decodeVincentJWTFromUri: (expectedAudience) => decodeVincentJWTFromUri({
            uri: window.location.href,
            expectedAudience: expectedAudience,
            requiredAppId: appId,
        }),
        removeVincentJWTFromURI: () => {
            const urlWithoutJWTSearchParam = removeSearchParam({
                paramName: constants_1.JWT_URL_KEY,
                uri: window.location.href,
            });
            window.history.replaceState({}, document.title, urlWithoutJWTSearchParam);
        },
    };
};
exports.getWebAuthClient = getWebAuthClient;
