// @lit-protocol/vincent-ability-sdk/dist/src/lib/handlers/evaluatePolicies.js
"use strict";
// src/lib/handlers/evaluatePolicies.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.evaluatePolicies = evaluatePolicies;
const zod_1 = require("zod");
const helpers_1 = require("../policyCore/helpers");
const resultCreators_1 = require("../policyCore/helpers/resultCreators");
/* eslint-disable @typescript-eslint/no-explicit-any */
async function evaluatePolicies({ vincentAbility, context, validatedPolicies, vincentAbilityApiVersion, }) {
    const evaluatedPolicies = [];
    let policyDeniedResult = undefined;
    const rawAllowedPolicies = {};
    for (const { policyPackageName, abilityPolicyParams } of validatedPolicies) {
        evaluatedPolicies.push(policyPackageName);
        const policy = vincentAbility.supportedPolicies.policyByPackageName[policyPackageName];
        try {
            const litActionResponse = await Lit.Actions.call({
                ipfsId: policy.ipfsCid,
                params: {
                    abilityParams: abilityPolicyParams,
                    context: {
                        abilityIpfsCid: context.abilityIpfsCid,
                        delegatorPkpEthAddress: context.delegation.delegatorPkpInfo.ethAddress,
                    },
                    vincentAbilityApiVersion,
                },
            });
            console.log(`evaluated ${String(policyPackageName)} policy, result is:`, JSON.stringify(litActionResponse));
            const result = parseAndValidateEvaluateResult({
                litActionResponse,
                vincentPolicy: policy.vincentPolicy,
            });
            if ((0, helpers_1.isPolicyDenyResponse)(result)) {
                policyDeniedResult = {
                    ...result,
                    packageName: policyPackageName,
                };
            }
            else {
                rawAllowedPolicies[policyPackageName] = {
                    result: result.result,
                };
            }
        }
        catch (err) {
            const denyResult = (0, helpers_1.createDenyResult)({
                runtimeError: err instanceof Error ? err.message : 'Unknown error',
            });
            policyDeniedResult = { ...denyResult, packageName: policyPackageName };
        }
    }
    if (policyDeniedResult) {
        return (0, resultCreators_1.createDenyEvaluationResult)({
            allowedPolicies: rawAllowedPolicies,
            evaluatedPolicies,
            deniedPolicy: policyDeniedResult,
        });
    }
    return (0, resultCreators_1.createAllowEvaluationResult)({
        evaluatedPolicies,
        allowedPolicies: rawAllowedPolicies,
    });
}
function parseAndValidateEvaluateResult({ litActionResponse, vincentPolicy, }) {
    let parsedLitActionResponse;
    try {
        parsedLitActionResponse = JSON.parse(litActionResponse);
    }
    catch (error) {
        console.log('rawLitActionResponse (parsePolicyExecutionResult)', litActionResponse);
        throw new Error(`Failed to JSON parse Lit Action Response: ${error instanceof Error ? error.message : String(error)}. rawLitActionResponse in request logs (parsePolicyExecutionResult)`);
    }
    try {
        console.log('parseAndValidateEvaluateResult', JSON.stringify(parsedLitActionResponse));
        if ((0, helpers_1.isPolicyDenyResponse)(parsedLitActionResponse) &&
            (parsedLitActionResponse.schemaValidationError || parsedLitActionResponse.runtimeError)) {
            console.log('parsedLitActionResponse is a deny response with a runtime error or schema validation error; skipping schema validation');
            return parsedLitActionResponse;
        }
        if (!(0, helpers_1.isPolicyResponse)(parsedLitActionResponse)) {
            throw new Error(`Invalid response from policy: ${JSON.stringify(parsedLitActionResponse)}`);
        }
        const { schemaToUse, parsedType } = (0, helpers_1.getSchemaForPolicyResponseResult)({
            value: parsedLitActionResponse,
            denyResultSchema: vincentPolicy.evalDenyResultSchema || zod_1.z.undefined(),
            allowResultSchema: vincentPolicy.evalAllowResultSchema || zod_1.z.undefined(),
        });
        console.log('parsedType', parsedType);
        const parsedResult = (0, helpers_1.validateOrDeny)(parsedLitActionResponse.result, schemaToUse, 'evaluate', 'output');
        if ((0, helpers_1.isPolicyDenyResponse)(parsedResult)) {
            // Failed schema parsing!
            return parsedResult;
        }
        if ((0, helpers_1.isPolicyDenyResponse)(parsedLitActionResponse)) {
            return (0, helpers_1.createDenyResult)({
                runtimeError: parsedLitActionResponse.runtimeError,
                schemaValidationError: parsedLitActionResponse.schemaValidationError,
                result: parsedResult,
            });
        }
        return (0, resultCreators_1.createAllowResult)({
            result: parsedResult,
        });
    }
    catch (err) {
        console.log('parseAndValidateEvaluateResult error; returning noResultDeny', err.message, err.stack);
        return (0, resultCreators_1.returnNoResultDeny)(err instanceof Error ? err.message : 'Unknown error');
    }
}
