// @lit-protocol/vincent-ability-sdk/dist/src/lib/handlers/vincentAbilityHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.vincentAbilityHandler = void 0;
exports.createAbilityExecutionContext = createAbilityExecutionContext;
const ethers_1 = require("ethers");
const helpers_1 = require("../abilityCore/helpers");
const typeGuards_1 = require("../abilityCore/helpers/typeGuards");
const validatePolicies_1 = require("../abilityCore/helpers/validatePolicies");
const zod_1 = require("../abilityCore/helpers/zod");
const assertSupportedAbilityVersion_1 = require("../assertSupportedAbilityVersion");
const getOnchainPolicyParams_1 = require("../policyCore/policyParameters/getOnchainPolicyParams");
const utils_1 = require("../utils");
const constants_1 = require("./constants");
const evaluatePolicies_1 = require("./evaluatePolicies");
/* eslint-disable @typescript-eslint/no-explicit-any */
function createAbilityExecutionContext({ vincentAbility, policyEvaluationResults, baseContext, }) {
    if (!policyEvaluationResults.allow) {
        throw new Error('Received denied policies to createAbilityExecutionContext()');
    }
    const newContext = {
        allow: true,
        evaluatedPolicies: policyEvaluationResults.evaluatedPolicies,
        allowedPolicies: {},
    };
    const policyByPackageName = vincentAbility.supportedPolicies.policyByPackageName;
    const allowedKeys = Object.keys(policyEvaluationResults.allowedPolicies);
    for (const packageName of allowedKeys) {
        const entry = policyEvaluationResults.allowedPolicies[packageName];
        const policy = policyByPackageName[packageName];
        const vincentPolicy = policy.vincentPolicy;
        if (!entry) {
            throw new Error(`Missing entry on allowedPolicies for policy: ${packageName}`);
        }
        const resultWrapper = {
            result: entry.result,
        };
        // TODO: Collect results of commit calls and add to the execution context result
        if (vincentPolicy.commit) {
            const commitFn = vincentPolicy.commit;
            resultWrapper.commit = (commitParams) => {
                return commitFn(commitParams, baseContext);
            };
        }
        newContext.allowedPolicies[packageName] =
            resultWrapper;
    }
    return newContext;
}
/** @hidden */
const vincentAbilityHandler = ({ vincentAbility, abilityParams, context, }) => {
    return async () => {
        (0, assertSupportedAbilityVersion_1.assertSupportedAbilityVersion)(vincentAbilityApiVersion);
        let policyEvalResults = undefined;
        const abilityIpfsCid = LitAuth.actionIpfsIds[0];
        const appDelegateeAddress = ethers_1.ethers.utils.getAddress(LitAuth.authSigAddress);
        // Build an initial baseContext -- we will add info as we execute, so if an error is encountered the consumer gets
        // all of the info we did find along the way
        const baseContext = {
            delegation: {
                delegateeAddress: appDelegateeAddress,
                // delegatorPkpInfo: null,
            },
            abilityIpfsCid,
            // appId: undefined,
            // appVersion: undefined,
        };
        try {
            const delegationRpcUrl = await Lit.Actions.getRpcUrl({ chain: 'yellowstone' });
            const parsedOrFail = (0, zod_1.validateOrFail)(abilityParams, vincentAbility.abilityParamsSchema, 'execute', 'input');
            if ((0, typeGuards_1.isAbilityFailureResult)(parsedOrFail)) {
                Lit.Actions.setResponse({
                    response: JSON.stringify({
                        abilityExecutionResult: parsedOrFail,
                    }),
                });
                return;
            }
            const userPkpInfo = await (0, helpers_1.getPkpInfo)({
                litPubkeyRouterAddress: constants_1.LIT_DATIL_PUBKEY_ROUTER_ADDRESS,
                yellowstoneRpcUrl: delegationRpcUrl,
                pkpEthAddress: context.delegatorPkpEthAddress,
            });
            baseContext.delegation.delegatorPkpInfo = userPkpInfo;
            const { decodedPolicies, appId, appVersion } = await (0, getOnchainPolicyParams_1.getPoliciesAndAppVersion)({
                delegationRpcUrl,
                appDelegateeAddress,
                agentWalletPkpEthAddress: context.delegatorPkpEthAddress,
                abilityIpfsCid,
            });
            baseContext.appId = appId.toNumber();
            baseContext.appVersion = appVersion.toNumber();
            const validatedPolicies = await (0, validatePolicies_1.validatePolicies)({
                decodedPolicies,
                vincentAbility,
                parsedAbilityParams: parsedOrFail,
                abilityIpfsCid,
            });
            console.log('validatedPolicies', JSON.stringify(validatedPolicies, utils_1.bigintReplacer));
            const policyEvaluationResults = await (0, evaluatePolicies_1.evaluatePolicies)({
                validatedPolicies,
                vincentAbility,
                context: baseContext,
                vincentAbilityApiVersion,
            });
            console.log('policyEvaluationResults', JSON.stringify(policyEvaluationResults, utils_1.bigintReplacer));
            policyEvalResults = policyEvaluationResults;
            if (!policyEvalResults.allow) {
                Lit.Actions.setResponse({
                    response: JSON.stringify({
                        abilityContext: {
                            ...baseContext,
                            policiesContext: policyEvaluationResults,
                        },
                        abilityExecutionResult: {
                            success: false,
                        },
                    }),
                });
                return;
            }
            const executeContext = createAbilityExecutionContext({
                vincentAbility,
                policyEvaluationResults,
                baseContext,
            });
            const abilityExecutionResult = await vincentAbility.execute({
                abilityParams: parsedOrFail,
            }, {
                ...baseContext,
                policiesContext: executeContext,
            });
            console.log('abilityExecutionResult', JSON.stringify(abilityExecutionResult, utils_1.bigintReplacer));
            Lit.Actions.setResponse({
                response: JSON.stringify({
                    abilityExecutionResult,
                    abilityContext: {
                        ...baseContext,
                        policiesContext: policyEvaluationResults,
                    },
                }),
            });
        }
        catch (err) {
            Lit.Actions.setResponse({
                response: JSON.stringify({
                    abilityContext: {
                        ...baseContext,
                        policiesContext: policyEvalResults,
                    },
                    abilityExecutionResult: {
                        success: false,
                        runtimeError: err instanceof Error ? err.message : String(err),
                    },
                }),
            });
        }
    };
};
exports.vincentAbilityHandler = vincentAbilityHandler;
