// @lit-protocol/vincent-ability-sdk/dist/src/lib/handlers/vincentPolicyHandler.js
"use strict";
// src/lib/handlers/vincentPolicyHandler.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.vincentPolicyHandler = vincentPolicyHandler;
const ethers_1 = require("ethers");
const helpers_1 = require("../abilityCore/helpers");
const assertSupportedAbilityVersion_1 = require("../assertSupportedAbilityVersion");
const helpers_2 = require("../policyCore/helpers");
const getOnchainPolicyParams_1 = require("../policyCore/policyParameters/getOnchainPolicyParams");
const utils_1 = require("../utils");
const constants_1 = require("./constants");
/** @hidden */
async function vincentPolicyHandler({ vincentPolicy, context, abilityParams, }) {
    (0, assertSupportedAbilityVersion_1.assertSupportedAbilityVersion)(vincentAbilityApiVersion);
    const { delegatorPkpEthAddress, abilityIpfsCid } = context; // FIXME: Set from ipfsCidsStack when it's shipped
    console.log('actionIpfsIds:', LitAuth.actionIpfsIds.join(','));
    const policyIpfsCid = LitAuth.actionIpfsIds[0];
    console.log('context:', JSON.stringify(context, utils_1.bigintReplacer));
    try {
        const delegationRpcUrl = await Lit.Actions.getRpcUrl({
            chain: 'yellowstone',
        });
        const userPkpInfo = await (0, helpers_1.getPkpInfo)({
            litPubkeyRouterAddress: constants_1.LIT_DATIL_PUBKEY_ROUTER_ADDRESS,
            yellowstoneRpcUrl: delegationRpcUrl,
            pkpEthAddress: delegatorPkpEthAddress,
        });
        const appDelegateeAddress = ethers_1.ethers.utils.getAddress(LitAuth.authSigAddress);
        console.log('appDelegateeAddress', appDelegateeAddress);
        const { decodedPolicies, appId, appVersion } = await (0, getOnchainPolicyParams_1.getPoliciesAndAppVersion)({
            delegationRpcUrl,
            appDelegateeAddress,
            agentWalletPkpEthAddress: delegatorPkpEthAddress,
            abilityIpfsCid,
        });
        const baseContext = {
            delegation: {
                delegateeAddress: appDelegateeAddress,
                delegatorPkpInfo: userPkpInfo,
            },
            abilityIpfsCid: abilityIpfsCid,
            appId: appId.toNumber(),
            appVersion: appVersion.toNumber(),
        };
        const onChainPolicyParams = await (0, getOnchainPolicyParams_1.getDecodedPolicyParams)({
            decodedPolicies,
            policyIpfsCid,
        });
        console.log('onChainPolicyParams:', JSON.stringify(onChainPolicyParams, utils_1.bigintReplacer));
        const evaluateResult = await vincentPolicy.evaluate({
            abilityParams,
            userParams: onChainPolicyParams,
        }, baseContext);
        console.log('evaluateResult:', JSON.stringify(evaluateResult, utils_1.bigintReplacer));
        Lit.Actions.setResponse({
            response: JSON.stringify({
                ...evaluateResult,
            }),
        });
    }
    catch (error) {
        console.log('Policy evaluation failed:', error.message, error.stack);
        Lit.Actions.setResponse({
            response: JSON.stringify((0, helpers_2.createDenyResult)({
                runtimeError: error instanceof Error ? error.message : String(error),
            })),
        });
    }
}
