// @lit-protocol/vincent-ability-sdk/dist/src/type-inference-verification/ability-definition-tests.js
"use strict";
// src/type-inference-verification/ability-definition-tests.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.testPolicyEvaluationResults = testPolicyEvaluationResults;
/**
 * Ability Definition Tests
 *
 * A focused test file to verify that VincentPolicyEvaluationResults inference works correctly,
 * using an ability with two policies - one simple and one with commit functionality.
 */
const zod_1 = require("zod");
const helpers_1 = require("../lib/abilityCore/helpers");
const vincentAbility_1 = require("../lib/abilityCore/vincentAbility");
const bundledPolicy_1 = require("../lib/policyCore/bundledPolicy/bundledPolicy");
const vincentPolicy_1 = require("../lib/policyCore/vincentPolicy");
// Base ability schema
const baseAbilitySchema = zod_1.z.object({
    action: zod_1.z.string(),
    target: zod_1.z.string(),
    amount: zod_1.z.number(),
});
/* eslint-disable @typescript-eslint/no-unused-vars */
/**
 * Simple test to verify inference of policy evaluation results
 * with a focus on the commit function type inference.
 */
function testPolicyEvaluationResults() {
    // Policy 1: Simple policy with object result schema
    const evalResultsPolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/simple-policy@1.0.0',
        abilityParamsSchema: zod_1.z.object({
            actionType: zod_1.z.string(),
            targetId: zod_1.z.string(),
        }),
        evalAllowResultSchema: zod_1.z.object({
            approved: zod_1.z.boolean(),
            reason: zod_1.z.string(),
        }),
        evaluate: async (params, { allow }) => {
            return allow({
                approved: true,
                reason: `Action ${params.abilityParams.actionType} approved`,
            });
        },
    });
    const simplePolicy = (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(evalResultsPolicy, '109i0ifj'),
        abilityParameterMappings: {
            action: 'actionType',
            target: 'targetId',
        },
    });
    // Policy 2: Policy with commit function
    const commitResultsPolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/commit-policy@1.0.0',
        abilityParamsSchema: zod_1.z.object({
            operation: zod_1.z.string(),
            resource: zod_1.z.string(),
            value: zod_1.z.number(),
        }),
        // Evaluation schema
        evalAllowResultSchema: zod_1.z.object({
            transactionId: zod_1.z.string(),
            status: zod_1.z.string(),
        }),
        // Commit phase schemas
        commitParamsSchema: zod_1.z.object({
            transactionId: zod_1.z.string(),
            status: zod_1.z.enum(['complete', 'reject']),
        }),
        commitAllowResultSchema: zod_1.z.object({
            confirmed: zod_1.z.boolean(),
            timestamp: zod_1.z.number(),
        }),
        evaluate: async (params, { allow }) => {
            const txId = `tx-${Date.now()}`;
            return allow({
                transactionId: txId,
                status: 'pending',
            });
        },
        commit: async (params, { allow }) => {
            // Access commit parameters
            const { transactionId, status } = params;
            console.log(transactionId);
            return allow({
                confirmed: status === 'complete',
                timestamp: Date.now(),
            });
        },
    });
    const commitPolicy = (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(commitResultsPolicy, 'ajialkjads'),
        abilityParameterMappings: {
            action: 'operation',
            target: 'resource',
            amount: 'value',
        },
    });
    // Define ability result schemas
    const executeSuccessSchema = zod_1.z.object({
        success: zod_1.z.boolean(),
        confirmed: zod_1.z.boolean().optional(),
        timestamp: zod_1.z.number().optional(),
        message: zod_1.z.string(),
    });
    const executeFailSchema = zod_1.z.object({
        errorCode: zod_1.z.number(),
        errorMessage: zod_1.z.string(),
    });
    const precheckSuccessSchema = zod_1.z.object({
        valid: zod_1.z.boolean(),
        validationMessage: zod_1.z.string().optional(),
    });
    const precheckFailSchema = zod_1.z.object({
        valid: zod_1.z.boolean(),
        issues: zod_1.z.array(zod_1.z.string()),
    });
    // Create ability with both policies
    return (0, vincentAbility_1.createVincentAbility)({
        packageName: '@lit-protocol/mahAbility@1.0.0',
        abilityDescription: 'Mah Ability',
        abilityParamsSchema: baseAbilitySchema,
        supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([simplePolicy, commitPolicy]),
        // Add schemas for ability results
        precheckSuccessSchema,
        precheckFailSchema,
        executeSuccessSchema,
        executeFailSchema,
        precheck: async ({ abilityParams }, { fail, succeed }) => {
            // Perform basic validation
            if (!abilityParams.action || !abilityParams.target || abilityParams.amount <= 0) {
                const issues = [];
                if (!abilityParams.action)
                    issues.push('Missing action');
                if (!abilityParams.target)
                    issues.push('Missing target');
                if (abilityParams.amount <= 0)
                    issues.push('Amount must be positive');
                return fail({
                    valid: false,
                    issues,
                });
            }
            return succeed({
                valid: true,
                validationMessage: 'All parameters are valid',
            });
        },
        // Execute function to demonstrate result type inference
        execute: async (_, { fail, succeed, policiesContext }) => {
            try {
                // Verify type inference works correctly when results.allow is true
                // Access specific policy results - now TypeScript knows this is defined
                if (policiesContext.allowedPolicies['@lit-protocol/simple-policy@1.0.0']) {
                    const simpleResult = policiesContext.allowedPolicies['@lit-protocol/simple-policy@1.0.0'].result;
                    // TypeScript should now correctly infer the result type
                    const { approved, reason } = simpleResult;
                    console.log(approved, reason);
                }
                // Type inference should work for the commit policy result
                if (policiesContext.allowedPolicies['@lit-protocol/commit-policy@1.0.0']) {
                    const commitPolicyResult = policiesContext.allowedPolicies['@lit-protocol/commit-policy@1.0.0'].result;
                    // No need for type guards anymore
                    const { transactionId, status } = commitPolicyResult;
                    console.log(status);
                    // The commit function should be available and properly typed
                    const commitFn = policiesContext.allowedPolicies['@lit-protocol/commit-policy@1.0.0'].commit;
                    // Call commit with properly typed parameters
                    const commitResult = await commitFn({
                        transactionId,
                        status: 'complete',
                    });
                    // Type inference should work for commit result
                    if (commitResult.allow) {
                        const { confirmed, timestamp } = commitResult.result;
                        return succeed({
                            success: true,
                            confirmed,
                            timestamp,
                            message: 'Operation executed successfully',
                        });
                    }
                }
                // If we reach here, something went wrong with the commit
                return fail({
                    errorCode: 400,
                    errorMessage: 'Failed to complete the transaction commit process',
                });
            }
            catch (error) {
                // Handle any unexpected errors
                return fail({
                    errorCode: 500,
                    errorMessage: 'Internal execution error',
                });
            }
        },
    });
}
