// @lit-protocol/vincent-ability-sdk/dist/src/type-inference-verification/ability-lifecycle-inference.js
"use strict";
// src/type-inference-verification/ability-lifecycle-inference.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = run;
const zod_1 = require("zod");
const helpers_1 = require("../lib/abilityCore/helpers");
const vincentAbility_1 = require("../lib/abilityCore/vincentAbility");
const bundledPolicy_1 = require("../lib/policyCore/bundledPolicy/bundledPolicy");
const resultCreators_1 = require("../lib/policyCore/helpers/resultCreators");
const vincentPolicy_1 = require("../lib/policyCore/vincentPolicy");
const abilityParamsSchema = zod_1.z.object({
    flag: zod_1.z.boolean(),
    value: zod_1.z.number(),
});
const aVincentPolicy = (0, vincentPolicy_1.createVincentPolicy)({
    packageName: 'a',
    abilityParamsSchema: zod_1.z.object({ foo: zod_1.z.string() }),
    evalAllowResultSchema: zod_1.z.object({ approved: zod_1.z.boolean() }),
    commit: async (_, ctx) => ctx.allow(),
    evaluate: async (_, ctx) => ctx.allow({ approved: true }),
});
const policyA = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema,
    bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(aVincentPolicy, 'cid-a'),
    abilityParameterMappings: {
        flag: 'foo',
    },
});
const bVincentPolicy = (0, vincentPolicy_1.createVincentPolicy)({
    packageName: 'b',
    abilityParamsSchema: zod_1.z.object({ bar: zod_1.z.number() }),
    evalAllowResultSchema: zod_1.z.string(),
    evaluate: async (_, ctx) => ctx.allow('hello'),
});
const policyB = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema,
    bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(bVincentPolicy, 'cid-b'),
    abilityParameterMappings: {
        value: 'bar',
    },
});
const ability = (0, vincentAbility_1.createVincentAbility)({
    packageName: '@lit-protocol/yesability3@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema,
    supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([policyA, policyB]),
    execute: async ({ abilityParams }, { policiesContext, succeed, fail }) => {
        // Type inference for allowedPolicies
        const a = policiesContext.allowedPolicies['a'];
        const b = policiesContext.allowedPolicies.b;
        // ✅ Expect result to be { approved: boolean }
        if (a) {
            console.log(a.result.approved);
            // ✅ Should compile - commit exists on 'a'
            if (a.commit) {
                const res = await a.commit();
                const wat = res.allow;
                console.log(wat);
                if (res.allow) {
                    // @ts-expect-error - 'success' doesn't exist because no commit result schema
                    console.log(res.result.success);
                }
            }
        }
        if (b) {
            // @ts-expect-error - 'approved' doesn't exist on string
            console.log(b.result.approved);
            // ✅ Expect result to be string
            const s = b.result;
            console.log(s);
            fail();
            // b.commit();
        }
        return succeed();
    },
    executeSuccessSchema: zod_1.z.undefined(),
    executeFailSchema: zod_1.z.undefined(),
});
async function run() {
    await ability.execute({
        abilityParams: {
            flag: true,
            value: 123,
        },
    }, {
        delegation: {
            delegatorPkpInfo: {
                tokenId: '90128301832',
                ethAddress: '0x102398103981032',
                publicKey: '0398103810938ef987ef978fe987ef',
            },
            delegateeAddress: 'y',
        },
        policiesContext: {
            evaluatedPolicies: ['a', 'b'],
            allowedPolicies: {
                a: {
                    result: { approved: true },
                    commit: async () => (0, resultCreators_1.createAllowResult)({ result: undefined }),
                },
                // @ts-expect-error - commit is inferred as `never` and omitted on purpose
                b: {
                    result: 'hello',
                },
            },
        },
    });
}
