// @lit-protocol/vincent-ability-sdk/dist/src/type-inference-verification/create-vincent-ability.js
"use strict";
// src/type-inference-verification/create-vincent-ability.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.ability = void 0;
const zod_1 = require("zod");
const helpers_1 = require("../lib/abilityCore/helpers");
const vincentAbility_1 = require("../lib/abilityCore/vincentAbility");
const bundledPolicy_1 = require("../lib/policyCore/bundledPolicy/bundledPolicy");
const vincentPolicy_1 = require("../lib/policyCore/vincentPolicy");
const abilityParams = zod_1.z.object({
    action: zod_1.z.string(),
    count: zod_1.z.number(),
});
// === Define a policy ===
const policySchema = zod_1.z.object({
    count: zod_1.z.number(),
});
const policyEvalAllow = zod_1.z.object({ ok: zod_1.z.boolean() });
const policyEvalDeny = zod_1.z.object({ reason: zod_1.z.string() });
const PolicyConfig = (0, vincentPolicy_1.createVincentPolicy)({
    packageName: 'limit-check',
    abilityParamsSchema: policySchema,
    evalAllowResultSchema: policyEvalAllow,
    evalDenyResultSchema: policyEvalDeny,
    evaluate: async ({ abilityParams }, context) => {
        return abilityParams.count < 10
            ? context.allow({ ok: true })
            : context.deny({ reason: 'Too high' });
    },
});
const bundled = (0, bundledPolicy_1.asBundledVincentPolicy)(PolicyConfig, 'QmCID123');
const policy = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema: abilityParams,
    bundledVincentPolicy: bundled,
    abilityParameterMappings: {
        count: 'count',
    },
});
exports.ability = (0, vincentAbility_1.createVincentAbility)({
    packageName: 'my-ability@1.0.0',
    abilityDescription: 'Yes Ability',
    abilityParamsSchema: abilityParams,
    supportedPolicies: (0, helpers_1.supportedPoliciesForAbility)([policy]),
    executeSuccessSchema: zod_1.z.object({ status: zod_1.z.string() }),
    executeFailSchema: zod_1.z.object({ error: zod_1.z.string() }),
    precheckSuccessSchema: zod_1.z.object({ passed: zod_1.z.boolean() }),
    precheckFailSchema: zod_1.z.object({ failCode: zod_1.z.literal('POLICY_FAIL') }),
    precheck: async ({ abilityParams }, { policiesContext, succeed, fail }) => {
        // ✅ Check context access
        if (policiesContext.allowedPolicies?.['limit-check']) {
            console.log('Policy check succeeded');
            const result = policiesContext.allowedPolicies['limit-check'].result;
            console.log(result);
        }
        //    ^? result: { ok: boolean }
        // ❌ Invalid key — should error
        // @ts-expect-error unknown policy name
        const wat = policiesContext.allowedPolicies['not-real'];
        console.log(wat);
        return policiesContext.allow ? succeed({ passed: true }) : fail({ failCode: 'POLICY_FAIL' });
    },
    execute: async ({ abilityParams }, { policiesContext, succeed, fail }) => {
        // const result = policiesContext.allowedPolicies['limit-check'].result;
        //    ^? result: { ok: boolean }
        return succeed({ status: 'executed' });
    },
});
