// @lit-protocol/vincent-ability-sdk/dist/src/type-inference-verification/create-policy-map-from-tool-policies.js
"use strict";
// src/type-inference-verification/create-policy-map-from-ability-policies.ts
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/no-unused-vars */
const zod_1 = require("zod");
const helpers_1 = require("../lib/abilityCore/helpers");
const bundledPolicy_1 = require("../lib/policyCore/bundledPolicy/bundledPolicy");
const vincentPolicy_1 = require("../lib/policyCore/vincentPolicy");
const PolicyConfig1 = (0, vincentPolicy_1.createVincentPolicy)({
    packageName: 'example-policy-1',
    abilityParamsSchema: zod_1.z.object({ x: zod_1.z.string() }),
    evalAllowResultSchema: zod_1.z.object({ pass: zod_1.z.boolean() }),
    evalDenyResultSchema: zod_1.z.object({ error: zod_1.z.string() }),
    evaluate: async ({ abilityParams }, ctx) => {
        return ctx.allow({ pass: true });
    },
});
const PolicyConfig2 = (0, vincentPolicy_1.createVincentPolicy)({
    packageName: 'example-policy-2',
    abilityParamsSchema: zod_1.z.object({ y: zod_1.z.number() }),
    evalAllowResultSchema: zod_1.z.object({ pass: zod_1.z.literal(true) }),
    evalDenyResultSchema: zod_1.z.object({ code: zod_1.z.number() }),
    evaluate: async ({ abilityParams }, ctx) => {
        return ctx.deny({ code: 403 });
    },
});
const bundled1 = (0, bundledPolicy_1.asBundledVincentPolicy)(PolicyConfig1, 'QmCID1');
const bundled2 = (0, bundledPolicy_1.asBundledVincentPolicy)(PolicyConfig2, 'QmCID2');
const policy1 = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema: zod_1.z.object({ x: zod_1.z.string() }),
    bundledVincentPolicy: bundled1,
    abilityParameterMappings: {
        x: 'x',
    },
});
const policy2 = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema: zod_1.z.object({ y: zod_1.z.number() }),
    bundledVincentPolicy: bundled2,
    abilityParameterMappings: {
        y: 'y',
    },
});
const result = (0, helpers_1.supportedPoliciesForAbility)([policy1, policy2]);
// ✅ Known package names — should succeed
const p1 = result.policyByPackageName['example-policy-1'];
const p2 = result.policyByPackageName['example-policy-2'];
console.log(p1, p2);
// ✅ Known CIDs — should succeed
const c1 = result.policyByIpfsCid['QmCID1'];
const c2 = result.policyByIpfsCid['QmCID2'];
console.log(c1, c2);
// @ts-expect-error should error: unknown package name
const failPkg = result.policyByPackageName['not-a-real-policy'];
// @ts-expect-error should error: unknown ipfsCid
const failCid = result.policyByIpfsCid['QmInvalidCID'];
