// @lit-protocol/vincent-ability-sdk/dist/src/type-inference-verification/parameter-inference-tests.js
"use strict";
// src/type-inference-verification/parameter-inference-tests.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.testBasicSchemaValidation = testBasicSchemaValidation;
exports.testNoSchemaValidation = testNoSchemaValidation;
exports.testStringSchemaValidation = testStringSchemaValidation;
exports.testDenyValidation = testDenyValidation;
exports.testContextSchemas = testContextSchemas;
/**
 * Parameter Inference Tests
 *
 * Tests to verify TypeScript correctly enforces type constraints
 * on parameters passed to allow() and deny() methods.
 */
const zod_1 = require("zod");
const bundledPolicy_1 = require("../lib/policyCore/bundledPolicy/bundledPolicy");
const vincentPolicy_1 = require("../lib/policyCore/vincentPolicy");
// Base ability schema for all tests
const baseAbilitySchema = zod_1.z.object({
    action: zod_1.z.string(),
    target: zod_1.z.string(),
    amount: zod_1.z.number(),
});
/**
 * Test 1: Basic schema validation for allow/deny
 */
function testBasicSchemaValidation() {
    const basicSchemaPolicyValid = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/test-policy@1.34.2',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        // Define schemas
        evalAllowResultSchema: zod_1.z.object({ success: zod_1.z.boolean() }),
        evalDenyResultSchema: zod_1.z.object({ errorCode: zod_1.z.number() }),
        evaluate: async (params, context) => {
            // Test TypeScript errors first
            // @ts-expect-error - No arguments when schema exists
            context.allow();
            // @ts-expect-error - Wrong type (string not matching schema)
            context.allow('not an object');
            // @ts-expect-error - Wrong shape (missing required field)
            context.allow({});
            // Valid call - matches schema
            return context.allow({ success: true });
        },
    });
    const basicSchemaPolicy = (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(basicSchemaPolicyValid, 'oi3wjfn9048w2j'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
    return basicSchemaPolicy;
}
/**
 * Test 2: No schema case validation
 */
function testNoSchemaValidation() {
    const noSchemaPolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/test-policy@1.34.2',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        // No schemas defined
        evaluate: async (params, context) => {
            // Test TypeScript errors first
            // @ts-expect-error - No schema means no args allowed
            context.allow('should error');
            // @ts-expect-error - No schema means no object allowed
            context.allow({ shouldError: true });
            // Valid calls - no args for allow with no schema
            // And string error for deny with no schema
            if (Math.random() > 0.5) {
                return context.allow();
            }
            else {
                // @ts-expect-error Can't return a string when no schema defined
                return context.deny('Error message');
            }
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(noSchemaPolicy, '1098jwsdoifosdji'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
}
/**
 * Test 3: String schema validation
 */
function testStringSchemaValidation() {
    const stringSchemaPolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/test-policy@1.34.2',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        // String schema
        evalAllowResultSchema: zod_1.z.string(),
        evaluate: async (params, context) => {
            // Test TypeScript errors first
            // @ts-expect-error - No arguments when schema exists
            context.allow();
            // @ts-expect-error - Wrong type (number not matching string schema)
            context.allow(123);
            // @ts-expect-error - Wrong type (object not matching string schema)
            context.allow({ notAString: true });
            // Valid - matches string schema
            return context.allow('Success message');
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(stringSchemaPolicy, '082u0fij0w9jf0sjdf'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
}
/**
 * Test 4: Deny parameter validation
 */
function testDenyValidation() {
    const denyValidationPolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/test-policy@1.34.2',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        // Define deny schema
        evalDenyResultSchema: zod_1.z.object({ code: zod_1.z.number(), message: zod_1.z.string() }),
        evaluate: async (params, context) => {
            // Test TypeScript errors first
            // @ts-expect-error - String only when schema exists
            context.deny('string only error');
            // @ts-expect-error - Wrong shape (missing required field)
            context.deny({ code: 400 });
            // @ts-expect-error - Wrong shape (missing required field)
            context.deny({ message: 'Error message' });
            // Valid - matches schema
            return context.deny({ code: 403, message: 'Forbidden' });
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(denyValidationPolicy, '198hf0298w3hfo9idfjh'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
}
/**
 * Test 5: Different schemas for different contexts
 */
function testContextSchemas() {
    const contextSchemasPolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/test-policy@1.34.2',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        // Different schemas for precheck and evaluate
        precheckAllowResultSchema: zod_1.z.object({ valid: zod_1.z.boolean() }),
        evalAllowResultSchema: zod_1.z.object({ approved: zod_1.z.boolean() }),
        precheck: async (params, context) => {
            // TypeScript error test
            // @ts-expect-error - Using eval schema in precheck
            context.allow({ approved: true });
            // Valid for precheck
            return context.allow({ valid: true });
        },
        evaluate: async (params, context) => {
            // TypeScript error test
            // @ts-expect-error - Using precheck schema in eval
            context.allow({ valid: true });
            // Valid for evaluate
            return context.allow({ approved: true });
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(contextSchemasPolicy, 'swmfd08238rhjdskjfn'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
}
