// @lit-protocol/vincent-ability-sdk/dist/src/type-inference-verification/schema-test.js
"use strict";
// src/type-inference-verification/schema-test.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.testRealPolicy = void 0;
exports.testContextSignature = testContextSignature;
exports.testWithoutSchema = testWithoutSchema;
exports.testStringSchema = testStringSchema;
exports.testDifferentContexts = testDifferentContexts;
exports.testPrimitiveSchemas = testPrimitiveSchemas;
const zod_1 = require("zod");
const bundledPolicy_1 = require("../lib/policyCore/bundledPolicy/bundledPolicy");
const vincentPolicy_1 = require("../lib/policyCore/vincentPolicy");
// Base ability schema for all tests
const baseAbilitySchema = zod_1.z.object({
    action: zod_1.z.string(),
    target: zod_1.z.string(),
    amount: zod_1.z.number(),
});
/**
 * Test 1: With Schema
 *
 * This test verifies that TypeScript properly enforces schema constraints
 * when schemas are provided.
 */
const objSchema = zod_1.z.object({ id: zod_1.z.string() });
// Function signature to test types on a PolicyContext
function testContextSignature(context) {
    // @ts-expect-error Should error - no args with schema
    context.allow();
    // @ts-expect-error Should error - wrong type
    context.allow('not an object');
    // @ts-expect-error Should error - wrong shape
    context.allow({ notId: 'wrong property' });
    // Valid - matches schema
    return context.allow({ id: 'valid-id' });
}
// Test in a real policy
const realPolicy = (0, vincentPolicy_1.createVincentPolicy)({
    packageName: '@lit-protocol/schema-test-policy@1.0.0',
    abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
    evalAllowResultSchema: objSchema,
    evaluate: async (params, { allow }) => {
        // These cause TypeScript errors:
        // @ts-expect-error - No args when schema exists
        allow();
        // @ts-expect-error - Wrong type
        allow(123);
        // @ts-expect-error - Wrong shape
        allow({ wrong: 'field' });
        // Valid case:
        return allow({ id: 'test-id' });
    },
});
exports.testRealPolicy = (0, vincentPolicy_1.createVincentAbilityPolicy)({
    abilityParamsSchema: baseAbilitySchema,
    bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(realPolicy, '1091jkfn2039j802jf'),
    abilityParameterMappings: {
        action: 'actionType',
    },
});
/**
 * Test 2: Without Schema
 *
 * This test verifies that TypeScript properly enforces constraints
 * when no schemas are provided (no args for allow, string or no args for deny).
 */
function testWithoutSchema() {
    // Function signature to test types on a PolicyContext
    function testContextSignature(context) {
        // @ts-expect-error Should error - no schema means no args allowed
        context.allow('no schema');
        // @ts-expect-error Should error - no schema means no args allowed
        context.allow({ no: 'schema' });
        // @ts-expect-error Can't return a string when no schema defined
        context.deny('Error message');
        // Valid - no args is allowed
        return context.deny();
    }
    testContextSignature.touch = true; // Avoid 'function is never referenced ts error breaking our inference test
    // Test in a real policy
    const thePolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/schema-test-policy@1.0.0',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        // No schemas defined
        evaluate: async (params, { allow, deny }) => {
            // These cause TypeScript errors:
            // @ts-expect-error - Cannot accept arguments when no schema
            allow('not allowed');
            // @ts-expect-error - Cannot accept arguments when no schema
            allow(123);
            // @ts-expect-error - Cannot accept object when no schema
            allow({ not: 'allowed' });
            // @ts-expect-error - Object not allowed for deny without schema
            deny({ not: 'allowed' });
            // Valid cases - only one will be returned
            if (Math.random() > 0.5) {
                return allow();
            }
            else {
                // @ts-expect-error Can't return a string when no schema defined
                return deny('Error message');
            }
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(thePolicy, '1091jkfn2039j802jf'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
}
/**
 * Test 3: String Schema
 *
 * This test verifies handling of string schemas, which are a special case.
 */
function testStringSchema() {
    // Test in a real policy
    const stringSchemaPolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/schema-test-policy@1.0.0',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        evalAllowResultSchema: zod_1.z.string(),
        evalDenyResultSchema: zod_1.z.string(),
        evaluate: async (params, { allow, deny }) => {
            // These cause TypeScript errors:
            // @ts-expect-error - No args when schema exists
            allow();
            // @ts-expect-error - Wrong type (number not string)
            allow(123);
            // @ts-expect-error - Wrong type (object not string)
            allow({ not: 'a string' });
            // @ts-expect-error - No args when schema exists
            deny();
            // @ts-expect-error - Wrong type (object not string)
            deny({ not: 'a string' });
            // Valid cases - only one will be returned
            if (Math.random() > 0.5) {
                return allow('Success message');
            }
            else {
                return deny('Error message');
            }
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(stringSchemaPolicy, '1091jkfn2039j802jf'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
}
/**
 * Test 4: Different Contexts
 *
 * This test verifies that different schemas are enforced correctly
 * in different contexts (precheck, evaluate, commit).
 */
function testDifferentContexts() {
    // Different schemas for different contexts
    const precheckSchema = zod_1.z.object({ validCheck: zod_1.z.boolean() });
    const evalSchema = zod_1.z.object({ status: zod_1.z.string() });
    const commitSchema = zod_1.z.object({ txId: zod_1.z.string() });
    const differentContextsPolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/schema-test-policy@1.0.0',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        // Define different schemas for different contexts
        precheckAllowResultSchema: precheckSchema,
        evalAllowResultSchema: evalSchema,
        commitParamsSchema: zod_1.z.object({ confirmation: zod_1.z.string() }),
        commitAllowResultSchema: commitSchema,
        precheck: async (params, { allow }) => {
            // These cause TypeScript errors:
            // @ts-expect-error - Wrong schema (using evaluate schema)
            allow({ status: 'wrong context' });
            // @ts-expect-error - Wrong schema (using commit schema)
            allow({ txId: 'wrong context' });
            // Valid for precheck
            return allow({ validCheck: true });
        },
        evaluate: async (params, { allow }) => {
            // These cause TypeScript errors:
            // @ts-expect-error - Wrong schema (using precheck schema)
            allow({ validCheck: true });
            // @ts-expect-error - Wrong schema (using commit schema)
            allow({ txId: 'wrong context' });
            // Valid for evaluate
            return allow({ status: 'approved' });
        },
        commit: async (params, { allow }) => {
            // Should have the right parameter type
            const confirmation = params.confirmation;
            // These cause TypeScript errors:
            // @ts-expect-error - Wrong schema (using precheck schema)
            allow({ validCheck: true });
            // @ts-expect-error - Wrong schema (using evaluate schema)
            allow({ status: 'wrong context' });
            // Valid for commit
            return allow({ txId: `tx-${confirmation}` });
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(differentContextsPolicy, '1091jkfn2039j802jf'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
}
/**
 * Test 5: Number and Boolean Schemas
 *
 * This test verifies that number and boolean schemas work correctly.
 */
function testPrimitiveSchemas() {
    // Test in a real policy
    const primitivePolicy = (0, vincentPolicy_1.createVincentPolicy)({
        packageName: '@lit-protocol/schema-test-policy@1.0.0',
        abilityParamsSchema: zod_1.z.object({ actionType: zod_1.z.string() }),
        evalAllowResultSchema: zod_1.z.number(),
        evalDenyResultSchema: zod_1.z.boolean(),
        evaluate: async (params, { allow, deny }) => {
            // These cause TypeScript errors:
            // @ts-expect-error - No args when schema exists
            allow();
            // @ts-expect-error - Wrong type (string not number)
            allow('not a number');
            // @ts-expect-error - Wrong type (object not number)
            allow({ value: 123 });
            // @ts-expect-error - No args when schema exists (except string error)
            deny();
            // @ts-expect-error - Wrong type (number not boolean)
            deny(123);
            // Valid cases - only one will be executed
            if (Math.random() > 0.5) {
                return allow(123);
            }
            else {
                return deny(false, 'Operation failed');
            }
        },
    });
    return (0, vincentPolicy_1.createVincentAbilityPolicy)({
        abilityParamsSchema: baseAbilitySchema,
        bundledVincentPolicy: (0, bundledPolicy_1.asBundledVincentPolicy)(primitivePolicy, '1091jkfn2039j802jf'),
        abilityParameterMappings: {
            action: 'actionType',
        },
    });
}
