// @lit-protocol/vincent-app-sdk/dist/src/abilityClient/index.js
"use strict";
/** The VincentAbilityClient is used to handle all interactions with VincentAbilities -- both local execution of its precheck()
 * method and the execution of its `execute()` method via the LIT Action runtime.  The VincentAbilityClient also handles calling policy
 * precheck methods when they are defined, and returns all policy precheck results along with the ability precheck result.
 *
 * @packageDocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateVincentAbilitySessionSigs = exports.getVincentAbilityClient = exports.disconnectVincentAbilityClients = exports.isAbilityResponseFailure = exports.isAbilityResponseSchemaValidationFailure = exports.isAbilityResponseRuntimeFailure = exports.isAbilityResponseSuccess = void 0;
// src/lib/abilityClient/index.ts
var typeGuards_1 = require("./typeGuards");
Object.defineProperty(exports, "isAbilityResponseSuccess", { enumerable: true, get: function () { return typeGuards_1.isAbilityResponseSuccess; } });
Object.defineProperty(exports, "isAbilityResponseRuntimeFailure", { enumerable: true, get: function () { return typeGuards_1.isAbilityResponseRuntimeFailure; } });
Object.defineProperty(exports, "isAbilityResponseSchemaValidationFailure", { enumerable: true, get: function () { return typeGuards_1.isAbilityResponseSchemaValidationFailure; } });
Object.defineProperty(exports, "isAbilityResponseFailure", { enumerable: true, get: function () { return typeGuards_1.isAbilityResponseFailure; } });
const getLitNodeClient_1 = require("../internal/LitNodeClient/getLitNodeClient");
/** This method closes any registered event listeners maintained by Vincent Ability Clients, allowing your process to exit gracefully.
 * @category API
 */
const disconnectVincentAbilityClients = getLitNodeClient_1.disconnectLitNodeClientInstance;
exports.disconnectVincentAbilityClients = disconnectVincentAbilityClients;
var vincentAbilityClient_1 = require("./vincentAbilityClient");
Object.defineProperty(exports, "getVincentAbilityClient", { enumerable: true, get: function () { return vincentAbilityClient_1.getVincentAbilityClient; } });
var generateVincentAbilitySessionSigs_1 = require("./execute/generateVincentAbilitySessionSigs");
Object.defineProperty(exports, "generateVincentAbilitySessionSigs", { enumerable: true, get: function () { return generateVincentAbilitySessionSigs_1.generateVincentAbilitySessionSigs; } });
