// @lit-protocol/vincent-app-sdk/dist/src/abilityClient/typeGuards.js
"use strict";
// src/lib/abilityClient/typeGuards.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAbilityResponseSuccess = isAbilityResponseSuccess;
exports.isAbilityResponseFailure = isAbilityResponseFailure;
exports.isAbilityResponseRuntimeFailure = isAbilityResponseRuntimeFailure;
exports.isAbilityResponseSchemaValidationFailure = isAbilityResponseSchemaValidationFailure;
exports.isAbilityResponse = isAbilityResponse;
exports.isRemoteVincentAbilityExecutionResult = isRemoteVincentAbilityExecutionResult;
/** Type guard function that returns true only if the passed value is a successful ability response
 * @category API
 * */
function isAbilityResponseSuccess(value) {
    return (typeof value === 'object' &&
        value !== null &&
        'success' in value &&
        value.success === true);
}
/** Type guard function that returns true only if the passed value is a failure ability response
 *  @category API */
function isAbilityResponseFailure(value) {
    return (typeof value === 'object' &&
        value !== null &&
        'success' in value &&
        value.success === false);
}
/** Type guard function that returns true only if the passed value is a failure ability response that was a runtime error
 * This could be caused by code `throw()`ing an error from inside ability or policy code, or by errors encountered in the
 * Vincent SDK wrapping code that handles the execution of the ability/policies.
 *
 * This could also be the result of a schema validation error -- use {@link isAbilityResponseSchemaValidationFailure} to check
 * to see if have a data validation error.
 *
 *  @category API*/
function isAbilityResponseRuntimeFailure(value) {
    if (!isAbilityResponseFailure(value))
        return false;
    return 'runtimeError' in value && typeof value.runtimeError === 'string';
}
/** Type guard function that returns true only if the passed value is a failure ability response that was caused by
 * a schema parsing failure.  All data inputs and outputs to / from Vincent Abilities and Policies is validated using ZOD schemas.
 *
 * If this returns true, you can parse the `schemaValidationError` property in the response - it is a ZodError object.
 * @category API */
function isAbilityResponseSchemaValidationFailure(value) {
    if (!isAbilityResponseFailure(value))
        return false;
    return ('schemaValidationError' in value && typeof value.schemaValidationError === 'object');
}
/** @hidden */
function isAbilityResponse(value) {
    return (typeof value === 'object' &&
        value !== null &&
        'success' in value &&
        typeof value.success === 'boolean');
}
/** @hidden */
function isRemoteVincentAbilityExecutionResult(value) {
    return (typeof value === 'object' &&
        value !== null &&
        'abilityExecutionResult' in value &&
        'abilityContext' in value &&
        isAbilityResponse(value.abilityExecutionResult) &&
        typeof value.abilityContext === 'object' &&
        value.abilityContext !== null);
}
