// @lit-protocol/vincent-app-sdk/dist/src/abilityClient/vincentAbilityClient.js
"use strict";
// src/abilityClient/vincentAbilityClient.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVincentAbilityClient = getVincentAbilityClient;
const ethers_1 = require("ethers");
const zod_1 = require("zod");
const constants_1 = require("@lit-protocol/constants");
const internal_1 = require("@lit-protocol/vincent-ability-sdk/internal");
const getLitNodeClient_1 = require("../internal/LitNodeClient/getLitNodeClient");
const generateVincentAbilitySessionSigs_1 = require("./execute/generateVincentAbilitySessionSigs");
const resultCreators_1 = require("./execute/resultCreators");
const resultCreators_2 = require("./precheck/resultCreators");
const runPolicyPrechecks_1 = require("./precheck/runPolicyPrechecks");
const typeGuards_1 = require("./typeGuards");
const YELLOWSTONE_RPC_URL = 'https://yellowstone-rpc.litprotocol.com/';
const BASE_RPC_URL = 'https://base.llamarpc.com';
const bigintReplacer = (key, value) => {
    return typeof value === 'bigint' ? value.toString() : value;
};
/** A VincentAbilityClient provides a type-safe interface for executing abilities, for both `precheck()`
 * and `execute()` functionality.
 *
 * ```typescript
 * import { disconnectVincentAbilityClients, getVincentAbilityClient, isAbilityResponseFailure } from '@lit-protocol/vincent-app-sdk/abilityClient';
 * import { bundledVincentAbility as uniswapBundledAbility } from '@lit-protocol/vincent-ability-uniswap-swap';
 * import { delegateeEthersSigner } = from './ethersSigner';
 * import { ETH_RPC_URL, BASE_RPC_URL } from './rpcConfigs';
 *
 * const uniswapAbilityClient = getVincentAbilityClient({
 *     bundledVincentAbility: uniswapBundledAbility,
 *     ethersSigner: delegateeEthersSigner,
 *   });
 *
 * // First, call `precheck()` to get a best-estimate result indicating that the ability execution in the LIT action runtime will not fail
 * const precheckResult = await uniswapSwapAbilityClient.precheck({
 *     ethRpcUrl: ETH_RPC_URL,
 *     rpcUrlForUniswap: BASE_RPC_URL,
 *     chainIdForUniswap: 8453, // Base
 *     tokenInAddress: '0x4200000000000000000000000000000000000006', // WETH
 *     tokenInDecimals: 18,
 *     tokenInAmount: 0.0000077,
 *     tokenOutAddress: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913', // USDC on Base
 *     tokenOutDecimals: 8,
 *   },
 *   {
 *     delegatorPkpEthAddress: '0x123456789123456789123456789...',
 *     agentAddress: '0xabc123...',
 *   });
 *
 * const uniswapSwapExecutionResult = await uniswapSwapAbilityClient.execute({
 *   ethRpcUrl: ETH_RPC_URL,
 *   rpcUrlForUniswap: BASE_RPC_URL,
 *   chainIdForUniswap: 8453,
 *   tokenInAddress: '0x4200000000000000000000000000000000000006', // WETH
 *   tokenInDecimals: 18,
 *   tokenInAmount: 0.0000077,
 *   tokenOutAddress: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913', // USDC on Base
 *   tokenOutDecimals: 8,
 * },
 * {
 *   delegatorPkpEthAddress: '0x123456789123456789123456789...',
 *   agentAddress: '0xabc123...',
 * });
 *
 * if(isAbilityResponseFailure(uniswapSwapExecutionResult)) {
 *   ...handle failure
 * } else {
 *  ...handle result
 * }
 * ```
 *
 * @typeParam IpfsCid {@removeTypeParameterCompletely}
 * @typeParam AbilityParamsSchema {@removeTypeParameterCompletely}
 * @typeParam PkgNames {@removeTypeParameterCompletely}
 * @typeParam PolicyMap {@removeTypeParameterCompletely}
 * @typeParam PoliciesByPackageName {@removeTypeParameterCompletely}
 * @typeParam ExecuteSuccessSchema {@removeTypeParameterCompletely}
 * @typeParam ExecuteFailSchema {@removeTypeParameterCompletely}
 * @typeParam PrecheckSuccessSchema {@removeTypeParameterCompletely}
 * @typeParam PrecheckFailSchema {@removeTypeParameterCompletely}
 *
 * @param params
 * @param params.ethersSigner  - An ethers signer that has been configured with your delegatee key
 * @param params.bundledVincentAbility  - The bundled vincent ability that you want to interact with
 * @param params.debug - Boolean to turn debug logging of the Lit Node Client on and off.
 *
 * @category API
 * */
function getVincentAbilityClient(params) {
    var _a, _b;
    const { bundledVincentAbility, ethersSigner, debug = false, registryRpcUrl: defaultRegistryRpcUrl, pkpInfoRpcUrl: defaultPkpInfoRpcUrl, } = params;
    const { ipfsCid, vincentAbility, vincentAbilityApiVersion } = bundledVincentAbility;
    (0, internal_1.assertSupportedAbilityVersion)(vincentAbilityApiVersion);
    const network = constants_1.LIT_NETWORK.Datil;
    const executeSuccessSchema = ((_a = vincentAbility.__schemaTypes.executeSuccessSchema) !== null && _a !== void 0 ? _a : zod_1.z.undefined());
    const executeFailSchema = ((_b = vincentAbility.__schemaTypes.executeFailSchema) !== null && _b !== void 0 ? _b : zod_1.z.undefined());
    return {
        async precheck(rawAbilityParams, { delegatorPkpEthAddress, agentAddress, registryRpcUrl, pkpInfoRpcUrl, rpcUrl, }) {
            var _a, _b, _c;
            if (!agentAddress) {
                throw new Error('AbilityClientContext.agentAddress is required for Vincent 2.0 flows.');
            }
            const resolvedRegistryRpcUrl = (_b = (_a = registryRpcUrl !== null && registryRpcUrl !== void 0 ? registryRpcUrl : rpcUrl) !== null && _a !== void 0 ? _a : defaultRegistryRpcUrl) !== null && _b !== void 0 ? _b : BASE_RPC_URL;
            const resolvedPkpInfoRpcUrl = (_c = pkpInfoRpcUrl !== null && pkpInfoRpcUrl !== void 0 ? pkpInfoRpcUrl : defaultPkpInfoRpcUrl) !== null && _c !== void 0 ? _c : YELLOWSTONE_RPC_URL;
            console.log('precheck', {
                rawAbilityParams,
                delegatorPkpEthAddress,
                agentAddress,
                resolvedRegistryRpcUrl,
                resolvedPkpInfoRpcUrl,
            });
            const delegateePkpEthAddress = ethers_1.ethers.utils.getAddress(await ethersSigner.getAddress());
            // This will be populated further during execution; if an error is encountered, it'll include as much data as we can give the caller.
            const baseContext = {
                delegation: {
                    delegateeAddress: delegateePkpEthAddress,
                    // delegatorPkpInfo: null,
                },
                abilityIpfsCid: ipfsCid,
                agentAddress,
                // appId: undefined,
                // appVersion: undefined,
            };
            const parsedParams = (0, internal_1.validateOrFail)(rawAbilityParams, vincentAbility.abilityParamsSchema, 'precheck', 'input');
            if ((0, typeGuards_1.isAbilityResponseFailure)(parsedParams)) {
                return (0, resultCreators_2.createAbilityPrecheckResponseFailureNoResult)({
                    ...parsedParams,
                    context: baseContext,
                });
            }
            const userPkpInfo = await (0, internal_1.getPkpInfo)({
                litPubkeyRouterAddress: internal_1.LIT_DATIL_PUBKEY_ROUTER_ADDRESS,
                yellowstoneRpcUrl: resolvedPkpInfoRpcUrl,
                pkpEthAddress: delegatorPkpEthAddress,
            });
            baseContext.delegation.delegatorPkpInfo = userPkpInfo;
            console.log('userPkpInfo', userPkpInfo);
            const { decodedPolicies, appId, appVersion } = await (0, internal_1.getPoliciesAndAppVersion)({
                registryRpcUrl: resolvedRegistryRpcUrl,
                appDelegateeAddress: delegateePkpEthAddress,
                agentAddress,
                abilityIpfsCid: ipfsCid,
            });
            baseContext.appId = appId.toNumber();
            baseContext.appVersion = appVersion.toNumber();
            console.log('Fetched policies and app info', { decodedPolicies, appId, appVersion });
            const baseAbilityContext = await (0, runPolicyPrechecks_1.runAbilityPolicyPrechecks)({
                bundledVincentAbility,
                abilityParams: parsedParams,
                decodedPolicies,
                context: {
                    ...baseContext,
                    registryRpcUrl: resolvedRegistryRpcUrl,
                },
            });
            if (!vincentAbility.precheck) {
                console.log('No ability precheck defined - returning baseContext policy evaluation results', {
                    rawAbilityParams,
                    delegatorPkpEthAddress,
                    resolvedRegistryRpcUrl,
                });
                if (!baseAbilityContext.policiesContext.allow) {
                    return (0, resultCreators_2.createAbilityPrecheckResponseFailureNoResult)({
                        context: baseAbilityContext,
                    });
                }
                return (0, resultCreators_2.createAbilityPrecheckResponseSuccessNoResult)({
                    context: baseAbilityContext,
                });
            }
            console.log('Executing ability precheck');
            const precheckResult = await vincentAbility.precheck({ abilityParams: parsedParams }, baseAbilityContext);
            if ((0, typeGuards_1.isAbilityResponseSchemaValidationFailure)(precheckResult) ||
                (0, typeGuards_1.isAbilityResponseRuntimeFailure)(precheckResult)) {
                console.log('Detected runtime or schema validation error in abilityPrecheckResult - returning as-is:', JSON.stringify({
                    isAbilityResponseRuntimeFailure: (0, typeGuards_1.isAbilityResponseRuntimeFailure)(precheckResult),
                    isAbilityResponseSchemaValidationFailure: (0, typeGuards_1.isAbilityResponseSchemaValidationFailure)(precheckResult),
                    precheckResult,
                }, bigintReplacer));
                // Runtime errors and schema validation errors will not have results; return them as-is.
                return precheckResult;
            }
            console.log('precheckResult()', JSON.stringify(precheckResult, bigintReplacer));
            return {
                ...precheckResult,
                context: baseAbilityContext,
            };
        },
        async execute(rawAbilityParams, context) {
            if (!context.agentAddress) {
                throw new Error('AbilityClientContext.agentAddress is required for Vincent 2.0 flows.');
            }
            const parsedParams = (0, internal_1.validateOrFail)(rawAbilityParams, vincentAbility.abilityParamsSchema, 'execute', 'input');
            const resolvedContext = context;
            if ((0, typeGuards_1.isAbilityResponseFailure)(parsedParams)) {
                return {
                    ...parsedParams,
                    context: resolvedContext,
                };
            }
            const litNodeClient = await (0, getLitNodeClient_1.getLitNodeClientInstance)({ network, debug });
            const sessionSigs = await (0, generateVincentAbilitySessionSigs_1.generateVincentAbilitySessionSigs)({ ethersSigner, litNodeClient });
            const result = await litNodeClient.executeJs({
                ipfsId: ipfsCid,
                sessionSigs,
                jsParams: {
                    abilityParams: parsedParams,
                    context: resolvedContext,
                    vincentAbilityApiVersion,
                },
            });
            const { success, response } = result;
            console.log('executeResult - raw result from `litNodeClient.executeJs()', {
                response,
                success,
            });
            if (success !== true) {
                return (0, resultCreators_1.createAbilityExecuteResponseFailureNoResult)({
                    runtimeError: `Remote ability failed with unknown error: ${JSON.stringify(response, bigintReplacer, 2)}`,
                });
            }
            let parsedResult = response;
            if (typeof response === 'string') {
                // lit-node-client returns a string if no signed data, even if the result could be JSON.parse'd :(
                try {
                    parsedResult = JSON.parse(response);
                    // eslint-disable-next-line @typescript-eslint/no-unused-vars
                }
                catch (e) {
                    return (0, resultCreators_1.createAbilityExecuteResponseFailureNoResult)({
                        runtimeError: `Remote ability failed with unknown error: ${JSON.stringify(response, bigintReplacer)}`,
                    });
                }
            }
            if (!(0, typeGuards_1.isRemoteVincentAbilityExecutionResult)(parsedResult)) {
                console.log('Result from `executeJs` was valid JSON, but not a vincentAbilityExecutionResult', { parsedResult, success });
                return (0, resultCreators_1.createAbilityExecuteResponseFailureNoResult)({
                    runtimeError: `Remote ability failed with unknown error: ${JSON.stringify(parsedResult, bigintReplacer)}`,
                });
            }
            const remoteVincentAbilityResult = parsedResult;
            console.log('Parsed executeJs vincentAbilityExecutionResult:', JSON.stringify(parsedResult, bigintReplacer));
            const { abilityContext, abilityExecutionResult } = remoteVincentAbilityResult;
            if ((0, typeGuards_1.isAbilityResponseSchemaValidationFailure)(abilityExecutionResult) ||
                (0, typeGuards_1.isAbilityResponseRuntimeFailure)(abilityExecutionResult)) {
                console.log('Detected runtime or schema validation error in abilityExecutionResult - returning as-is:', JSON.stringify({
                    isAbilityResponseRuntimeFailure: (0, typeGuards_1.isAbilityResponseRuntimeFailure)(abilityExecutionResult),
                    isAbilityResponseSchemaValidationFailure: (0, typeGuards_1.isAbilityResponseSchemaValidationFailure)(abilityExecutionResult),
                    abilityExecutionResult,
                }, bigintReplacer));
                // Runtime errors and schema validation errors will not have results; return them as-is.
                return (0, resultCreators_1.createAbilityExecuteResponseFailureNoResult)({
                    ...(abilityExecutionResult.runtimeError
                        ? { runtimeError: abilityExecutionResult.runtimeError }
                        : {}),
                    ...(abilityExecutionResult.schemaValidationError
                        ? { schemaValidationError: abilityExecutionResult.schemaValidationError }
                        : {}),
                    context: remoteVincentAbilityResult.abilityContext,
                });
            }
            // Policy eval happens before `execute()` is ever called
            // As a result, when policies return a `deny` result, there will be no ability result
            // so we need to skip trying to run result through the success/fail schema logic
            if (!abilityContext.policiesContext.allow) {
                return (0, resultCreators_1.createAbilityExecuteResponseFailureNoResult)({
                    context: abilityContext,
                });
            }
            const resultSchemaDetails = (0, internal_1.getSchemaForAbilityResult)({
                value: abilityExecutionResult,
                successResultSchema: executeSuccessSchema,
                failureResultSchema: executeFailSchema,
            });
            const { schemaToUse, parsedType } = resultSchemaDetails;
            console.log(`Parsing ability result using the ${parsedType} Zod schema`);
            // Parse returned result using appropriate execute zod schema
            const executeResult = (0, internal_1.validateOrFail)(abilityExecutionResult.result, schemaToUse, 'execute', 'output');
            console.log('Zod parse result:', executeResult);
            if ((0, typeGuards_1.isAbilityResponseFailure)(executeResult)) {
                // Parsing the result threw a zodError
                return executeResult;
            }
            console.log('Raw abilityExecutionResult was:', abilityExecutionResult);
            // We parsed the result -- it may be a success or a failure; return appropriately.
            if ((0, typeGuards_1.isAbilityResponseFailure)(abilityExecutionResult)) {
                return (0, resultCreators_1.createAbilityExecuteResponseFailure)({
                    ...(abilityExecutionResult.runtimeError
                        ? { runtimeError: abilityExecutionResult.runtimeError }
                        : {}),
                    ...(abilityExecutionResult.schemaValidationError
                        ? { schemaValidationError: abilityExecutionResult.schemaValidationError }
                        : {}),
                    result: executeResult,
                    context: remoteVincentAbilityResult.abilityContext,
                });
            }
            const res = executeResult;
            return (0, resultCreators_1.createAbilityExecuteResponseSuccess)({
                result: res,
                context: remoteVincentAbilityResult.abilityContext,
            });
        },
    };
}
