// @lit-protocol/vincent-contracts-sdk/dist/src/internal/wrapped-keys/getVincentWrappedKeysAccs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVincentWrappedKeysAccs = getVincentWrappedKeysAccs;
const ethers_1 = require("ethers");
const constants_1 = require("@lit-protocol/constants");
const contracts_sdk_1 = require("@lit-protocol/contracts-sdk");
const constants_2 = require("../../constants");
const pkpInfo_1 = require("../../utils/pkpInfo");
const CHAIN_YELLOWSTONE = 'yellowstone';
/**
 * Creates access control condition to validate Vincent delegatee and platform user authorization
 * via the Vincent registry contract's isDelegateePermitted method and the PKP NFT contract's ownerOf method
 *
 * Creates access control conditions for both the delegatee and the platform user.
 * Delegatee authorization is validated by the Vincent registry contract's isDelegateePermitted method.
 * Platform user authorization is validated by validating the owner of the delegator PKP token is the requester by checking
 * the Lit PKP NFT contract's ownerOf method.
 *
 * @param delegatorPkpEthAddress - The user PKP address used for ownership checks (not the agent address)
 * @param agentAddress - The agent smart account address used for permission checks
 *
 * @returns EvmContractConditions - Access control conditions authorizing a valid delegatee OR a platform user that is the owner of the delegator's PKP token
 */
async function getVincentWrappedKeysAccs({ delegatorPkpEthAddress, agentAddress, }) {
    if (!ethers_1.ethers.utils.isAddress(delegatorPkpEthAddress)) {
        throw new Error(`delegatorPkpEthAddress is not a valid Ethereum Address: ${delegatorPkpEthAddress}`);
    }
    if (!ethers_1.ethers.utils.isAddress(agentAddress)) {
        throw new Error(`agentAddress is not a valid Ethereum Address: ${agentAddress}`);
    }
    const delegatorPkpTokenId = (await (0, pkpInfo_1.getPkpTokenId)({
        pkpEthAddress: delegatorPkpEthAddress,
        signer: ethers_1.ethers.Wallet.createRandom().connect(new ethers_1.ethers.providers.StaticJsonRpcProvider(constants_1.LIT_RPC.CHRONICLE_YELLOWSTONE)), // Read only; signer identity is irrelevant in this code path :)
    })).toString();
    return Promise.all([
        getDelegateeAccessControlConditions({
            agentAddress,
        }),
        Promise.resolve({ operator: 'or' }),
        getPlatformUserAccessControlConditions({
            delegatorPkpTokenId,
        }),
    ]);
}
async function getDelegateeAccessControlConditions({ agentAddress, }) {
    const contractInterface = new ethers_1.ethers.utils.Interface(constants_2.COMBINED_ABI.fragments);
    const fragment = contractInterface.getFunction('validateAbilityExecutionAndGetPolicies');
    const functionAbi = {
        type: 'function',
        name: fragment.name,
        inputs: fragment.inputs.map((input) => ({
            name: input.name,
            type: input.type,
        })),
        outputs: fragment.outputs?.map((output) => ({
            name: output.name,
            type: output.type,
        })),
        stateMutability: fragment.stateMutability,
    };
    return {
        contractAddress: constants_2.VINCENT_DIAMOND_CONTRACT_ADDRESS_PROD,
        chain: CHAIN_YELLOWSTONE,
        functionAbi,
        functionName: 'validateAbilityExecutionAndGetPolicies',
        functionParams: [':userAddress', agentAddress, ':currentActionIpfsId'],
        returnValueTest: {
            key: 'isPermitted',
            comparator: '=',
            value: 'true',
        },
    };
}
/**
 * Delegated wrapped keys access control condition is 'the owner of the PKP may decrypt this'
 * - e.g. the Vincent 'user PKP' can decrypt the wrapped keys for all of their agent PKPs
 *
 * This contrasts from the original wrapped keys access control condition, which was 'the PKP itself may decrypt this'
 *
 * @param delegatorPkpTokenId
 */
async function getPlatformUserAccessControlConditions({ delegatorPkpTokenId, }) {
    const contractAddresses = await contracts_sdk_1.LitContracts.getContractAddresses(constants_1.LIT_NETWORK.Datil, new ethers_1.ethers.providers.StaticJsonRpcProvider(constants_1.LIT_RPC.CHRONICLE_YELLOWSTONE));
    const pkpNftContractInfo = contractAddresses.PKPNFT;
    if (!pkpNftContractInfo?.address) {
        throw new Error('PKP NFT contract address not found for Datil network');
    }
    return {
        contractAddress: pkpNftContractInfo.address,
        chain: CHAIN_YELLOWSTONE,
        functionAbi: {
            type: 'function',
            name: 'ownerOf',
            inputs: [
                {
                    name: 'tokenId',
                    type: 'uint256',
                },
            ],
            outputs: [
                {
                    name: '',
                    type: 'address',
                },
            ],
            stateMutability: 'view',
        },
        functionName: 'ownerOf',
        functionParams: [delegatorPkpTokenId],
        returnValueTest: {
            key: '',
            comparator: '=',
            value: ':userAddress',
        },
    };
}
