// @lit-protocol/vincent-contracts-sdk/dist/src/internal/user/UserView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllRegisteredAgentPkpEthAddresses = getAllRegisteredAgentPkpEthAddresses;
exports.getPermittedAppVersionForPkp = getPermittedAppVersionForPkp;
exports.getAllPermittedAppIdsForPkp = getAllPermittedAppIdsForPkp;
exports.getPermittedAppsForPkps = getPermittedAppsForPkps;
exports.getAllAbilitiesAndPoliciesForApp = getAllAbilitiesAndPoliciesForApp;
exports.validateAbilityExecutionAndGetPolicies = validateAbilityExecutionAndGetPolicies;
exports.getLastPermittedAppVersionForPkp = getLastPermittedAppVersionForPkp;
exports.getUnpermittedAppsForPkps = getUnpermittedAppsForPkps;
exports.isDelegateePermitted = isDelegateePermitted;
const constants_1 = require("../../constants");
const utils_1 = require("../../utils");
const pkpInfo_1 = require("../../utils/pkpInfo");
const policyParams_1 = require("../../utils/policyParams");
async function getAllRegisteredAgentPkpEthAddresses(params) {
    const { contract, args: { userPkpAddress, offset }, } = params;
    try {
        const pkpTokenIds = await contract.getAllRegisteredAgentPkps(userPkpAddress, offset);
        const pkpEthAdddresses = [];
        for (const tokenId of pkpTokenIds) {
            const pkpEthAddress = await (0, pkpInfo_1.getPkpEthAddress)({ signer: contract.signer, tokenId });
            pkpEthAdddresses.push(pkpEthAddress);
        }
        return pkpEthAdddresses;
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        if (decodedError.includes('NoRegisteredPkpsFound')) {
            return [];
        }
        throw new Error(`Failed to Get All Registered Agent PKPs: ${decodedError}`);
    }
}
async function getPermittedAppVersionForPkp(params) {
    const { contract, args: { pkpEthAddress, appId }, } = params;
    try {
        const pkpTokenId = await (0, pkpInfo_1.getPkpTokenId)({ pkpEthAddress, signer: contract.signer });
        const appVersion = await contract.getPermittedAppVersionForPkp(pkpTokenId, appId);
        if (!appVersion)
            return null;
        return appVersion;
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Get Permitted App Version For PKP: ${decodedError}`);
    }
}
async function getAllPermittedAppIdsForPkp(params) {
    const { contract, args: { pkpEthAddress, offset }, } = params;
    try {
        const pkpTokenId = await (0, pkpInfo_1.getPkpTokenId)({ pkpEthAddress, signer: contract.signer });
        const appIds = await contract.getAllPermittedAppIdsForPkp(pkpTokenId, offset);
        return appIds.map((id) => id);
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Get All Permitted App IDs For PKP: ${decodedError}`);
    }
}
async function getPermittedAppsForPkps(params) {
    const { contract, args: { pkpEthAddresses, offset, pageSize = constants_1.DEFAULT_PAGE_SIZE }, } = params;
    try {
        const pkpTokenIds = [];
        for (const pkpEthAddress of pkpEthAddresses) {
            const pkpTokenId = await (0, pkpInfo_1.getPkpTokenId)({ pkpEthAddress, signer: contract.signer });
            pkpTokenIds.push(pkpTokenId);
        }
        const results = await contract.getPermittedAppsForPkps(pkpTokenIds, offset, pageSize);
        return results.map((result) => ({
            pkpTokenId: result.pkpTokenId.toString(),
            permittedApps: result.permittedApps.map((app) => ({
                appId: app.appId,
                version: app.version,
                versionEnabled: app.versionEnabled,
                isDeleted: app.isDeleted,
            })),
        }));
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Get Permitted Apps For PKPs: ${decodedError}`);
    }
}
async function getAllAbilitiesAndPoliciesForApp(params) {
    const { contract, args: { pkpEthAddress, appId }, } = params;
    try {
        const pkpTokenId = await (0, pkpInfo_1.getPkpTokenId)({ pkpEthAddress, signer: contract.signer });
        const abilities = await contract.getAllAbilitiesAndPoliciesForApp(pkpTokenId, appId);
        return (0, policyParams_1.decodePermissionDataFromChain)(abilities);
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Get All Abilities And Policies For App: ${decodedError}`);
    }
}
async function validateAbilityExecutionAndGetPolicies(params) {
    const { contract, args: { delegateeAddress, pkpEthAddress, abilityIpfsCid }, } = params;
    try {
        const pkpTokenId = await (0, pkpInfo_1.getPkpTokenId)({ pkpEthAddress, signer: contract.signer });
        const validationResult = await contract.validateAbilityExecutionAndGetPolicies(delegateeAddress, pkpTokenId, abilityIpfsCid);
        const { policies } = validationResult;
        const decodedPolicies = (0, policyParams_1.decodePolicyParametersForOneAbility)({ policies });
        return {
            ...validationResult,
            appId: validationResult.appId,
            appVersion: validationResult.appVersion,
            decodedPolicies,
        };
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Validate Ability Execution And Get Policies: ${decodedError}`);
    }
}
async function getLastPermittedAppVersionForPkp(params) {
    const { contract, args: { pkpEthAddress, appId }, } = params;
    try {
        const pkpTokenId = await (0, pkpInfo_1.getPkpTokenId)({ pkpEthAddress, signer: contract.signer });
        const lastPermittedVersion = await contract.getLastPermittedAppVersionForPkp(pkpTokenId, appId);
        if (!lastPermittedVersion)
            return null;
        return lastPermittedVersion;
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Get Last Permitted App Version: ${decodedError}`);
    }
}
async function getUnpermittedAppsForPkps(params) {
    const { contract, args: { pkpEthAddresses, offset }, } = params;
    try {
        // Convert PKP ETH addresses to token IDs
        const pkpTokenIds = [];
        for (const pkpEthAddress of pkpEthAddresses) {
            const pkpTokenId = await (0, pkpInfo_1.getPkpTokenId)({ pkpEthAddress, signer: contract.signer });
            pkpTokenIds.push(pkpTokenId);
        }
        // Call the contract method with token IDs
        const results = await contract.getUnpermittedAppsForPkps(pkpTokenIds, offset);
        // Convert BigNumber token IDs back to strings for the response
        return results.map((result) => ({
            pkpTokenId: result.pkpTokenId.toString(),
            unpermittedApps: result.unpermittedApps.map((app) => ({
                appId: app.appId,
                previousPermittedVersion: app.previousPermittedVersion,
                versionEnabled: app.versionEnabled,
                isDeleted: app.isDeleted,
            })),
        }));
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Get Unpermitted Apps For PKPs: ${decodedError}`);
    }
}
async function isDelegateePermitted(params) {
    const { contract, args: { delegateeAddress, pkpEthAddress, abilityIpfsCid }, } = params;
    try {
        const pkpTokenId = await (0, pkpInfo_1.getPkpTokenId)({ pkpEthAddress, signer: contract.signer });
        const isPermitted = await contract.isDelegateePermitted(delegateeAddress, pkpTokenId, abilityIpfsCid);
        return isPermitted;
    }
    catch (error) {
        const decodedError = (0, utils_1.decodeContractError)(error, contract);
        throw new Error(`Failed to Check If Delegatee Is Permitted: ${decodedError}`);
    }
}
