// @account-kit/smart-contracts/dist/esm/src/msca/plugins/multisig/signer.js
import { hashMessage, hashTypedData, } from "viem";
import { MultisigPlugin, MultisigPluginAbi } from "./plugin.js";
export const multisigSignMethods = ({ client, accountAddress, signer, threshold, pluginAddress = MultisigPlugin.meta.addresses[client.chain.id], }) => {
    const get712Wrapper = async (msg) => {
        const [, name, version, chainId, verifyingContract, salt] = await client.readContract({
            abi: MultisigPluginAbi,
            address: pluginAddress,
            functionName: "eip712Domain",
            account: accountAddress,
        });
        return {
            domain: {
                chainId: Number(chainId),
                name,
                salt,
                verifyingContract,
                version,
            },
            types: {
                AlchemyMultisigMessage: [{ name: "message", type: "bytes" }],
            },
            message: {
                message: msg,
            },
            primaryType: "AlchemyMultisigMessage",
        };
    };
    const prepareSign = async (params) => {
        const messageHash = params.type === "personal_sign"
            ? hashMessage(params.data)
            : hashTypedData(params.data);
        return {
            type: "eth_signTypedData_v4",
            data: await get712Wrapper(messageHash),
        };
    };
    const formatSign = async (signature) => {
        return signature;
    };
    return {
        prepareSign,
        formatSign,
        getDummySignature: async () => {
            const [, thresholdRead] = await client.readContract({
                abi: MultisigPluginAbi,
                address: pluginAddress,
                functionName: "ownershipInfoOf",
                args: [accountAddress],
            });
            const actualThreshold = thresholdRead === 0n ? threshold : thresholdRead;
            // (uint upperLimitPreVerificationGas, uint upperLimitMaxFeePerGas, uint upperLimitMaxPriorityFeePerGas)
            // the first sig will be on "actual" with v = 32
            return ("0x" +
                "FF".repeat(32 * 3) +
                "fffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa3c" +
                "fffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c".repeat(Number(actualThreshold) - 1));
        },
        signUserOperationHash: (uoHash) => {
            return signer().signMessage({ raw: uoHash });
        },
        async signMessage({ message }) {
            const { type, data } = await prepareSign({
                type: "personal_sign",
                data: message,
            });
            return type === "personal_sign"
                ? signer().signMessage(data)
                : signer().signTypedData(data);
        },
        signTypedData: async (typedDataDefinition) => {
            const { type, data } = await prepareSign({
                type: "eth_signTypedData_v4",
                data: typedDataDefinition,
            });
            return type === "personal_sign"
                ? signer().signMessage(data)
                : signer().signTypedData(data);
        },
    };
};
