// @lit-protocol/vincent-ability-sdk/dist/src/lib/abilityCore/abilityConfig/context/resultCreators.js
"use strict";
// src/lib/abilityCore/abilityConfig/context/resultCreators.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSuccess = createSuccess;
exports.createSuccessNoResult = createSuccessNoResult;
exports.createFailure = createFailure;
exports.createFailureNoResult = createFailureNoResult;
const types_1 = require("./types");
/**
 * Wraps a success result with payload
 */
function createSuccess(result) {
    return {
        success: true,
        result,
        [types_1.YouMustCallContextSucceedOrFail]: 'AbilityResult',
    };
}
/**
 * Wraps a success result without payload
 */
function createSuccessNoResult() {
    return {
        success: true,
        [types_1.YouMustCallContextSucceedOrFail]: 'AbilityResult',
    };
}
/**
 * Wraps a failure result with payload
 */
function createFailure(result) {
    return {
        success: false,
        result,
        [types_1.YouMustCallContextSucceedOrFail]: 'AbilityResult',
    };
}
/**
 * Wraps a failure result without payload
 */
function createFailureNoResult() {
    return {
        success: false,
        result: undefined,
        [types_1.YouMustCallContextSucceedOrFail]: 'AbilityResult',
    };
}
