// @lit-protocol/vincent-app-sdk/dist/src/jwt/accessors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAppInfo = getAppInfo;
exports.getPKPInfo = getPKPInfo;
exports.getRole = getRole;
exports.getAuthentication = getAuthentication;
exports.getPublicKey = getPublicKey;
exports.getIssuerAddress = getIssuerAddress;
exports.getSubjectAddress = getSubjectAddress;
exports.getAudience = getAudience;
const constants_1 = require("./constants");
const typeGuards_1 = require("./typeGuards");
/**
 * @category API > Accessors
 */
function getAppInfo(decodedJWT) {
    if (!(0, typeGuards_1.isAppUser)(decodedJWT)) {
        throw new Error('JWT is not app specific');
    }
    return { appId: decodedJWT.payload.app.id, version: decodedJWT.payload.app.version };
}
/**
 * @category API > Accessors
 */
function getPKPInfo(decodedJWT) {
    (0, typeGuards_1.assertIsPKPSignedVincentJWT)(decodedJWT);
    return decodedJWT.payload.pkpInfo;
}
/**
 * @category API > Accessors
 */
function getRole(decodedJWT) {
    return decodedJWT.payload.role;
}
/**
 * @category API > Accessors
 */
function getAuthentication(decodedJWT) {
    (0, typeGuards_1.assertIsPKPSignedVincentJWT)(decodedJWT);
    return decodedJWT.payload.authentication;
}
/**
 * @category API > Accessors
 */
function getPublicKey(decodedJWT) {
    return decodedJWT.payload.publicKey;
}
/**
 * @category API > Accessors
 */
function getIssuerAddress(decodedJWT) {
    return decodedJWT.payload.iss;
}
/**
 * @category API > Accessors
 */
function getSubjectAddress(decodedJWT) {
    if (!decodedJWT.payload.sub) {
        throw new Error(constants_1.JWT_ERROR.INVALID_JWT + ' - Missing subject address');
    }
    return decodedJWT.payload.sub;
}
/**
 * @category API > Accessors
 */
function getAudience(decodedJWT) {
    const aud = decodedJWT.payload.aud;
    if (!aud) {
        throw new Error(constants_1.JWT_ERROR.INVALID_AUDIENCE + ' - Missing audience');
    }
    return Array.isArray(aud) ? aud : [aud];
}
