// @lit-protocol/vincent-app-sdk/dist/src/jwt/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAuthentication = exports.getPublicKey = exports.getAudience = exports.getIssuerAddress = exports.getSubjectAddress = exports.getRole = exports.getPKPInfo = exports.getAppInfo = exports.isDelegatee = exports.isAnyVincentJWT = exports.isPlatformUser = exports.isAppUser = exports.verifyVincentDelegateeJWT = exports.verifyVincentPlatformJWT = exports.verifyVincentAppUserJWT = exports.isExpired = exports.createAppUserJWT = exports.createDelegateeJWT = exports.createPlatformUserJWT = void 0;
/** The `jwt` module provides helper methods that allow you to decode and validate Vincent-specific JWTs.
 *
 * Vincent JWTs are signed using `alg: ES256K`.  The signed data is EIP191 compliant.
 *
 *
 * Vincent JWTs are issued by the Vincent Dashboard when a user provides delegation permission for your app to their agent PKP.
 * They are passed to your web app using a redirectUri which you configure on your app.
 *
 * The methods exported by the `jwt` module are low-level - you probably will just want to use {@link webAuthClient.getWebAuthClient | getWebAuthClient} to get
 * a {@link webAuthClient.WebAuthClient | WebAuthClient} which handles the redirect process, parsing the JWT from the URL, and verifying it for you.
 *
 * @packageDocumentation
 *
 */
var create_1 = require("./core/create");
Object.defineProperty(exports, "createPlatformUserJWT", { enumerable: true, get: function () { return create_1.createPlatformUserJWT; } });
Object.defineProperty(exports, "createDelegateeJWT", { enumerable: true, get: function () { return create_1.createDelegateeJWT; } });
Object.defineProperty(exports, "createAppUserJWT", { enumerable: true, get: function () { return create_1.createAppUserJWT; } });
var isExpired_1 = require("./core/isExpired");
Object.defineProperty(exports, "isExpired", { enumerable: true, get: function () { return isExpired_1.isExpired; } });
var verify_1 = require("./core/verify");
Object.defineProperty(exports, "verifyVincentAppUserJWT", { enumerable: true, get: function () { return verify_1.verifyVincentAppUserJWT; } });
Object.defineProperty(exports, "verifyVincentPlatformJWT", { enumerable: true, get: function () { return verify_1.verifyVincentPlatformJWT; } });
Object.defineProperty(exports, "verifyVincentDelegateeJWT", { enumerable: true, get: function () { return verify_1.verifyVincentDelegateeJWT; } });
var typeGuards_1 = require("./typeGuards");
Object.defineProperty(exports, "isAppUser", { enumerable: true, get: function () { return typeGuards_1.isAppUser; } });
Object.defineProperty(exports, "isPlatformUser", { enumerable: true, get: function () { return typeGuards_1.isPlatformUser; } });
Object.defineProperty(exports, "isAnyVincentJWT", { enumerable: true, get: function () { return typeGuards_1.isAnyVincentJWT; } });
Object.defineProperty(exports, "isDelegatee", { enumerable: true, get: function () { return typeGuards_1.isDelegatee; } });
var accessors_1 = require("./accessors");
Object.defineProperty(exports, "getAppInfo", { enumerable: true, get: function () { return accessors_1.getAppInfo; } });
Object.defineProperty(exports, "getPKPInfo", { enumerable: true, get: function () { return accessors_1.getPKPInfo; } });
Object.defineProperty(exports, "getRole", { enumerable: true, get: function () { return accessors_1.getRole; } });
Object.defineProperty(exports, "getSubjectAddress", { enumerable: true, get: function () { return accessors_1.getSubjectAddress; } });
Object.defineProperty(exports, "getIssuerAddress", { enumerable: true, get: function () { return accessors_1.getIssuerAddress; } });
Object.defineProperty(exports, "getAudience", { enumerable: true, get: function () { return accessors_1.getAudience; } });
Object.defineProperty(exports, "getPublicKey", { enumerable: true, get: function () { return accessors_1.getPublicKey; } });
Object.defineProperty(exports, "getAuthentication", { enumerable: true, get: function () { return accessors_1.getAuthentication; } });
