// @lit-protocol/vincent-app-sdk/dist/src/jwt/typeGuards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAppUser = isAppUser;
exports.isPlatformUser = isPlatformUser;
exports.isDelegatee = isDelegatee;
exports.isAnyVincentJWT = isAnyVincentJWT;
exports.assertIsPKPSignedVincentJWT = assertIsPKPSignedVincentJWT;
exports.assertJWTAPIVersion = assertJWTAPIVersion;
const constants_1 = require("./constants");
const index_1 = require("./core/utils/index");
/**
 * Check if a decoded JWT is an app-specific JWT (role === 'app-user')
 *
 * @category API > Type Guards
 * */
function isAppUser(decodedJWT) {
    var _a;
    return ((_a = decodedJWT.payload) === null || _a === void 0 ? void 0 : _a.role) === 'app-user';
}
/** Check if a decoded JWT is a general platform-user JWT
 *
 * @category API > Type Guards
 * */
function isPlatformUser(decodedJWT) {
    var _a;
    return ((_a = decodedJWT.payload) === null || _a === void 0 ? void 0 : _a.role) === 'platform-user';
}
/** Check if a decoded JWT is a delegatee token (role === 'app-delegatee')
 *
 * @category API > Type Guards
 * */
function isDelegatee(decodedJWT) {
    var _a;
    return ((_a = decodedJWT.payload) === null || _a === void 0 ? void 0 : _a.role) === 'app-delegatee';
}
/** Check if the decoded JWT matches any known Vincent JWT variant
 *
 * @category API > Type Guards
 * */
function isAnyVincentJWT(decodedJWT) {
    return isPlatformUser(decodedJWT) || isAppUser(decodedJWT) || isDelegatee(decodedJWT);
}
/**
 * Assert that the JWT contains expected fields for a PKP-authenticated JWT.
 * Used to validate `VincentJWT` and `VincentJWTAppSpecific` before accessing `.payload.pkp` or `.authentication`.
 *
 * @internal
 */
function assertIsPKPSignedVincentJWT(decodedJWT) {
    const { authentication, pkpInfo } = decodedJWT.payload;
    if (!(0, index_1.isDefinedObject)(authentication)) {
        throw new Error(`${constants_1.JWT_ERROR.INVALID_JWT}: Missing "authentication" field in JWT payload.`);
    }
    if (!(0, index_1.isDefinedObject)(pkpInfo)) {
        throw new Error(`${constants_1.JWT_ERROR.INVALID_JWT}: Missing "pkpInfo" field in JWT payload.`);
    }
}
function assertJWTAPIVersion(apiVersion) {
    if (constants_1.VINCENT_JWT_API_VERSION !== apiVersion) {
        throw new Error(`Invalid JWT API version. Expected ${constants_1.VINCENT_JWT_API_VERSION}, got ${apiVersion}`);
    }
}
