// @saucereon/send-tokens-ability/dist/src/lib/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeFailSchema = exports.executeSuccessSchema = exports.precheckFailSchema = exports.precheckSuccessSchema = exports.abilityParamsSchema = exports.KNOWN_ERRORS = void 0;
const zod_1 = require("zod");
exports.KNOWN_ERRORS = {
    INSUFFICIENT_BALANCE: 'INSUFFICIENT_BALANCE',
    INVALID_AMOUNT: 'INVALID_AMOUNT', // Here for example purposes
};
/**
 * Tool parameters schema - defines the input parameters for the native send tool
 */
exports.abilityParamsSchema = zod_1.z.object({
    to: zod_1.z.string().regex(/^0x[a-fA-F0-9]{40}$/, 'Invalid Ethereum address'),
    amount: zod_1.z
        .string()
        .regex(/^\d*\.?\d+$/, 'Invalid amount format')
        .refine((val) => parseFloat(val) > 0, 'Amount must be greater than 0'),
    rpcUrl: zod_1.z
        .string()
        .url('Invalid RPC URL format')
        .optional()
        .default('https://yellowstone-rpc.litprotocol.com/'),
});
/**
 * Precheck success result schema
 */
exports.precheckSuccessSchema = zod_1.z.object({
    availableBalance: zod_1.z.string(),
});
/**
 * Precheck failure result schema
 */
exports.precheckFailSchema = zod_1.z.object({
    reason: zod_1.z.union([
        zod_1.z.literal(exports.KNOWN_ERRORS['INSUFFICIENT_BALANCE']),
        zod_1.z.literal(exports.KNOWN_ERRORS['INVALID_AMOUNT']), // Here for schema example purposes
    ]),
    error: zod_1.z.string(),
});
/**
 * Execute success result schema
 */
exports.executeSuccessSchema = zod_1.z.object({
    txHash: zod_1.z.string(),
    to: zod_1.z.string(),
    amount: zod_1.z.string(),
    timestamp: zod_1.z.number(),
});
/**
 * Execute failure result schema
 */
exports.executeFailSchema = zod_1.z.object({
    error: zod_1.z.string(),
});
