// @lit-protocol/vincent-app-sdk/dist/src/abilityClient/execute/generateVincentAbilitySessionSigs.js
"use strict";
// src/abilityClient/execute/generateVincentAbilitySessionSigs.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateVincentAbilitySessionSigs = void 0;
const auth_helpers_1 = require("@lit-protocol/auth-helpers");
const constants_1 = require("@lit-protocol/constants");
/** @hidden */
const generateVincentAbilitySessionSigs = async ({ litNodeClient, ethersSigner, }) => {
    return litNodeClient.getSessionSigs({
        chain: 'ethereum',
        resourceAbilityRequests: [
            { resource: new auth_helpers_1.LitPKPResource('*'), ability: constants_1.LIT_ABILITY.PKPSigning },
            { resource: new auth_helpers_1.LitActionResource('*'), ability: constants_1.LIT_ABILITY.LitActionExecution },
            // This is required for abilities that use Vincent Wrapped Keys to be able to decrypt the previously encrypted wrapped key
            {
                resource: new auth_helpers_1.LitAccessControlConditionResource('*'),
                ability: constants_1.LIT_ABILITY.AccessControlConditionDecryption,
            },
        ],
        authNeededCallback: async ({ resourceAbilityRequests, uri }) => {
            const [walletAddress, nonce] = await Promise.all([
                ethersSigner.getAddress(),
                litNodeClient.getLatestBlockhash(),
            ]);
            const toSign = await (0, auth_helpers_1.createSiweMessageWithRecaps)({
                uri: uri || 'http://localhost:3000',
                expiration: new Date(Date.now() + 1000 * 60 * 10).toISOString(),
                resources: resourceAbilityRequests || [],
                walletAddress,
                nonce,
                litNodeClient,
            });
            return await (0, auth_helpers_1.generateAuthSig)({ signer: ethersSigner, toSign });
        },
    });
};
exports.generateVincentAbilitySessionSigs = generateVincentAbilitySessionSigs;
