// @lit-protocol/vincent-app-sdk/dist/src/abilityClient/execute/resultCreators.js
"use strict";
// src/abilityClient/execute/resultCreators.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAllowEvaluationResult = createAllowEvaluationResult;
exports.createDenyEvaluationResult = createDenyEvaluationResult;
exports.createAbilityExecuteResponseSuccess = createAbilityExecuteResponseSuccess;
exports.createAbilityExecuteResponseSuccessNoResult = createAbilityExecuteResponseSuccessNoResult;
exports.createAbilityExecuteResponseFailure = createAbilityExecuteResponseFailure;
exports.createAbilityExecuteResponseFailureNoResult = createAbilityExecuteResponseFailureNoResult;
function createAllowEvaluationResult(evaluatedPolicies, allowedPolicies) {
    return {
        allow: true,
        evaluatedPolicies,
        allowedPolicies,
    };
}
function createDenyEvaluationResult(evaluatedPolicies, allowedPolicies, deniedPolicy) {
    return {
        allow: false,
        evaluatedPolicies,
        allowedPolicies,
        deniedPolicy,
    };
}
function createAbilityExecuteResponseSuccess(params) {
    return {
        success: true,
        result: params.result,
        context: params.context,
    };
}
function createAbilityExecuteResponseSuccessNoResult(params) {
    return {
        success: true,
        result: undefined,
        context: params === null || params === void 0 ? void 0 : params.context,
    };
}
function createAbilityExecuteResponseFailure(params) {
    return {
        success: false,
        runtimeError: params.runtimeError,
        schemaValidationError: params.schemaValidationError,
        result: params.result,
        context: params.context,
    };
}
function createAbilityExecuteResponseFailureNoResult(params) {
    return {
        success: false,
        runtimeError: params.runtimeError,
        schemaValidationError: params.schemaValidationError,
        result: undefined,
        context: params.context,
    };
}
