// @lit-protocol/vincent-app-sdk/dist/src/jwt/core/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPlatformUserJWT = createPlatformUserJWT;
exports.createAppUserJWT = createAppUserJWT;
exports.createDelegateeJWT = createDelegateeJWT;
const utils_1 = require("ethers/lib/utils");
const constants_1 = require("../constants");
const base64_1 = require("./utils/base64");
const ensureHex = (s) => (0, utils_1.hexlify)(s, { allowMissingPrefix: true });
function createES256KSigner(wallet) {
    return async (data) => {
        const messageBytes = typeof data === 'string' ? (0, utils_1.toUtf8Bytes)(data) : data;
        const sig = await wallet.signMessage(messageBytes);
        const { r, s } = (0, utils_1.splitSignature)(sig);
        const sigBytes = new Uint8Array(64);
        sigBytes.set((0, utils_1.arrayify)(r), 0);
        sigBytes.set((0, utils_1.arrayify)(s), 32);
        return (0, base64_1.toBase64Url)(sigBytes);
    };
}
async function createJWS({ payload, wallet, config }) {
    const { expiresInMinutes, audience, subjectAddress, role } = config;
    const iat = Math.floor(Date.now() / 1000);
    const exp = (payload.nbf || Math.floor(Date.now() / 1000) + expiresInMinutes * 60);
    const header = { alg: 'ES256K', typ: 'JWT' };
    const iss = ensureHex(await wallet.getAddress());
    const publicKey = ensureHex(wallet.publicKey);
    const _payload = {
        ...payload,
        iat,
        exp,
        iss,
        publicKey,
        aud: audience,
        role,
        ...(subjectAddress ? { sub: subjectAddress } : {}),
        __vincentJWTApiVersion: constants_1.VINCENT_JWT_API_VERSION,
    };
    const signingInput = [
        (0, base64_1.toBase64Url)((0, utils_1.toUtf8Bytes)(JSON.stringify(header))),
        (0, base64_1.toBase64Url)((0, utils_1.toUtf8Bytes)(JSON.stringify(_payload))),
    ].join('.');
    const signature = await createES256KSigner(wallet)(signingInput);
    // JWS Compact Serialization
    // https://www.rfc-editor.org/rfc/rfc7515#section-7.1
    return [signingInput, signature].join('.');
}
/**
 * Create JWT for a platform user
 * @category API > Create
 * */
async function createPlatformUserJWT(config) {
    const { pkpWallet, pkpInfo, authentication, audience, expiresInMinutes, payload = {} } = config;
    return createJWS({
        payload: {
            ...payload,
            pkpInfo,
            authentication,
        },
        wallet: pkpWallet,
        config: { audience, expiresInMinutes, role: 'platform-user' },
    });
}
/** Create JWT for an app-scoped user
 * @category API > Create
 * */
async function createAppUserJWT(config) {
    const { app, pkpWallet, pkpInfo, authentication, audience, expiresInMinutes, payload = {}, } = config;
    return createJWS({
        payload: {
            ...payload,
            pkpInfo,
            app,
            authentication,
        },
        wallet: pkpWallet,
        config: { audience, expiresInMinutes, role: 'app-user' },
    });
}
/**
 * Creates a JWT for an app delegatee (Ethereum account that may act on behalf of a user).
 *
 * You must provide a valid `subjectAddress`, which must be a valid delegator for your Delegatee address.
 *
 * @category API > Create
 */
async function createDelegateeJWT(config) {
    const { ethersWallet, subjectAddress, audience, expiresInMinutes, payload = {} } = config;
    return createJWS({
        payload,
        wallet: ethersWallet,
        config: { audience, expiresInMinutes, subjectAddress, role: 'app-delegatee' },
    });
}
